---
title: "4_1_figure_1_2_3"
author: "Victor Yuan"
date: "14/11/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

# Setup

## Libraries

```{r}
#install.packages('ggforce')
library(RColorBrewer)
library(egg)
library(scales)
library(here)
library(ggrepel)
library(forcats)
library(GGally)
library(irlba)
library(tidyverse)
font_size <- 10
```

## Data

```{r}
# parent directory
base_path <- here("data", "main", "interim")

# sample data
pDat <- readRDS(file.path(base_path, '3_1_pDat_filt.rds'))

# methylation data
betas <- readRDS(file.path(base_path, '3_1_betas_filt.rds'))

# annotation
anno <- readRDS(file.path('Z:/', 
                          'Victor', 'Repositories', 'EPIC_annotation',
                          'hg19_epic_annotation.rds'))
# snp betas cor matrix
snp_betas_cor <- readRDS(file.path(base_path, '1_1_cor_matrix_snp_betas.rds'))

#color code
color_code <- readRDS(file.path(base_path, '2_3_color_code.rds'))
color_code_tissue <- setNames(color_code$Colors_Tissue, color_code$label)
colors <- color_code_tissue[unique(pDat$Tissue)]

#list of genes for search list
gene <- tibble(gene = str_split(anno$genes_symbol, ', ') %>%
                 unlist() %>%
                 unique() %>%
                 na.omit() %>%
                 sort())

#pull out cpg name for search list use
cpg <- unique(anno$cpg) 
cpg <- cpg[cpg %in% rownames(betas)] %>% as_tibble()

pDat_filt <- pDat %>% 
  filter(maternal_contamination_norm_flip < 0.35,
         !Sample_Name %in% c('PM364_hofb_cs', 'PL293_v_R2', 'PM366_vc_R2', 'P131_hofb_cs', 
                             'PM324_V4', 'PM324_V1', 'PM139_vc', 'PM77_vc'),
         !Tissue %in% c('Villi maternal', 'Trophoblasts enz', 'Mixture cs', 
                        'Dead Cells and Lymphocytes cs'),)
# filter to first trimester
betas_filt <- betas[,pDat_filt$Sample_Name]


enrich <- readRDS(file.path(base_path, '2_4_enrich.rds'))
kegg <- enrich$kegg
go <- enrich$GO
```
# Figure 2 gene plot

```{r}
stat_plot <- function(cpg_name = NULL, gene_name = NULL, trimester = 'Third'){
  
  ######## FILTER ANNOTATION TO RELEVANT CPGS, DEPENDING ON INPUT #################
  if (is.null(cpg_name)){
     
    # If a gene symbol is supplied,
    # filter to gene in annotation
    anno_gene <- anno %>%
      filter(grepl(paste0('\\<', gene_name, '\\>'), genes_symbol),
             cpg %in% rownames(betas)) %>%
      arrange(desc(start))
    
    # stop if gene symbol is invalid
    if (nrow(anno_gene) == 0) {
      stop('No cpgs found, maybe they were filtered out.')
    }
  } else if (gene_name == ''){
      
      # If a cpg is provided,
      # filter annotation to cpgs
      anno_gene <- anno %>%
        filter(cpg %in% cpg_name,
               cpg %in% rownames(betas)) %>% 
        arrange(chr, desc(start))
      
      # stop if no cpgs were found
      if (nrow(anno_gene) == 0) {
        stop('None of the selected CpGs exist in our processed data.')
      }
      
      if (length(cpg_name) < length(intersect(cpg_name, rownames(betas)))) {
        print(paste0('The following selected CpGs do not exist in our processed data:\n', 
                     setdiff(cpg_name, rownames(betas))))
      }
      
  } else {
      stop('Enter valid cpg(s) OR one specific gene.')
  }
  
  ######## SETUP INPUT TO PLOTS ###############
  #
  # now we wrangle the betas/annotations/pdata information into a format usable for plotting
  #
  betas_gene <- t(betas[anno_gene$cpg,,drop = FALSE]*100) %>% as_tibble %>% 
    mutate(Sample_Name = colnames(betas)) %>%
    
    # reshape into longer format
    pivot_longer(cols = -Sample_Name,
                 names_to = 'cpg', 
                 values_to = 'beta')  %>%
    
    # add tissue and trimester info
    left_join(pDat %>% select(Sample_Name, Trimester, Tissue), by = 'Sample_Name') %>%
    
    # calculate mean and sd for each cpg for each group
    group_by(Tissue, Trimester, cpg) %>%
    summarize(mean = mean(beta),
              sd = sd(beta)) %>%
    mutate(lower = mean-sd, upper = mean+sd) %>%
    
    # add cpg info
    left_join(anno_gene, by = 'cpg') %>%
    
    # order on cpg position
    ungroup() %>%
    arrange(desc(start)) %>%
    mutate(cpg = factor(cpg, levels = unique(cpg)),
           Trimester_Tissue = paste0(Trimester, ' - ', Tissue)) %>%
    filter(Trimester == trimester)
  
  ##### GENERATE INDIVIDUAL PLOT TRACKS #####
  font_size <- font_size+0.6
  point_size <- 0.5
  
  # Generate  methylation track
  p1 <- betas_gene %>%
    ggplot() +
    
    # geom_ribbon is to generate the alternating shaded background
    geom_ribbon(aes(x = as.numeric(cpg), ymin = 0, ymax = 25),
                fill = 'grey', alpha = 0.25)+
    geom_ribbon(aes(x = as.numeric(cpg), ymin = 50, ymax = 75),
                fill = 'grey', alpha = 0.25)+
    
    # geom_linerange and geom_point is for the actual methylation data
    geom_linerange(alpha = 0.5, size = point_size, 
                   aes(x = cpg, ymin =lower, ymax = upper, color = Tissue),
                   show.legend = FALSE) +
    geom_point(alpha = 0.75, aes(x = cpg, y = mean, color = Tissue), size = point_size*2) +
    
    # we want to facet by trimester
    facet_wrap(vars(Trimester), ncol = 1, labeller = label_both) +
    
    # cosmetics
    theme_bw(base_size = font_size) +
    theme(axis.text.x = element_blank(), 
          axis.text.y = element_text(size = 6),
          axis.ticks = element_blank(),
          axis.title.y = element_text(angle = 0, vjust = 0.5, hjust = 0, size = font_size),
          
          legend.text = element_text(),
          legend.key.size = unit(0, 'lines'),
          
          panel.grid.minor.y = element_blank(),
          panel.grid.major.y = element_blank(),
          panel.spacing.y = unit(0.5, 'cm'),
          panel.border = element_blank(),
          axis.line.y = element_line(color = 'black', size = 0.5, linetype = 'solid'),
          
          strip.background = element_blank(),
          strip.text = element_text(face = 'bold', hjust = 0),
          
          plot.margin=margin(l=-0.1, unit="cm")) +
    scale_y_continuous(limits = c(0, 100), 
                       expand = c(0, 0), 
                       breaks = c(0, 50, 100),
                       labels = function(x)paste0(x, '%')) +
    scale_x_discrete(expand = c(0.01,0.01))+
    scale_color_manual(values= color_code_tissue[unique(betas_gene$Tissue)],
                       guide = guide_legend(override.aes = list(size = point_size*4)),
                       labels = function(x)gsub(' cs', '', x)) +
    labs(y = 'DNA\nmethylation', 
         x = '', 
         color = '', 
         title = if_else(!is.null(gene_name), gene_name, '')) 
    
  # plot cpg annotations
  p2 <- betas_gene %>% 
    
    # First we need to process the annotation data
    mutate(# absent/presence for different genomic elements
           enhancer = !is.na(enhancers_id),
           pmd = !is.na(pmd_id),
           imprinted_dmr_general = 
             !is.na(imprint_tissue_specificity) & imprint_tissue_specificity == 'other',
           imprinted_dmr_placenta = 
             !is.na(imprint_tissue_specificity) & imprint_tissue_specificity ==
             'placental-specific')%>%
    
    select(cpg, enhancer, pmd, 
           imprinted_dmr_general, imprinted_dmr_placenta, cpg_id, start) %>%
    mutate_if(is.logical, as.character) %>%
    
    # reshape
    pivot_longer(cols = c(-cpg, -start),
               names_to = 'cpg_element', 
               values_to = 'presence') %>%
    distinct() %>%
    
    # arrange plot orde for each track
    arrange(desc(start)) %>%
    mutate(cpg = factor(cpg, levels = unique(cpg)),
           cpg_element = factor(cpg_element, 
                                levels = c('enhancer', 'pmd', 'imprinted_dmr_general',
                                           'imprinted_dmr_placenta', 'cpg_id')),
           presence = factor(ifelse(presence == "TRUE", 'Present',
                                    ifelse(presence == "FALSE", 'Absent', presence)),
                             levels = c('sea', 'shelf', 'shore', 'island',
                                        'Present', 'Absent', ' ', '  ')),
           # for facet label
           group = 'Annotations') %>%
    
    # plot code
    ggplot(aes(x = cpg, y = cpg_element, fill = presence)) +
    geom_tile(color = '#fcfcfc') +
    guides(fill=guide_legend(ncol=3, override.aes = list(colour = 'black'))) +
    facet_wrap(vars(group), ncol = 1) +
    theme_bw(base_size = font_size) +
    theme(axis.text.x = element_blank(),
          axis.ticks = element_blank(),
          axis.text.y = element_text(vjust = 0.5, size = 6),
          
          legend.text = element_text(),
          legend.key.height = unit(0.25, 'cm'),
          legend.key.width = unit(0.25, 'cm'),
          
          plot.margin = margin(l=-0.1,unit="cm"),
          panel.border = element_blank(),
          axis.line.y = element_line(color = 'black', size = 0.5, linetype = 'solid'),
          
          strip.background = element_blank(),
          strip.text = element_text(face = 'bold', hjust = 0)) +
    scale_fill_manual(values = c('Present' = '#cccccc', 'Absent' = '#f7f7f7', 
                                 ' ' = 'white', '  ' = 'white',
                                 'sea' = '#ffffcc', 'shelf' = '#a1dab4', 'shore' = '#41b6c4',
                                 'island' = '#225ea8'), drop = F,
                      labels = stringr::str_to_title) +
    scale_y_discrete(expand = c(0,0),
                     labels = function(x)gsub('cpg_id', 'Relation to CpG Islands',
                       gsub('imprinted_dmr_placenta', 'Imprinted DMR (Placenta)',
                       gsub('imprinted_dmr_general', 'Imprinted DMR (General)',
                       gsub('pmd', 'Placental PMD',
                       gsub('enhancer', 'Enhancer', 
                            x)))))) +
    scale_x_discrete(expand = c(0,0)) +
    guides(fill=guide_legend(ncol=2, override.aes = list(colour = 'white'))) +
    labs(x = '', y = '', fill = '') 
  
  # transcript annotations
  
  # cpgs can have more than one mapping to a gene element per transcript
  # cpgs mapped to intronexonboundaries can also be part of exons and introns
  # 5 UTR and 3 UTR can also overlap with exons and introns
  
  # If there are multiple mappings then only the following will be displayed in order of priority:
  # 5UTR / 3UTR, exon / intron, intronexonboundary
  
  p3 <- betas_gene %>% 
    select(cpg, end, genes_id, genes_tx_id, genes_symbol) %>%
    distinct() %>%
    
    # split comma separate list and put each entry into a new row
    mutate(genes_tx_id = str_split(genes_tx_id, ', '),
           genes_id = str_split(genes_id, ', '),
           genes_symbol = str_split(genes_symbol, ', ')) %>%
    unnest(c(genes_tx_id, genes_id, genes_symbol)) %>%
    
    # paste together transcript ID and genes symbol
    mutate(genes_tx_id_symbol = if_else(genes_symbol != 'no_associated_gene', 
                                        paste0(genes_tx_id, ' (', genes_symbol, ')'),
                                        genes_tx_id)) %>%
    
    ### ORDER TRANSCRIPTS IN PLOT 
    #
    # note that the first level appears on the bottom of the graph (y = 0 if it was continuous)
    
    # calculate earliest end of each symbol
    group_by(genes_symbol) %>%
    mutate(genes_symbol_end = min(end)) %>%
    ungroup() %>%
    
    # reorder gene symbols by their earliest end position
    mutate(genes_symbol = fct_reorder(genes_symbol, genes_symbol_end)) %>%
    
    # reorder transcripts based on gene symbol levels and then by their end position
    arrange(genes_symbol, end) %>%
    mutate(genes_tx_id_symbol = 
             factor(genes_tx_id_symbol, levels = unique(genes_tx_id_symbol)) %>%
             fct_explicit_na()) %>%
    
    ####
    # fill all combinations of cpg and gene_tx_id with NA if missing, so that they show in plot
    complete(cpg, genes_tx_id_symbol) %>%
    
    # filter out to unique mappings using the priority described above
    # strategy is to order by factor level and take the first occurence
    mutate(genes_id = factor(genes_id, 
                             levels = c('1to5kb', 'promoter', '5UTR', '3UTR', 'exon',
                                        'intron', 'intronexonboundary', 'intergenic'))) %>%
    
    group_by(cpg, genes_tx_id_symbol) %>%
    arrange(genes_id)  %>%
    dplyr::slice(1) %>%
    
    # reorder factor levels for appearance in plot legend
    mutate(genes_id = factor(genes_id, 
                             levels = c('1to5kb', 'promoter', '5UTR', 'exon', 'intron',
                                        'intronexonboundary', '3UTR', 'intergenic')),
           group = 'Transcripts') %>%
    
    #plot
    ggplot(aes(x = cpg, y = genes_tx_id_symbol, fill = genes_id)) +
    geom_tile(color = '#fcfcfc', alpha = 0.6) +
    facet_wrap(vars(group), ncol = 1) +
    guides(fill=guide_legend(ncol=2, override.aes = list(alpha=0.7, 
                                                         color = 'black'))) +
    theme_bw(base_size = font_size) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 6), 
          axis.text.y = element_text(vjust = 0.5, size = 6),
          axis.ticks = element_blank(),
          
          plot.margin = margin(l=-0.1,unit="cm"),
          
          legend.text = element_text(margin = margin(t = -1, b = -1)),
          legend.key.height = unit(0.25, 'cm'),
          legend.key.width = unit(0.25, 'cm'),
          
          panel.grid = element_blank(),
          panel.border = element_blank(),
          axis.line = element_line(color = 'black'),
          axis.line.y = element_line(color = 'black', size = 0.5, linetype = 'solid'),
          
          strip.background = element_blank(),
          strip.text = element_text(face = 'bold', hjust = 0)) +
    scale_fill_brewer(na.value = '#f7f7f7', breaks = c('1to5kb', 'promoter', '5UTR', 'exon', 
                                                        'intron', 'intronexonboundary', '3UTR',
                                                        'intergenic'),
                      palette = 'Paired', direction = -1) +
    scale_y_discrete(expand = c(0,0), 
                     labels = function(x)(if_else(x == '(Missing)', '', x))) +
    scale_x_discrete(expand = c(0,0)) +
    guides(fill=guide_legend(override.aes = list(colour = 'white'))) +
    labs(x = '', y = '', fill = '')
  
  # the size of each plot depends on the number of unique transcripts
  # More transcripts means a longer transcript track
  # so the plot height depends on the number of transcripts
  
  # number of transcripts
  n_trans <- betas_gene$genes_tx_id %>% str_split(', ') %>% unlist() %>% unique %>% length()
  
  # size of p1
  p1_h <- 18
  
  # combine tracks plot
  ggarrange(p1, p2, p3, ncol = 1, heights = c(p1_h-(n_trans-7), 5, n_trans))
}

p <- stat_plot(gene_name = 'DNMT1', trimester = 'Third');p #h: 5, w:6.5

```

# Figure S* gene plots

```{r}

ps1 <- stat_plot(gene_name = 'DNMT3A', trimester = 'Third');ps1 #h:3.5, w:10

ps2 <- stat_plot(gene_name = 'DNMT3B', trimester = 'Third');ps2 #h:3.5, w: 4.25
ps3 <-  stat_plot(gene_name = 'DNMT3L', trimester = 'Third');ps3 #h:3.5, w: 3.5


ps4 <-  stat_plot(gene_name = 'TET1', trimester = 'Third');ps4 #h:3.5, w: 5
```

# Figure 1 enrichment

```{r}
base_path <- file.path('data', 'main','interim')

tests <- readRDS(here(base_path, '2_5_enrich_tests.rds'))


plot_bar <- function(x){
  x %>%
    # ordering in plot
    mutate(# linerange ymin ymax values
           ymin = pmin(Observed_p_in, Expected_p_in),
           ymax = pmax(Observed_p_in, Expected_p_in)) %>%
    
    ggplot() +
    
    # for cell type dmc enrichment lines and points
    geom_linerange(aes(x = genomic_feature, 
                        ymin = ymin, ymax = ymax, 
                        color = Celltype),
             stat = 'identity', 
             position = position_dodge(width = 0.75),
             #fatten = 0.01, 
             size = 1) +
    #geom_point(aes(x = genomic_feature,  y = Observed_p_in,
    #               color = Celltype, alpha = bonferroni001),
    #         stat = 'identity',
    #         position = position_dodge(width = 0.75),
             #fatten = 0.01, 
    #         size = 0.03) +
    
    # for expected frequency in legend
    geom_linerange(aes(x = genomic_feature, 
                      ymin = Expected_p_in, 
                      ymax = Expected_p_in, 
                      linetype = 'Expected frequency'),
                   size = 0.5) +
    
    # for generating the expected frequency bars
    geom_errorbar(aes(x = genomic_feature, 
                      ymin = Expected_p_in, 
                      ymax = Expected_p_in, 
                      linetype = 'Expected frequency'),
                  show.legend = FALSE) +

    # add significance asterisks
    geom_text(data = . %>%
                filter(bonferroni001),
              aes(x = genomic_feature,
                  y = ymax + 0.035, 
                  group = Celltype, 
                  size = bonferroni001),
              position = position_dodge(0.85), 
              label = '*') +
    
    facet_grid(genomic_feature_category~ Direction , 
               labeller = labeller(Direction = c('Hyper' = 'More methylated',
                                                 'Hypo' = 'Less methylated')),
               scales = 'free', space = 'free', drop = TRUE) +
    coord_flip() +
    theme_bw(base_size = font_size+1) + 
    
    theme(legend.position = 'top', 
          legend.spacing.x = unit(-0.2, 'cm'),
          strip.background = element_blank(),
          strip.text.x = element_text(face = 'bold'),
          strip.text.y = element_blank(),
          axis.ticks.y = element_blank(),
          axis.text = element_text(font_size*2),
          panel.grid.major.y = element_blank(),
          panel.border = element_rect(color = 'grey'),
          panel.spacing.y = unit(0.15, 'cm')) +
    

    labs(color = '', x = '', y = '', shape = '', linetype = '', size = '') +
    scale_color_manual(values = colors,
                       breaks = '',
                       guide = guide_legend(override.aes = list(size = 0.75),
                                            order = 3)) +
    scale_linetype(guide =guide_legend(order = 2,
                                       override.aes = list(size = 1))) +
    scale_size_manual(values = c('TRUE' = 1.5), 
                      na.translate = F, 
                      labels = c('Enriched\n(bonferroni p < 0.001)'), 
                      guide = guide_legend(override.aes = list(size = 4),
                                            order = 1)) +
    scale_y_continuous(limits = c(0,1), expand = c(0.075,0),
                       breaks = c(0, 0.5, 1),
                       labels = function(x)scales::percent(x, accuracy = 1)) 
    #scale_alpha_discrete(range = c(0.6,1), breaks = c('TRUE', 'FALSE'), guide = FALSE)
}

p1 <- tests %>% 
  filter(Trimester == 'Third',
         genomic_feature_category %in% c('pmd', 'gene', 'enhancer', 'cpg_island')) %>%
  arrange(genomic_feature_category, genomic_feature) %>%
  mutate(genomic_feature = as_factor(genomic_feature)) %>%
  plot_bar();p1 #h:4.25, w: 3.25

p2 <- tests %>% 
  filter(Trimester == 'First',
         genomic_feature_category %in% c('pmd', 'gene', 'enhancer', 'cpg_island')) %>%
  arrange(genomic_feature_category, genomic_feature) %>%
  mutate(genomic_feature = as_factor(genomic_feature)) %>%
  plot_bar() 
```


# Number of DMCs

```{r}
dmcs <- readRDS(file.path(base_path, '2_4_dmcs.rds'))

# define cutoffs
p_thresh <- 0.01
b_thresh <- 0.25

# count number of significant cpgs per cell
dmcs_sig <- dmcs %>% 
  group_by(Group1) %>% 
  summarize(b001db50_all = sum(bonferroni < p_thresh & abs(delta_b) > b_thresh), 
            b001db50_hypo = sum(bonferroni < p_thresh & delta_b < -b_thresh),
            b001db50_hyper = sum(bonferroni < p_thresh & delta_b > b_thresh)) %>%
  separate(Group1, into = c('Trimester', 'Celltype'), sep = '\\.') %>%
  gather('group', 'n_cpg', -Trimester, -Celltype) %>%
  separate(group, into = c('pvalue', 'delta_b')) %>%
  mutate(Group1_label = paste0(Trimester, ' - ', case_when(
    Celltype == 'Endo_cs' ~ 'Endothelial cs',
    Celltype == 'Hofb_cs' ~ 'Hofbauer cs',
    Celltype == 'Strom_cs' ~ 'Stromal cs',
    Celltype == 'Troph_cs' ~ 'Trophoblasts cs'
  )))

plot_object <- dmcs_sig %>% 
  # calculate proportions
  spread(key = delta_b, value = n_cpg) %>%
  select(-Trimester, -Celltype, -pvalue) %>%
  pivot_longer(cols = c(hyper, hypo),
               names_to = 'direction',
               values_to = 'n') %>%
  separate(Group1_label, into = c('Trimester', 'Celltype'), sep = ' - ') %>% {
    
    ggplot(data = ., aes(x = Celltype)) +
      geom_bar(stat = 'identity', aes(y = n, fill = direction)) +
      geom_text(data = select(., Trimester:all) %>%
                  distinct(),
                aes(label = scales::comma(all), y = all),
                size = 3,
                hjust = 0,
                nudge_y = 15000) +
      facet_wrap(~Trimester, ncol = 1, labeller = 'label_both') +
      scale_x_discrete(labels = function(x)gsub(' cs', '', x)) +
      scale_y_continuous(labels = scales::number_format(), 
                         expand = c(0, 0),
                         limits = c(0, 200000)) +
      scale_fill_manual(values = c('hyper' = '#e6e6e6', 'hypo' = 'grey'), 
                        labels = c('Less methylated', 'More methylated'),
                        breaks = c('hypo', 'hyper'),
                        guide = guide_legend(override.aes = list(size = 1))) +
      guides(fill = guide_legend(override.aes = list(size = 0.1)))+
      theme_bw(base_size = font_size+0.5) +
      theme(panel.border = element_blank(),
            axis.text.x = element_blank(),
            axis.ticks = element_blank(),
            panel.grid = element_blank(),
            strip.background = element_blank(),
            strip.text = element_text(hjust = 0, face = 'bold'),
            legend.position = 'top',
            legend.box.margin=margin(-5,-5,-5,-5),
            legend.margin=margin(0,0,0,0),
            legend.key.size = unit('0.4', 'cm'),
            legend.text = element_text(vjust = 0.5)) +
      coord_flip() +
      labs(x = '', y = '', fill = '') };plot_object #h: 2.75, w: 3

dmcs_sig %>% 
  # calculate proportions
  spread(key = delta_b, value = n_cpg) %>%
  mutate(hyper_p = hyper*100/all,
         hypo_p = hypo*100/all) %>%
  select(-Trimester, -Celltype, -pvalue) %>%
  
  mutate(Hypermethylated = paste0(hyper, ' (', prettyNum(hyper_p, digits = 2), '%)'),
         Hypomethylated = paste0(hypo, ' (', prettyNum(hypo_p, digits = 2), '%)')) %>%
  
  # specify color of cells
  mutate(Hypermethylated = cell_spec(Hypermethylated, color = 'white', 
                                     background = spec_color(hyper_p, end = 0.5, direction = -1)),
         Hypomethylated = cell_spec(Hypomethylated, color = 'white', 
                                     background = spec_color(hypo_p, end = 0.5, direction = -1))) %>%
  separate(Group1_label, into = c('Trimester', 'Celltype'), sep = ' - ') %>%
  select(-(hyper:hypo_p)) %>%
  dplyr::rename(Num_sig = all) %>%
  kable(escape = F) %>%
  kable_styling(full_width = F) %>%
  column_spec(1, bold = T) %>%
  collapse_rows(columns = 1)
```

# DMCs genomic viewer

```{r}
#install.packages('ggforce')

library(annotatr)
library(RColorBrewer)
library(egg)
library(scales)
library(here)
library(ggrepel)
library(forcats)
library(GGally)
library(irlba)
library(tidyverse)
font_size <- 10

# parent directory
base_path <- here("data", "main", "interim")

# sample data
pDat <- readRDS(file.path(base_path, '3_1_pDat_filt.rds'))

# methylation data
betas <- readRDS(file.path(base_path, '3_1_betas_filt.rds'))

# annotation
anno <- readRDS(file.path('Z:/', 
                          'Victor', 'Repositories', 'EPIC_annotation',
                          'hg19_epic_annotation.rds'))

#color code
color_code <- readRDS(file.path(base_path, '2_3_color_code.rds'))
color_code_tissue <- setNames(color_code$Colors_Tissue, color_code$label)

#list of genes for search list
gene <- tibble(gene = str_split(anno$genes_symbol, ', ') %>%
                 unlist() %>%
                 unique() %>%
                 na.omit() %>%
                 sort())

#pull out cpg name for search list use
cpg <- unique(anno$cpg) 
cpg <- cpg[cpg %in% rownames(betas)] %>% as_tibble()

pDat_filt <- pDat %>% 
  filter(maternal_contamination_norm_flip < 0.35,
         !Sample_Name %in% c('PM364_hofb_cs', 'PL293_v_R2', 'PM366_vc_R2', 'P131_hofb_cs', 
                             'PM324_V4', 'PM324_V1', 'PM139_vc', 'PM77_vc'),
         !Tissue %in% c('Villi maternal', 'Trophoblasts enz', 'Mixture cs', 
                        'Dead Cells and Lymphocytes cs'),)
# filter to first trimester
betas_filt <- betas[,pDat_filt$Sample_Name]
```

## setup annotation

```{r}
#ect annotations for intersection with regions
# Note inclusion of custom annotation, and use of shortcuts
annots <- c('hg19_cpgs', 'hg19_basicgenes', 'hg19_genes_intergenic',
           'hg19_genes_intronexonboundaries')

# Build the annotations (a single GRanges object)
annotations <- build_annotations(genome = 'hg19', annotations = annots)
annotations <- annotations %>%
  as_tibble() %>%
  mutate(type = gsub('genes_', '',
                     gsub('hg19_', '', type)),
         type = fct_relevel(type,
                            c('1to5kb', 'promoters', '5UTRs', 
                              'exons', 'introns', 'intronexonboundaries', 
                              '3UTRs',
                              'cpg_islands', 'cpg_shores', 'cpg_shelves', 'cpg_inter'))) 

anno <- anno %>%
  filter(cpg %in% rownames(betas_filt))
```

## setup function

```{r}
igv <- function(gene_name = NULL, 
                
                start_extend = 2500, # how far to extend the plot left, in genomic coordinates 
                end_extend = 2500,   # how far to extend right
                
                show_transcript_gene = TRUE, # show transcripts associated gene symbol
                breaks_width = 15000, # determines the spacing of x axis breaks
                
                highlight_region = NULL, # a df with start end columns, indicating regions to highlight
                
                trimester = 'Third', # can be Third, First, or Both
                sex = 'Both', # can be Male, Female, or Both
                
                plot_sizes = c(3,1), # a 2-element numeric vector indicating the relative plot heights 
                                    # of the methylation plot to the transcript plot
                transcript_height = 1 # scale factor for height of transcript elements. e.g. 1.1 = 110%
                
                # required components:
                # pDat_filt with columns Sample_Name, Trimester, Sex, and Tissue
                # betas_filt data.frame with column names == pDat$Sample_Name
                # anno data frame that is contains a column cpg, genes_symbol, chr, start, end
                # the cpgs in anno and betas_filt need to be exactly the same, and same order
                # annotation data.frame that I built from annotatr, and slightly cleaned
                # colors, a named vector corresponding to the color codes for each celltype
                ) {
  
  # subset to relevant samples
  if (trimester %in% c('Third', 'First')) {
    pDat_filt <- pDat_filt %>%
      filter(Trimester == trimester)
  }
  
  if (!trimester %in% c('Third', 'First', 'Both')) {
    stop('trimester must be set to one of "Third", "First", or "Both"')
  }
  
  if (sex %in% c('Male', 'Female')) {
    pDat_filt <- pDat_filt %>%
      filter(Sex == sex)
  }
  
  if (!sex %in% c('Male', 'Female', 'Both')) {
    stop('sex must be set to one of "Male", "Female", or "Both"')
  }
  
   
  # subset to cpgs associated with gene
  anno_gene_subset <- anno %>%
    filter(grepl(paste0('\\<', gene_name, '\\>'), genes_symbol)) %>% 
    dplyr::select(chr, start, end) 
  
  c <- unique(anno_gene_subset$chr)
  s <- min(anno_gene_subset$start) # first cpg
  e <- max(anno_gene_subset$start) # last cpg
  s_extend <- s - start_extend
  e_extend <- e + end_extend
  
  region_of_interest <- anno %>%
    filter(chr == c, start > s_extend, end < e_extend)
  
  # subset betas
  b <- betas_filt[region_of_interest$cpg, pDat_filt$Sample_Name] %>%
    as.data.frame() %>%
    bind_cols(cpg = rownames(.), .) %>%
    pivot_longer(cols = -cpg,
                 names_to = 'Sample_Name',
                 values_to = 'beta') %>%
    left_join(anno, by = 'cpg') %>% 
    right_join(pDat_filt, by = 'Sample_Name') 
  
  
  p1 <- b %>%
    
    # calculate mean beta for each cpg for each tissue
    group_by(cpg, Tissue) %>%
    summarize(mean = mean(beta), start = dplyr::first(start)) %>%
    
    # plot methylation means
    ggplot(aes(x = start)) +
    geom_linerange(ymin = 0, aes(ymax = mean, color = Tissue)) +
    
    facet_wrap(vars(Tissue), ncol = 1, strip.position = 'left',
               labeller = labeller(Tissue = function(x)gsub(' cs', '', x))) +
    
    theme_bw(base_size = font_size) +
    theme(strip.background = element_blank(),
          strip.text.y.left = element_text(angle = 0, hjust = 1),
          strip.placement = 'outside',
          axis.ticks = element_blank(),
          axis.text.x = element_blank(),
          axis.title.x = element_blank(),
          axis.line = element_line(size = 0.5),
          panel.border = element_blank(),
          panel.grid.minor = element_blank(),
          panel.spacing = unit(0, 'lines'),
          #panel.grid = element_blank()
          ) +
    scale_y_continuous(limits = c(0,1), breaks = c(0,1), expand = c(0,0), labels = NULL) +
    scale_x_continuous(limits = c(s_extend, e_extend), expand = c(0,0), 
                       breaks = breaks_width(breaks_width),
                       labels = scales::number) +
    scale_color_manual(values = color_code_tissue, guide = 'none') +
    labs(y = '', x = '', title = gene_name)
  
  p2 <- annotations %>%
    filter(seqnames == as.character(c), start > s_extend, end < e_extend,
           type %in% c('exons', 'introns', 'promoters')) %>%
    
    mutate(symbol = ifelse(is.na(symbol), '', symbol)) %>%
    ggplot(aes(x = start, xend = end)) +
    geom_segment(y= 0.5, yend = 0.5, aes(size = type, color = type)) +
    theme_bw(base_size = font_size)+
    theme(strip.background = element_blank(),
          strip.text.y.left  = element_text(angle = 0),
          strip.placement = 'outside',
          legend.position = 'bottom',
          plot.title =  element_blank(),
          plot.subtitle = element_blank(),
          axis.text.y = element_blank(),
          axis.ticks = element_blank(),
          axis.line = element_line(size = 0.5),
          panel.spacing = unit(0, 'lines'),
          panel.grid.major.y = element_blank(),
          panel.grid.minor.y = element_blank(),
          panel.grid.minor.x = element_blank(),
          panel.border = element_blank(),
          legend.key.size = unit(0.5, 'cm'),
          legend.spacing = unit(0.25, 'cm')) +
    scale_size_manual(values = c('exons' = 3, 'introns' = 1, 'intronexonboundaries' = 3,
                                 '5UTRs' = 5, '3UTRs' = 6,
                                 '1to5kb' = 2,
                                 'promoters' = 2,
                                 
                                 'cpg_inter' = 1,
                                 'cpg_islands' = 5,
                                 'cpg_shores' = 4,
                                 'cpg_shelves' = 3)*transcript_height) +
    scale_color_manual(values = c('exons' = 'black', 'introns' = 'black', 'intronexonboundaries' = 'black',
                                 '5UTRs' = 'grey', '3UTRs' = 'grey',
                                 '1to5kb' = 'grey',
                                 'promoters' = 'forestgreen',
                                 
                                 'cpg_inter' = 1,
                                 'cpg_islands' = 5,
                                 'cpg_shores' = 4,
                                 'cpg_shelves' = 3)) +
    scale_x_continuous(limits = c(s_extend,e_extend), 
                       expand = c(0,0), 
                       labels = scales::number,
                       breaks = breaks_width(breaks_width)) +
    scale_y_continuous(expand = c(0,0)) +
    labs(x = 'position', size = '', color = '')
  
  if (show_transcript_gene) {
    p2 <- p2 + 
      facet_wrap(~symbol + tx_id, ncol = 1, strip.position = 'left') 
  } else if (show_transcript_gene == FALSE) {
    p2 <- p2 + 
      facet_wrap(~tx_id, ncol = 1, strip.position = 'left') 
  }
  
  if (!is.logical(show_transcript_gene)) {
    stop('show_transcript_gene must be set to TRUE or FALSE')
  }
  
  if (!is.null(highlight_region)) {
    p1 <- p1 + 
      geom_rect(data = highlight_region,
                aes(xmin = start, xmax = end),
                ymin = 0, ymax = 1,
                alpha = 0.1)
    p2 <- p2 + 
      geom_rect(data = highlight_region,
                aes(xmin = start, xmax = end),
                ymin = 0, ymax = 1,
                alpha = 0.1)
  }
  egg::ggarrange(p1,p2, ncol = 1, heights = plot_sizes) #H:4.25, #w: 3.25

}
```

## GCM1

```{r}
igv(gene_name = 'GCM1')
```

## INHBA

```{r}
anno %>%
  filter(grepl('INHBA$', genes_symbol)) %>%
  pull(start) %>% range

c <- 'chr7'
s <- 41731552
e <- 41742795
s_extend <- s - 2500
e_extend <- e + 7500


region_of_interest <- anno %>%
  filter(chr == c, start > s_extend, end < e_extend,
         cpg %in% rownames(betas_filt))

# add betas, annotation, and pdata
b <- betas_filt[region_of_interest$cpg,pDat_filt$Sample_Name] %>%
  as.data.frame() %>%
  bind_cols(cpg = rownames(.), .) %>%
  pivot_longer(cols = -cpg,
               names_to = 'Sample_Name',
               values_to = 'beta') %>%
  left_join(anno, by = 'cpg') %>% 
  right_join(pDat_filt, by = 'Sample_Name') 



p1 <- b %>%
  filter(Trimester == 'Third') %>%
  group_by(cpg, Tissue) %>%
  summarize(mean = mean(beta), start = dplyr::first(start)) %>%
  ggplot(aes(x = start, color = Tissue)) +
  geom_linerange(ymin = 0, aes(ymax = mean)) +
  annotate(geom = 'rect', ymin = 0, ymax = 1, alpha = 0.1,
           xmin =41745000-200, 
           xmax = 41745000+1250
           ) +
  annotate(geom = 'rect', ymin = 0, ymax = 1, alpha = 0.1,
           xmin = 41739801-1250, 
           xmax = 41739742 + 250
           ) +
  annotate(geom = 'rect', ymin = 0, ymax = 1, alpha = 0.1,
           xmin = 41733356-1900, 
           xmax = 41733356 -950
           ) +
  facet_wrap(vars(Tissue), ncol = 1, strip.position = 'left',
             labeller = labeller(Tissue = function(x)gsub(' cs', '', x))) +
  theme_bw(base_size = font_size) +
  theme(strip.background = element_blank(),
        strip.text.y.left = element_text(angle = 0, hjust = 1),
        strip.placement = 'outside',
        axis.ticks = element_blank(),
        axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.line = element_line(size = 0.5),
        panel.border = element_blank(),
        panel.grid.minor = element_blank(),
        panel.spacing = unit(0, 'lines'),
        #panel.grid = element_blank()
        ) +
  scale_y_continuous(limits = c(0,1), breaks = c(0,1), expand = c(0,0), labels = NULL) +
  scale_x_continuous(limits = c(s_extend, e_extend), expand = c(0,0), 
                     breaks = breaks_width(15000),
                     labels = scales::number) +
  scale_color_manual(values = color_code_tissue,guide = 'none') +
  labs(y = '', x = '', title = 'INHBA');p1

p2 <- annotations %>%
  filter(seqnames == as.character(c), start > s_extend, end < e_extend,
         type %in% c('exons', 'introns', 'promoters')) %>%
  
  filter(symbol != 'INHBA-AS1') %>%
  
  ggplot(aes(x = start, xend = end)) +
  geom_segment(y= 0.5, yend = 0.5, aes(size = type, color = type))  +
  annotate(geom = 'rect', ymin = 0, ymax = 1, alpha = 0.1,
           xmin =41745000-200, 
           xmax = 41745000+1250
           ) +
  annotate(geom = 'rect', ymin = 0, ymax = 1, alpha = 0.1,
           xmin = 41739801-1250, 
           xmax = 41739742 + 250
           ) +
  annotate(geom = 'rect', ymin = 0, ymax = 1, alpha = 0.1,
           xmin = 41733356-1900, 
           xmax = 41733356 -950
           ) +
  facet_wrap(~tx_id, ncol = 1, strip.position = 'left') +
  theme_bw(base_size = font_size)+
  theme(strip.background = element_blank(),
        strip.text.y.left  = element_text(angle = 0),
        strip.placement = 'outside',
        legend.position = 'bottom',
        plot.title =  element_blank(),
        plot.subtitle = element_blank(),
        axis.text.y = element_blank(),
        axis.ticks = element_blank(),
        axis.line = element_line(size = 0.5),
        panel.spacing = unit(0, 'lines'),
        panel.grid.major.y = element_blank(),
        panel.grid.minor.y = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.border = element_blank(),
        legend.key.size = unit(0.5, 'cm'),
        legend.spacing = unit(0.25, 'cm')) +
  scale_size_manual(values = c('exons' = 7, 'introns' = 1, 'intronexonboundaries' = 3,
                               '5UTRs' = 5, '3UTRs' = 6,
                               '1to5kb' = 2,
                               'promoters' = 2,
                               
                               'cpg_inter' = 1,
                               'cpg_islands' = 5,
                               'cpg_shores' = 4,
                               'cpg_shelves' = 3)) +
  scale_color_manual(values = c('exons' = 'black', 'introns' = 'black', 'intronexonboundaries' = 'black',
                               '5UTRs' = 'grey', '3UTRs' = 'grey',
                               '1to5kb' = 'grey',
                               'promoters' = 'forestgreen',
                               
                               'cpg_inter' = 1,
                               'cpg_islands' = 5,
                               'cpg_shores' = 4,
                               'cpg_shelves' = 3)) +
  scale_x_continuous(limits = c(s_extend,e_extend), 
                     expand = c(0,0), 
                     labels = scales::number,
                     breaks = breaks_width(15000)) +
  scale_y_continuous(expand = c(0,0)) +
  labs(x = 'position', size = '', color = '');p2
egg::ggarrange(p1,p2, ncol = 1, heights = c(3, 1)) #H:4.25, #w: 3.25


igv(gene_name = 'INHBA')
```

## TFAP2C

```{r}
igv(gene_name = 'TFAP2C', show_transcript_gene = FALSE, breaks_width = 10000,
    highlight_region = tibble(start = c(55200000-600),
                              end = c(55200000 + 4000)))#H:3.25, #w: 3.25
```

## MMP2

```{r}
igv(gene_name = 'MMP2', start_extend = 2500, end_extend = 2500,
    show_transcript_gene = FALSE,
    highlight_region = tibble(start = c(55515000 - 2750),
                              end = c(55515000 -500)))#H:4.25, #w: 3.25
```

## SLC1A5

```{r}
igv(gene_name = 'SLC1A5', start_extend = 2500, end_extend = 2500,
    show_transcript_gene = FALSE,
    highlight_region = tibble(start = c(47280000 + 7600),
                              end = c(47280000 +12500)))#H:4.25, #w: 3.25
```

## RRM2

```{r}
igv(gene_name = 'RRM2', start_extend = 2600, end_extend = 500)#H:4.25, #w: 3.25
```

## GATA3

```{r}
igv(gene_name = 'GATA3', start_extend = 2600, end_extend = 500,
    highlight_region = tibble(start = c(8090000 - 300),
                              end = c(8105000-1000)),
    exon_size = 3)#H:4.25, #w: 3.25
```


## TEAD3

```{r}
igv(gene_name = 'TEAD3', start_extend = 2600, end_extend = 500,
    highlight_region = tibble(start = c(35460000 - 7000),
                              end = c(35460000 + 7000)))#H:4.25, #w: 3.25
```

## JUNB

```{r}
igv(gene_name = 'JUNB', start_extend = 2600, end_extend = 2600)#H:4.25, #w: 3.25
```

## ST3GAL1

```{r}
igv(gene_name = 'ST3GAL1', start_extend = 5000, end_extend = 500)#H:4.25, #w: 3.25
```

## PKM2

```{r}
#igv(gene_name = 'PKM2', start_extend = 2500, end_extend = 2500)#H:4.25, #w: 3.25
```

## NDRG1

```{r}
igv(gene_name = 'NDRG1', start_extend = 2500, end_extend = 500,
    breaks_width = 25000,
    show_transcript_gene = FALSE,
    transcript_height = 0.5,
     highlight_region = tibble(start = c(134250000 +8000),
                              end = c(134250000 + 14000)))#H:4.25, #w: 3.25
```

## CHI3L2

```{r}
igv(gene_name = 'CHI3L2', start_extend = 2500, end_extend = 2500)#H:4.25, #w: 3.25
```

## FAM3B

ok

```{r}
igv(gene_name = 'FAM3B', start_extend = 2500, end_extend = 2500)#H:4.25, #w: 3.25
```

## SYNE1

too many transcripts

```{r}
igv(gene_name = 'SYNE1', start_extend = 2500, end_extend = 2500)#H:4.25, #w: 3.25
```

## AGAP1

too large of a region

```{r}
igv(gene_name = 'AGAP1', start_extend = 2500, end_extend = 2500)#H:4.25, #w: 3.25
```

## ADAM12

good but lots of data

```{r}
igv(gene_name = 'ADAM12', start_extend = 2500, end_extend = 7500,
    show_transcript_gene = FALSE, 
    breaks_width = 100000,
    highlight_region = tibble(start = c(128050000 + 10000, 
                                        128000000 - 15000,
                                        127900000 - 4000,
                                        127800000 - 10000),
                              end = c(128050000 + 38000, 
                                      128000000 + 5000,
                                      127900000 + 18000,
                                      127800000 + 25000))
      )#H:4.25, #w: 3.25
```

## CGA

```{r}
igv(gene_name = 'CGA', start_extend = 500, end_extend = 500,
    breaks_width = 2500,
    show_transcript_gene = FALSE,
    highlight_region = tibble(start = c(87805000-100),
                              end = c(87805000+100)))#H:4.25, #w: 3.25
```

## FN1

```{r}
igv(gene_name = 'FN1', start_extend = 500, end_extend = 500)#H:4.25, #w: 3.25
```

## PKM2

```{r}
#igv(gene_name = 'PKM2', start_extend = 500, end_extend = 500)#H:4.25, #w: 3.25
```

## KRT15 

```{r}
igv(gene_name = 'KRT15', start_extend = 500, end_extend = 500)#H:4.25, #w: 3.25
```

## DNMT1 

```{r}
igv(gene_name = 'DNMT1', start_extend = 2500, end_extend = 2500)#H:4.25, #w: 3.25
```

## DNMT3A 

```{r}
igv(gene_name = 'DNMT3A', start_extend = 2500, end_extend = 2500)#H:4.25, #w: 3.25
```

## DNMT3B 

```{r}
igv(gene_name = 'DNMT3B', start_extend = 2500, end_extend = 2500)#H:4.25, #w: 3.25
```

## DNMT3L 

```{r}
igv(gene_name = 'DNMT3L', start_extend = 2500, end_extend = 2500)#H:4.25, #w: 3.25
```

## TET1 

```{r}
igv(gene_name = 'TET1', start_extend = 10000, end_extend = 2500)#H:4.25, #w: 3.25
```

# PCA

```{r}
# plot without legend
p <- pDat %>%
  filter(Trimester != 'Second') %>%
  select(Sample_Name, Case_ID, Trimester, Tissue, contains('PC')) %>%
  ggplot(aes(x = PC1_raw, y = PC2_raw, color = Tissue, shape = Trimester)) +
  geom_point(alpha = 0.9, size = 0.75) +
  theme_bw(base_size = font_size+1) +
  theme(panel.grid.minor = element_blank(),
        strip.background = element_blank(),
        axis.ticks = element_blank(),
        panel.grid = element_blank(),
        panel.border = element_blank(),
        axis.line = element_line(color = 'black'),
        legend.position = '',
        legend.spacing.x = unit(-0.1, 'cm'),
        legend.spacing.y = unit(-0.2, 'cm')) +
  scale_x_continuous(breaks = c(-100, 0, 100)) +
  scale_y_continuous(breaks = c(-100, 0, 100)) +
  scale_color_manual(values = color_code_tissue[pDat$Tissue],
                     guide = guide_legend(override.aes = list(size = 3, alpha = 1),
                                          ncol = 2),
                     labels = function(x)gsub(' cs', '', x)) +
  scale_shape_manual(values = c(19, 3)) +
  guides(shape = guide_legend(override.aes = list(size = 1))) +
  labs(x = 'PC1 (41%)', y = 'PC2 (23%)', color = '') +
  coord_equal();p #h: 1.75, w: 2.75

#just for the legend
p <- pDat %>%
  filter(Trimester != 'Second') %>%
  select(Sample_Name, Case_ID, Trimester, Tissue, contains('PC')) %>%
  ggplot(aes(x = PC1_raw, y = PC2_raw, color = Tissue, shape = Trimester)) +
  geom_point(alpha = 0.9, size = 0.75) +
  theme_bw(base_size = font_size) +
  theme(panel.grid.minor = element_blank(),
        strip.background = element_blank(),
        axis.ticks = element_blank(),
        panel.grid = element_blank(),
        panel.border = element_blank(),
        axis.line = element_line(color = 'black'),
        legend.position = 'right',
        legend.key.size = unit(0.2, 'cm')) +
  scale_x_continuous(breaks = c(-100, 0, 100)) +
  scale_y_continuous(breaks = c(-100, 0, 100)) +
  scale_color_manual(values = color_code_tissue[pDat$Tissue],
                     guide = guide_legend(override.aes = list(size = 3, alpha = 1),
                                          ncol = ),
                     labels = function(x)gsub(' cs', '', x)) +
  scale_shape_manual(values = c(19, 3)) +
  guides(shape = guide_legend(override.aes = list(size = 1.5)),
         color = guide_legend(override.aes = list(size = 1.5))) +
  labs(x = 'PC1 (41%)', y = 'PC2 (23%)', color = '') +
  coord_equal();p
  
```

PC association

```{r}
library(yahew)
# compute pca
set.seed(1)
pca <- prcomp_irlba(t(betas_filt), n = 20, center = T, scale = F)

# add pc scores to pdata
pca_scores <- pca$x[,1:20] %>% as.data.frame()
colnames(pca_scores) <- paste0(colnames(pca_scores), '_processed')
pDat_filt <- pDat_filt %>% 
  bind_cols(pca_scores)

# create proportion variance explained data frame
pc_info <- summary(pca)$importance %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column(var = 'variable') %>%
  gather(key = 'PC', value = 'value', -variable) %>%
  as_tibble() %>%
  mutate(PC = factor(as.character(PC), levels = paste0('PC', 1:20)),
         Label = ifelse(variable == 'Proportion of Variance',
                        paste0(PC, ' (', prettyNum(value*100, digits = 2), '%)'),
                        as.character(PC))) %>%
  arrange(variable, PC)

  
# correlate PCs with phenodata
pc_cor <- lmmatrix(dep = pca$x[,1:20, drop = F],
                   ind = pDat_filt %>%
                   dplyr::select(Case_ID, Tissue, Sex, Trimester, #bio
                                 Week, Chip_number, Row_numeric, Row_factor, Batch_BSC, # batch
                                 DNA_loaded, 
                                 failed_probes,
                                 maternal_contamination_norm_flip),
                   metric = 'Pvalue')
# plot data
pc_cor <- pc_cor %>% as.data.frame() %>% 
  
  # add dep variables
  mutate(dep = rownames(pc_cor)) %>%
  
  # reshape
  gather(PC, pval, -dep) %>%
  
  # pvalue categories
  mutate(pval_cat = factor(case_when(
    pval > 0.05  ~ '> 0.05',
    pval < 0.05 & pval > 0.01 ~ '< 0.05',
    pval < 0.01 & pval > 0.001 ~ '< 0.01',
    pval < 0.001 ~ '< 0.001'
  ), levels = c('> 0.05', '< 0.05','< 0.01', '< 0.001')),
  
  # make PC is encoded with proper levels!!!
  PC = factor(PC, levels = paste0('PC', 1:20))) %>% as_tibble()

# create color palette
colpal <- c('white', '#fee8c8', '#fdbb84', '#e34a33')
names(colpal) <- levels(pc_cor$pval_cat)

# relevel
pc_cor <- pc_cor %>% 
  mutate(dep = factor(dep, levels = c('Case_ID', 'Tissue', 'Sex', 'Trimester',
                                     'Week', 'Chip_number', 'Row_numeric', 'Row_factor', 
                                     'Batch_BSC', # batch
                                     'DNA_loaded',
                                     'failed_probes',
                                     'maternal_contamination_norm_flip')))
                                     
p7 <- ggplot(pc_cor, aes(x = PC, y = dep, fill = pval_cat)) +
  geom_tile(col = 'lightgrey') + theme_bw() +
  scale_x_discrete(expand = c(0, 0), labels = 1:20) +
  scale_y_discrete(expand = c(0, 0)) +
  scale_fill_manual(values = colpal)  + 
  labs(y = '', fill = 'P value')

p8 <- ggplot(pc_info %>% filter(variable == 'Proportion of Variance') %>%
               mutate(value = value*100), 
             aes(x = PC, y = value)) +
  geom_bar(stat = 'identity') +
  theme_bw(base_size = 8) + 
  scale_x_discrete(expand = c(0, 0), labels = 1:20) + 
  scale_y_continuous(expand = c(0, 0)) +
  labs(y = '% variance explained')

egg::ggarrange(p7, p8, heights = c(3,1))


# create another without all variables
var_interest <- c('Tissue', 'Sex', 'Trimester', 
                  'Week', 'Chip_number', 'Row_numeric', 'Row_factor', 'Batch_BSC')

p9 <- pc_cor %>% 
  filter(dep %in% var_interest) %>%
  mutate(dep = ifelse(dep %in% var_interest[4:8], paste0('TECH_', dep), as.character(dep))) %>%
  mutate(dep = factor(dep, levels = c('TECH_Week', 'TECH_Chip_number', 'TECH_Row_numeric',
                                      'TECH_Row_factor', 'TECH_Batch_BSC', 
                                      'Sex', 'Trimester', 'Tissue'))) %>%
  ggplot(aes(x = PC, y = dep, fill = pval_cat)) +
  geom_tile(col = 'lightgrey') + theme_bw() +
  scale_x_discrete(expand = c(0, 0), labels = 1:20) +
  scale_y_discrete(expand = c(0, 0)) +
  theme_bw(base_size = 8) + 
  scale_fill_manual(values = colpal)  + 
  labs(y = '', fill = 'P value')

egg::ggarrange(p9, p8, heights = c(3,1))

pDat_filt %>% 
  dplyr::count(Trimester, Tissue, Sex)
```

# Number of DMCs in genes


just DNMTs and TET1 for supp figure

```{r}
dnmt <- anno %>%
  filter(grepl('DNMT|TET1', genes_symbol)) %>%
  inner_join(dmcs, by = c('cpg' = 'gene'))

         
dnmt %>%
  
  filter(grepl('promoter', genes_id) |
           grepl('1to5kb', genes_id)) %>%

  # boolean vector indicating if cpg is a differentially methylated for 1 or more cell type
  mutate(sig = (bonferroni < 0.01 & abs(delta_b > 0.1))) %>%
        # bonferroni = scales::pvalue(bonferroni, accuracy = 0.01), 
         #delta_b = scales::number(delta_b, accuracy = 0.01)) %>%
  #select(Group1, cpg, contains('gene'), sig, bonferroni, delta_b) %>% View()
  
  group_by(cpg) %>%
  summarize(genes_symbol = dplyr::first(genes_symbol), sig = any(sig)) %>%
  select(cpg, genes_symbol, sig) %>%
  
  # put gene symbol on rows
  separate_rows(genes_symbol, sep = ', ') %>%
  distinct() %>%
  
  # calculate % dmcs per gene
  group_by(genes_symbol) %>%
  summarize(p = sum(sig)/n())
```

all genes

```{r}
# what prop many cpgs are dm
dmcs %>% 
  group_by(gene) %>% 
  summarize(n = any(bonferroni < 0.01 & abs(delta_b) > 0.1)) %>% 
  summarize(sum(n)/n()) # 0.627

# what prop of cpgs are dm, outside of genes
dmcs %>% 
  filter(gene %in% (anno %>% filter(is.na(genes_symbol)) %>% pull(cpg))) %>%
  group_by(gene) %>% 
  summarize(n = any(bonferroni < 0.01 & abs(delta_b) > 0.1)) %>% 
  summarize(sum(n)/n()) # 0.800

# what prop of cpgs are dm, inside of genes
dmcs %>% 
  filter(gene %in% (anno %>% filter(!is.na(genes_symbol)) %>% pull(cpg))) %>%
  group_by(gene) %>% 
  summarize(n = any(bonferroni < 0.01 & abs(delta_b) > 0.1)) %>% 
  summarize(sum(n)/n()) # 0.574


all_genes <- anno %>%
  mutate(genes_symbol = gsub('NA, ', '', genes_symbol), # if cpg is otherwise associated with a gene,
         genes_symbol = ifelse(is.na(genes_symbol), 'NA', genes_symbol)) %>%
  inner_join(dmcs, by = c('cpg' = 'gene')) %>%
  
  filter(grepl('promoter', genes_id) |
           grepl('1to5kb', genes_id)) %>%
  
  # boolean vector indicating if cpg is a differentially methylated for 1 or more cell type
  mutate(sig = (bonferroni < 0.01 & abs(delta_b > 0.1))) %>%
        # bonferroni = scales::pvalue(bonferroni, accuracy = 0.01), 
         #delta_b = scales::number(delta_b, accuracy = 0.01)) %>%
  #select(Group1, cpg, contains('gene'), sig, bonferroni, delta_b) %>% View()
  
  group_by(cpg) %>%
  summarize(genes_symbol = dplyr::first(genes_symbol), sig = any(sig)) %>%
  dplyr::select(cpg, genes_symbol, sig) %>%
  
  # put gene symbol on rows
  separate_rows(genes_symbol, sep = ', ') %>%
  distinct() %>%
  
  # calculate % dmcs per gene
  group_by(genes_symbol) %>%
  summarize(n_dm =  sum(sig),
            p_dm = n_dm/n(),
            n = n())

all_genes %>% 
  mutate(n_interval = cut(n, c(0, 10, 25, 50, 100, 250, Inf))) %>%
  ggplot(aes(x = n_interval, y = p_dm)) +
  geom_violin()

p3 <- all_genes %>% 
  mutate(n_interval = cut(n, c(0, 10, 25, 50, 100, 250, Inf))) %>%
  filter(n > 5, n < 200,
         n_dm > 4) %>%
  {
    ggplot(data = ., aes(x = n, y = n_dm)) +
      geom_point(alpha = 0.1, size = 0.5) +
      stat_smooth(se = FALSE, color = '#1b9e77', 
                  alpha = 0.25, size = 0.25, method = 'loess') +
      geom_abline(intercept = 0, slope = 1, color = '#7570b3', alpha = 1, size = 0.25) +
      
      # highlight genes
      geom_text_repel(data = . %>% 
                        filter(genes_symbol %in% c('DNMT1', 'DNMT3L')),
                      aes(label = genes_symbol),
                      segment.color = '#d95f02',
                      size = 3,
                      segment.size = 0.1,
                      nudge_y = 150) +
      geom_point(data = . %>% 
                        filter(genes_symbol %in% c('DNMT1', 'DNMT3L')),
                 color = '#c95353', size = 0.5) +
      
      theme_bw(base_size = font_size) +
      theme(panel.grid = element_blank(),
            panel.border = element_blank(),
            axis.line = element_line(),
            axis.ticks = element_blank(),
            axis.title.y = element_text(angle = 0, vjust = 0.5, hjust = 0),
            plot.margin = margin(0.1, 0.5, 0.1, 0.1, "cm")) +
      scale_x_continuous(expand = c(0,0), limits = c(5, 200)) +
      scale_y_continuous(expand = c(0,0), limits = c(5,200)) +
      #coord_cartesian(xlim = c(5, 200),
      #                ylim = c(5, 200)) +
      labs(x = 'CpGs per gene\n(n)', y = 'Differentially\nmethylated\n(n)')
    }; p3 # h: 1.75, w: 3

p4 <- all_genes %>%
  filter(n > 10) %>%
  ggplot(aes(x = p_dm)) +
  geom_histogram(binwidth = 0.05, color = 'white', size = 1,
                 fill = '#c2c2c2') +
  geom_vline(xintercept = all_genes %>%
               filter(n > 10) %>% summarize(m = mean(p_dm)) %>% pull(m), 
             color = 'black', 
             linetype = 'dashed',
             size = 0.75) + # average
  theme_bw(base_size = font_size) +
  theme(axis.title.y = element_text(angle = 0, vjust = 0.5),
        panel.grid = element_blank(),
        panel.grid.major.y = element_line(color = 'grey'),
        panel.border = element_blank(),
        axis.ticks = element_blank(),
        axis.line = element_line()) +
  scale_x_continuous(expand = c(0,0), labels = scales::percent) +
  scale_y_continuous(expand = c(0,0)) +
  labs(x = 'Gene-wise percentage of CpGs\nthat are differentially methylated', y = 'Count');p4 #h: 1.75, w:2.75



p5 <- tibble(
  set = c('All', 'Outside genes', 'In genes'),
  `Differentially methylated` = c(0.627, 0.800, 0.574),
  `Not differentially methylated` = 1-`Differentially methylated`
) %>%
  pivot_longer(
    cols = c(`Differentially methylated`, `Not differentially methylated`),
    names_to = 'DM',
    values_to = 'p'
  ) %>%
  mutate(DM = fct_rev(DM)) %>%
  filter(DM == 'Differentially methylated') %>%
  {
    ggplot(data = ., aes(x = set, y = p, fill = DM)) +
      geom_bar(stat = 'identity') +
      geom_text(aes(label = scales::percent(p, accuracy = 1)),
                nudge_y = +0.1,
                size = 1.75) +
      theme_bw(base_size = 6) +
      theme(panel.border = element_blank(),
            panel.grid = element_blank(),
            axis.ticks = element_blank(),
            axis.text.y = element_blank(),
            axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 4),
            axis.line.x = element_line()) +
      scale_y_continuous(expand = c(0,0), breaks = c(0,0.5,1), limits = c(0,1),
                         labels = scales::percent) +
      scale_fill_manual(values = c('grey'), guide = FALSE) +
      labs(x = '', fill = '', y = '', x = '') 
  };p5 #h: 1.25, w: 0.75

  geom_vline(xintercept = 0.627, color = 'black', linetype = 'dashed') + # out of all cpgs
  geom_vline(xintercept = 0.800, color = 'red', linetype = 'dashed') + # outside of genes
  geom_vline(xintercept = 0.574, color = 'green', linetype = 'dashed') + # inside genes
    
all_genes %>% 
  mutate(n_interval = cut(n, c(0, 10, 25, 50, 100, 250, Inf))) %>%
  arrange(desc(n_interval), desc(p_dm)) %>% DT::datatable()
```

# Figure S* contamination

*From 11_QC_Report.Rmd*

## XY

```{r}
# load data
pDat_preprocess <- readRDS(here::here('data', 'main', 'interim', '1_1_pDat.rds'))
pDat_preprocess <- pDat_preprocess %>% 
  mutate(Tissue = ifelse(Tissue == 'Dead Cells and Lymphocytes', 'Dead cells', Tissue))

convert_names2 <- function(x) {
  gsub('_hofb_cs', ' Hofbauer', 
       gsub('_endo_cs', ' Endothelial',
            gsub('_v', ' Villi',
                 gsub('_strom_cs', ' Stromal',
                      gsub('_troph_cs', ' Trophoblasts', x)))))
}

# xy sample check
p6 <- pDat_preprocess %>%
  ggplot(aes(x = normalized_X_intensity, y = normalized_Y_intensity, fill = Sex)) +
  geom_hline(yintercept = 0.5, linetype = 'dashed', col = '#9B9B9B') + 
  geom_vline(xintercept = 0.95, linetype = 'dashed', col = '#9B9B9B') +
  geom_text_repel(data = pDat_preprocess %>% 
                    filter(Sex == 'M', 
                           normalized_X_intensity > 0.95,
                           normalized_Y_intensity < 0.5) %>%
                    mutate(Sample_Name = convert_names2(Sample_Name)), 
            aes(label = Sample_Name), 
            segment.size = 0.1,
            size = 1.75, 
            force = 15, nudge_x = -0.2, nudge_y = -0.1) +
  geom_point(shape = 21, col = 'white', size = 1.5, stroke = 0.01) +
  theme_bw(base_size = font_size) +
  theme(axis.ticks = element_blank(),
        axis.title.y = element_text(angle = 0, vjust = 0.5, hjust = 0, size = font_size),
        
        legend.key.size = unit(0, 'lines'),
        
        panel.grid = element_blank(),
        panel.background = element_blank(),
        panel.border = element_blank(),
        
        axis.line = element_line(color = 'black', size = 0.5, linetype = 'solid'))  +
  labs(x = 'Normalized X intensity', y = 'Normalized\nY intensity');p6 # h:1.5, w: 3.25
```

## Genotype correlations

```{r}
# genotype correlations
label_size <- 3.5
point_size <- 7.5
r_size <- 2.5
colors <- c('red', 'red', rev(viridis_pal(option = 'D')(9))[c(1,1,1:8)])
r_limit <- c(0.65,1)

# organize by trimester
ind_first <- pDat_preprocess %>% 
  add_count(Case_ID) %>%
  filter(Case_ID == 'PL289',
         Tissue != 'Dead cells') %>%
  arrange(n, Case_ID) %>%
  pull(Sample_Name)

pl289 <- snp_betas_cor[ind_first,ind_first]

convert_names <- function(x) {
  gsub('_hofb_cs', 'Hofbauer', 
       gsub('_endo_cs', 'Endothelial',
            gsub('_v', 'Villi',
                 gsub('_strom_cs', 'Stromal',
                      gsub('_troph_cs', 'Trophoblasts', 
                           gsub('PL289', '', x))))))
}
rownames(pl289) <- convert_names(rownames(pl289))
colnames(pl289) <- convert_names(colnames(pl289))

pl289 <- pl289[c(1,2,4,5,3), c(1,2,4,5,3)]
pl289[lower.tri(pl289, diag = TRUE)] <- NA

p7 <- pl289[,-1] %>%
  as_tibble %>%
  mutate(sample2 = factor(rownames(pl289), levels = rownames(pl289))) %>%
  pivot_longer(cols = -sample2,
               names_to = 'sample1',
               values_to = 'cor') %>%
  filter(!is.na(cor)) %>%
  mutate(sample1 = factor(sample1, levels = rev(rownames(pl289)))) %>%
  
  ggplot(aes(x = sample1, y = sample2, fill = cor)) +
  geom_tile(color = '#fcfcfc', alpha = 1, linejoin = 'bevel') +
  geom_text(color = 'white',
            size = 2,
            aes( label = scales::number(cor, accuracy = 0.01))) +
  theme_bw(base_size = font_size) +
  theme(axis.ticks = element_blank(),
        axis.text.x = element_text(angle = 90, hjust = 1),
        
        legend.key.size = unit(3, 'mm'),
          
        panel.grid = element_blank(),
        panel.border = element_blank()) +
  scale_x_discrete(expand = c(0,0)) +
  scale_y_discrete(expand = c(0,0)) +
  scale_fill_viridis_c(limits = c(0.7, 1), direction = -1) +
  labs(x = '', y = '', fill = 'Genetic\ncorrelation', title = 'PL289');p7 #1.75, w: 2.25
  

```

## SNP betas distributions

```{r}
#snp betas
snp_betas <- readRDS(here::here('data', 'main', 'interim', '1_1_snp_betas.rds'))

snp_betas_melt <- t(snp_betas) %>% 
  as_tibble %>% 
  mutate(Sample_Name = colnames(snp_betas)) %>%
  left_join(pDat %>% select(Sample_Name, Tissue, Trimester), by = 'Sample_Name') %>%
  gather(key = 'SNP', value = 'Beta', -Sample_Name, -Tissue, -Trimester)

p8 <- snp_betas_melt %>%
  filter(Trimester != 'Second') %>%
  ggplot(aes(x = SNP, y = Beta)) +
  geom_point(alpha = 0.5, shape = 16, color = 'black', size = 0.5) + 
  labs(x = 'SNPs (n=59)', y = '') +
  facet_grid(Trimester ~ Tissue,
             switch = 'y',
             labeller = labeller(.rows = label_both, .cols = function(x)gsub(' cs', '', x))) +
  
  # cosmetics
  theme_bw(base_size = font_size) +
  theme(axis.text.x = element_blank(), 
        axis.text.y = element_text(size = font_size-2),
        axis.ticks = element_blank(),
        axis.title.y = element_text(angle = 0, vjust = 0.5, hjust = 0, size = font_size),
        
        legend.text = element_text(),
        legend.key.size = unit(0, 'lines'),
        
        panel.grid.minor.y = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.spacing.y = unit(0.5, 'cm'),
        axis.line.y = element_line(color = 'black', size = 0.5, linetype = 'solid'),
        
        strip.background = element_blank(),
        strip.placement = 'outside',
        strip.text.y = element_text(face = 'bold', hjust = 1, angle = 180),
        
        plot.margin=margin(l=-0.1, unit="cm")) +
  #scale_color_manual(values= color_code_tissue[unique(snp_betas_melt$Tissue)],
  #                   labels = function(x)gsub(' cs', '', x),
  #                   guide = 'none') +
  scale_y_continuous(limits = c(0, 1), 
                     breaks = c(0, 0.5, 1),
                     labels = scales::percent) +
  scale_x_discrete(breaks = NULL, expand = c(0.02, 0.02));p8 # h: 2.25 w: 6

```

## Theor

```{r}
df <- data.frame(contamination = seq(0, 1, 0.05),
           probability = c(seq(0,0.5,0.05), seq(0.45, 0, -0.05)))

p9 <- df %>% 
  ggplot(aes(x = contamination, y = probability)) +
  geom_smooth(stat = 'smooth', method = 'loess', size = 0.5) + 
  geom_point(size = 0.75) +
  labs(x = 'Maternal contamination', 
       y = 'P(SNP falls\noutside of\nexpected\ndistribution)') +
  theme_bw(base_size = font_size) +
  theme(panel.border = element_blank(),
        panel.grid = element_blank(),
        axis.line = element_line(),
        axis.title.y = element_text(angle = 0, vjust = 0.5)) +
  scale_x_continuous(expand = c(0.01,0.01),
                     labels = function(x)scales::percent(x, accuracy = 1)) +
  scale_y_continuous(expand = c(0.01,0.01), limits = c(-0.01,0.51),
                     labels = function(x)scales::percent(x, accuracy = 1));p9

p10 <- ggplot(pDat_preprocess %>% filter(Sex == 'M'), 
       aes(x = normalized_Y_intensity, y = Prob_SNP_outlier)) +
  geom_smooth(method = 'loess', size = 0.5)+
  geom_point(size = 1.25, stroke = 0.01, shape = 21, color = 'white', fill = 'black') +
  theme_bw(base_size = font_size) +
  theme(panel.border = element_blank(),
        panel.grid = element_blank(),
        axis.line = element_line(),
        axis.title.y = element_text(angle = 0, vjust = 0.5)) +
  scale_x_continuous(expand = c(0,0)) +
  scale_y_continuous(expand = c(0,0), limits = c(0,0.5),
                     labels = function(x)scales::percent(x, accuracy = 1)) +
  scale_fill_viridis_c(option = 'inferno') +
  labs(x = 'Normalized Y intensity', y = 'P(SNP is outlier)');p10 #h:1.25, #w:2.75
```

## Predictor

```{r}
pDat_1_3 <- readRDS(here::here('data', 'main', 'interim', '1_3_pDat.rds'))
pDat_2_3 <- readRDS(here::here('data', 'main', 'interim', '2_3_pDat_contam.rds'))

fit_ycor <-lm(normalized_Y_intensity ~ cor_to_reference, pDat_1_3 %>% filter(Sex == 'M'))

p11 <- pDat_2_3 %>%
  select(Sex, contains('maternal_contamination'), cor_to_reference, normalized_Y_intensity) %>%
  gather(key = step, value = Contamination, -Sex, -cor_to_reference) %>%
  mutate(step = as.factor(case_when(
    step == 'normalized_Y_intensity' ~ '1. Fit linear model in males',
    step == 'maternal_contamination' ~ '2. Predict on females',
    step == 'maternal_contamination_norm' ~ '3. Normalize distribution\nto 0-1',
    step == 'maternal_contamination_norm_flip' ~ '4. Reverse, such that 1 = 100%'))) %>% 
  {
    ggplot(., aes(x = cor_to_reference, y = Contamination, color = Sex, fill = Sex)) +
      geom_smooth(data = . %>% filter(step == '1. Raw data', Sex == 'M'), method = 'lm',
                  show.legend = F) +
      geom_abline(data = data.frame(slope = fit_ycor$coefficients[2], 
                                    intercept = fit_ycor$coefficients[1],
                                    step = c('2. Predict on females'),
                                    Sex = c('M')),
                  aes(slope = slope, intercept = intercept, color = Sex),
                  show.legend = FALSE) +
      geom_point(alpha = 1, size = 1.25, shape = 21, color = 'white', stroke = 0.1) +
      facet_wrap(~step, ncol = 4) +
      theme_bw(base_size = font_size) +
      theme(strip.background = element_blank(),
            strip.text = element_text(face = 'bold'),
            panel.grid.minor = element_blank(),
            axis.title.y = element_text(angle = 0, vjust = 0.5),
            legend.key.size = unit(1.5, 'mm')) +
      labs(x = 'Genetic correlation to reference',
           y = 'Measure of\nmaternal\ncontamination', color = '') +
      guides(color = guide_legend(override.aes = list(size = 1.5)))
  };p11 # h:1.5, w: 6


```

# Hofb and STB analysis

```{r, message = FALSE, warning = FALSE}
library(limma)
library(lumi)
library(broom)
library(biobroom)
library(DMRcate)
library(missMethyl)
library(tidyverse)
library(pheatmap)
library(ggridges)
library(umap)
library(pheatmap)
library(viridis)
library(here)

pDat <- readRDS(here('data', 'main', 'interim', '2_3_pDat_contam.rds')) # for mat contam

betas <- readRDS(here::here('data', 'main', 'interim', '1_4_betas_noob_filt.rds'))
pca <- readRDS(here::here('data', 'main', 'interim','2_4_pca.rds'))

betas_bmiq <- readRDS(here::here('data', 'main', 'interim', '1_6_bmiq.rds'))

pDat$Sentrix == colnames(betas_bmiq) 
pDat$Sentrix == colnames(betas) 
colnames(betas) <- pDat %>% pull(Sample_Name)
colnames(betas_bmiq) <- pDat %>% pull(Sample_Name)

color_code <- readRDS(here::here('data', 'main', 'interim', '2_3_color_code.rds'))
color_code_tissue <- setNames(color_code$Colors_Tissue, color_code$Tissue)
colors <- color_code_tissue[unique(pDat$Tissue)]

# filter samples
pDat_filt <-  pDat %>%
  filter(!Tissue %in% c('Villi maternal',  'Mixture'),
         !Sample_Name %in% c('PM364_hofb_cs', 'PL293_v_R2', 'PM366_vc_R2', 'P131_hofb_cs'))

betas_filt <- betas[, pDat_filt %>%
                      pull(Sample_Name)]
```

## STB and dead cells PCA

```{r}
pDat %>% dplyr::count(Trimester, Tissue)

p_sd_pca <- pDat %>% 
  filter(Trimester != 'Second',
         !Tissue %in% c('Villi maternal',  'Dead Cells and Lymphocytes')) %>%
  mutate(Tissue = ifelse(Tissue %in% c('Syncytiotrophoblast', 'Villi', 'Trophoblasts'),
                         Tissue, NA)) %>%
  arrange(desc(Tissue)) %>%
  ggplot(aes(x = PC1_raw, y = PC2_raw, color = Tissue)) +
  geom_point(alpha = 0.9, size = 0.75) +
  theme_bw(base_size = font_size) +
  theme(panel.grid.minor = element_blank(),
        strip.background = element_blank(),
        axis.ticks = element_blank(),
        panel.grid = element_blank(),
        panel.border = element_blank(),
        axis.line = element_line(color = 'black'),
        plot.title = element_blank(),
        plot.subtitle = element_blank(),
        legend.position = 'top',        
        legend.key.size = unit(0.35, 'cm'),
        legend.spacing = unit(0, 'cm')) +
  scale_x_continuous(breaks = c(-100, 0, 100)) +
  scale_y_continuous(breaks = c(-100, 0, 100)) +
  scale_color_manual(values = colors,
                     breaks = c('Syncytiotrophoblast', 'Villi', 'Trophoblasts'),
                     na.value = 'grey',
                     guide = guide_legend(override.aes = list(size = 3, alpha = 1),
                                          ncol = 1),
                     labels = function(x)gsub(' cs', '', x)) +
  scale_shape_manual(values = c(19, 3)) +
  guides(shape = guide_legend(override.aes = list(size = 1))) +
  labs(x = 'PC1 (41%)', y = 'PC2 (23%)', color = '') +
  coord_equal();p_sd_pca #h: 2.5, w: 2.75


#just for the legend
p <- pDat %>% 
  filter(Trimester != 'Second',
         Tissue != 'Villi maternal') %>%
  mutate(Tissue = ifelse(Tissue %in% c('Syncytiotrophoblast', 'Dead Cells and Lymphocytes'),
                         Tissue, NA)) %>%
  ggplot(aes(x = PC1_raw, y = PC2_raw, color = Tissue)) +
  geom_point(alpha = 0.9, size = 0.75) +
  theme_bw(base_size = font_size) +
  theme(panel.grid.minor = element_blank(),
        strip.background = element_blank(),
        axis.ticks = element_blank(),
        panel.grid = element_blank(),
        panel.border = element_blank(),
        axis.line = element_line(color = 'black'),
        legend.position = 'right',
        legend.key.size = unit(0.2, 'cm')) +
  scale_x_continuous(breaks = c(-100, 0, 100)) +
  scale_y_continuous(breaks = c(-100, 0, 100)) +
  scale_color_manual(values = colors,
                     breaks = c('Syncytiotrophoblast', 'Dead Cells and Lymphocytes'),
                     na.value = 'grey',
                     guide = guide_legend(override.aes = list(size = 3, alpha = 1),
                                          ncol = 1),
                     labels = function(x)gsub(' cs', '', x)) +
  scale_shape_manual(values = c(19, 3)) +
  guides(shape = guide_legend(override.aes = list(size = 1.5)),
         color = guide_legend(override.aes = list(size = 1.5))) +
  labs(x = 'PC1 (41%)', y = 'PC2 (23%)', color = '') +
  coord_equal();p
```

## STB pheatmap

```{r}
pheatmap <- readRDS(here::here('data', 'main', 'interim', '2_15_pheatmap_stb_troph.rds'))
pheatmap #800 h, 1000 w
```

## Density plots

```{r}
stb_troph_dens <- readRDS(here::here('data', 'main', 'interim', '2_15_stb-troph-density-data.rds'))
p_std <- stb_troph_dens %>%
  ggplot(aes(y = label, x = beta, fill = Tissue)) +
  geom_density_ridges(alpha = 0.75, color = 'white', panel_scaling = FALSE, size = 0.5) +
  facet_grid(rows = vars(gene), 
             scales = 'free_y',
             switch = 'y',
             space = 'free') +
  scale_fill_manual(values = colors) +
  scale_x_continuous(limits = c(0,1), labels = scales::percent) +
  theme_bw(base_size = font_size+1) +
  theme(panel.border = element_blank(),
        axis.line = element_line(),
        panel.grid.major.y = element_blank(),
        panel.grid.minor.x = element_blank(),
        axis.ticks = element_blank(),
        axis.text.y = element_text(size = font_size-4),
        strip.placement = 'outside',
        strip.background = element_blank(),
        strip.text.y.left = element_text(angle = 0, hjust = 1),
        legend.position = 'top',
        legend.key.size = unit(0.35, 'cm'),
        legend.spacing = unit(0, 'cm'),
        plot.title = element_blank(),
        plot.subtitle = element_blank()) +
#  guides(fill = guide_legend(direction = 'vertical')) ++
  guides(fill = 'none') +
  labs(x = '', y = '', fill = '');p_std ###fig_height <- 2.5;fig_width <- 4
```

## Hofb

```{r}
pheatmap <- readRDS(here::here('data', 'main', 'interim', '2_15_pheatmap_450k_cord_hofb.rds'))
pheatmap #1000h, 1000w
```

# missMethyl enrichment

```{r}
# plot results
go %>%
  filter(Trimester == 'First') %>%
  # take top n significant
  group_by(Celltype) %>%
  dplyr::slice(1:10) %>%
  
  ggplot(aes(x = Order, y = Generatio, fill = neg_log_P)) +
  geom_segment(aes(x = Order, xend = Order, y = 0, yend = Generatio)) +
  geom_point(stat = 'identity', shape = 21, color = 'black', size = 3) +
  theme_bw() +
  facet_wrap(vars(Celltype), scales = 'free_y') +
  scale_x_continuous(breaks = go$Order,
                     labels = go$TERM) +
  coord_flip() +
  scale_y_continuous(expand = c(0,0), limits = c(0,1)) +
  scale_fill_viridis_c(option = 'B', begin = 0, limits = c(0, NA)) +
  labs(fill = '-log10(P)', x = '', 
       y = '(# of genes with DMCs) / (# of genes associated with GO term)')

# plot results
go %>%
  filter(Trimester == 'Third') %>%
  # take top n significant
  group_by(Celltype) %>%
  dplyr::slice(1:10) %>%
  
  ggplot(aes(x = Order, y = Generatio, fill = neg_log_P)) +
  geom_segment(aes(x = Order, xend = Order, y = 0, yend = Generatio)) +
  geom_point(stat = 'identity', shape = 21, color = 'black', size = 3) +
  theme_bw() +
  facet_wrap(vars(Celltype), scales = 'free_y') +
  scale_x_continuous(breaks = go$Order,
                     labels = go$TERM) +
  coord_flip() +
  scale_y_continuous(expand = c(0,0), limits = c(0,1)) +
  scale_fill_viridis_c(option = 'B', begin = 0, limits = c(0, NA)) +
  labs(fill = '-log10(P)', x = '', 
       y = '(# of genes with DMCs) / (# of genes associated with GO term)')

# plot results
kegg %>%
  # take top n significant
  group_by(Label) %>%
  dplyr::slice(1:10) %>%
  
  ggplot(aes(x = Order, y = Generatio, fill = neg_log_P)) +
  geom_segment(aes(x = Order, xend = Order, y = 0, yend = Generatio)) +
  geom_point(stat = 'identity', shape = 21, color = 'black', size = 3) +
  theme_bw() +
  facet_wrap(vars(Label), scales = 'free_y') +
  scale_x_continuous(breaks = kegg$Order,
                     labels = kegg$Description) +
  coord_flip() +
  scale_y_continuous(expand = c(0,0), limits = c(0,1)) +
  scale_fill_viridis_c(option = 'B', begin = 0, limits = c(0, NA)) +
  labs(fill = '-log10(P)', x = '', 
       y = '(# of genes with DMCs) / (# of genes associated with GO term)')
```

# contamination levels

```{r}
#Before contamination
p <- pDat %>%
  filter(!Tissue %in% c('Villi maternal', 'Trophoblasts enz', 'Mixture cs', 'Dead Cells and Lymphocytes cs'),
         maternal_contamination_norm_flip < 0.35) %>%
  arrange(Tissue, maternal_contamination_norm_flip) %>% 
  mutate(Sample_Name = factor(as.character(Sample_Name), levels = Sample_Name),
         Trimester = gsub('Third', 'Term', Trimester)) %>%
  ggplot(aes(x = Sample_Name, y = maternal_contamination_norm_flip, color = Tissue)) +
  geom_point(size = 0.9) + 
  geom_segment(aes(y = 0, yend = maternal_contamination_norm_flip, 
                   x = Sample_Name, xend = Sample_Name),
               size = 0.5)+
  scale_color_manual(values = color_code_tissue,
                     labels = function(x)gsub(' cs', '', x),
                     guide = guide_legend(override.aes = list(size = 1))) +
  scale_y_continuous(limits = c(0,1), expand = c(0,0), breaks = c(0,0.5,1), 
                     labels = scales::percent) +
  scale_x_discrete(expand = c(0.05, 0.05)) +
  facet_grid(~Trimester, scales = 'free_x', space = 'free_x') +
  theme_bw(base_size = font_size-2) +
  theme(axis.text.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.grid.major.x = element_blank(),
        axis.ticks.x = element_blank(),
        panel.border = element_blank(),
        axis.line = element_line(),
        legend.key.size = unit(0.2, 'cm'),
        axis.title.y = element_text(angle = 0, vjust = 0.5),
        strip.background = element_blank())  +
  labs(x = 'Samples', y = 'Maternal\ncontamination', color = '');p
```

