---
title: "4_1_figure1"
author: "Victor Yuan"
date: "14/11/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

# Setup

## Libraries

```{r}
#install.packages('ggforce')
library(RColorBrewer)
library(tidyverse)
library(egg)
library(scales)
library(here)
library(ggrepel)
library(forcats)
```

## Data

```{r}
# parent directory
base_path <- file.path("Z:", "Victor", "Projects", "NIH - cells", "data", "main", "interim")

# sample data
pDat <- readRDS(file.path(base_path, '3_1_pDat_filt.rds'))

# methylation data
betas <- readRDS(file.path(base_path, '3_1_betas_filt.rds'))

# annotation
anno <- readRDS(file.path('Z:', 'Victor', 'Repositories', 'EPIC_annotation',
                          'hg19_epic_annotation.rds'))

#color code
color_code <- readRDS(file.path(base_path, '2_3_color_code.rds'))
color_code_tissue <- setNames(color_code$Colors_Tissue, color_code$label)
colors <- color_code_tissue[unique(pDat$Tissue)]

#list of genes for search list
gene <- tibble(gene = str_split(anno$genes_symbol, ', ') %>%
                 unlist() %>%
                 unique() %>%
                 na.omit() %>%
                 sort())

#pull out cpg name for search list use
cpg <- unique(anno$cpg) 
cpg <- cpg[cpg %in% rownames(betas)] %>% as_tibble()
```
# Figure 1 gene plot

```{r}
stat_plot <- function(cpg_name = NULL, gene_name = NULL, trimester = 'Third'){
  
  ######## FILTER ANNOTATION TO RELEVANT CPGS, DEPENDING ON INPUT #################
  if (is.null(cpg_name)){
     
    # If a gene symbol is supplied,
    # filter to gene in annotation
    anno_gene <- anno %>%
      filter(grepl(paste0('\\<', gene_name, '\\>'), genes_symbol),
             cpg %in% rownames(betas)) %>%
      arrange(desc(start))
    
    # stop if gene symbol is invalid
    if (nrow(anno_gene) == 0) {
      stop('No cpgs found, maybe they were filtered out.')
    }
  } else if (gene_name == ''){
      
      # If a cpg is provided,
      # filter annotation to cpgs
      anno_gene <- anno %>%
        filter(cpg %in% cpg_name,
               cpg %in% rownames(betas)) %>% 
        arrange(chr, desc(start))
      
      # stop if no cpgs were found
      if (nrow(anno_gene) == 0) {
        stop('None of the selected CpGs exist in our processed data.')
      }
      
      if (length(cpg_name) < length(intersect(cpg_name, rownames(betas)))) {
        print(paste0('The following selected CpGs do not exist in our processed data:\n', 
                     setdiff(cpg_name, rownames(betas))))
      }
      
  } else {
      stop('Enter valid cpg(s) OR one specific gene.')
  }
  
  ######## SETUP INPUT TO PLOTS ###############
  #
  # now we wrangle the betas/annotations/pdata information into a format usable for plotting
  #
  betas_gene <- t(betas[anno_gene$cpg,,drop = FALSE]*100) %>% as_tibble %>% 
    mutate(Sample_Name = colnames(betas)) %>%
    
    # reshape into longer format
    pivot_longer(cols = -Sample_Name,
                 names_to = 'cpg', 
                 values_to = 'beta')  %>%
    
    # add tissue and trimester info
    left_join(pDat %>% select(Sample_Name, Trimester, Tissue), by = 'Sample_Name') %>%
    
    # calculate mean and sd for each cpg for each group
    group_by(Tissue, Trimester, cpg) %>%
    summarize(mean = mean(beta),
              sd = sd(beta)) %>%
    mutate(lower = mean-sd, upper = mean+sd) %>%
    
    # add cpg info
    left_join(anno_gene, by = 'cpg') %>%
    
    # order on cpg position
    ungroup() %>%
    arrange(desc(start)) %>%
    mutate(cpg = factor(cpg, levels = unique(cpg)),
           Trimester_Tissue = paste0(Trimester, ' - ', Tissue)) %>%
    filter(Trimester == trimester)
  
  ##### GENERATE INDIVIDUAL PLOT TRACKS #####
  font_size <- 6
  point_size <- 0.5
  
  # Generate  methylation track
  p1 <- betas_gene %>%
    ggplot() +
    
    # geom_ribbon is to generate the alternating shaded background
    geom_ribbon(aes(x = as.numeric(cpg), ymin = 0, ymax = 25),
                fill = 'grey', alpha = 0.25)+
    geom_ribbon(aes(x = as.numeric(cpg), ymin = 50, ymax = 75),
                fill = 'grey', alpha = 0.25)+
    
    # geom_linerange and geom_point is for the actual methylation data
    geom_linerange(alpha = 0.5, size = point_size, 
                   aes(x = cpg, ymin =lower, ymax = upper, color = Tissue),
                   show.legend = FALSE) +
    geom_point(alpha = 0.75, aes(x = cpg, y = mean, color = Tissue), size = point_size*2) +
    
    # we want to facet by trimester
    facet_wrap(vars(Trimester), ncol = 1, labeller = label_both) +
    
    # cosmetics
    theme_bw(base_size = font_size) +
    theme(axis.text.x = element_blank(), 
          axis.text.y = element_text(size = font_size-2),
          axis.ticks = element_blank(),
          axis.title.y = element_text(angle = 0, vjust = 0.5, hjust = 0, size = font_size),
          
          legend.text = element_text(size = font_size),
          legend.key.size = unit(0, 'lines'),
          
          panel.grid.minor.y = element_blank(),
          panel.grid.major.y = element_blank(),
          panel.spacing.y = unit(0.5, 'cm'),
          panel.border = element_blank(),
          axis.line.y = element_line(color = 'black', size = 0.5, linetype = 'solid'),
          
          strip.background = element_blank(),
          strip.text = element_text(face = 'bold', hjust = 0),
          
          plot.margin=margin(l=-0.1, unit="cm")) +
    scale_y_continuous(limits = c(0, 100), 
                       expand = c(0, 0), 
                       breaks = c(0, 50, 100),
                       labels = function(x)paste0(x, '%')) +
    scale_x_discrete(expand = c(0.01,0.01))+
    scale_color_manual(values= color_code_tissue[unique(betas_gene$Tissue)],
                       guide = guide_legend(override.aes = list(size = point_size*4))) +
    labs(y = 'DNA\nmethylation', 
         x = '', 
         color = '', 
         title = if_else(!is.null(gene_name), gene_name, '')) 
    
  # plot cpg annotations
  p2 <- betas_gene %>% 
    
    # First we need to process the annotation data
    mutate(# absent/presence for different genomic elements
           enhancer = !is.na(enhancers_id),
           pmd = !is.na(pmd_id),
           imprinted_gene_placenta = !is.na(imprinted_gene_placenta),
           imprinted_gene_general = !is.na(imprinted_gene_general),
           imprinted_dmr_general = !is.na(imprinted_dmr_general),
           imprinted_dmr_placenta = !is.na(imprinted_dmr_placenta)) %>%
    mutate_if(is.logical, as.character) %>%
    select(cpg, enhancer, pmd, imprinted_gene_general:imprinted_dmr_placenta, cpg_id, start) %>%
    
    # reshape
    pivot_longer(cols = c(-cpg, -start),
               names_to = 'cpg_element', 
               values_to = 'presence') %>%
    distinct() %>%
    
    # arrange plot orde for each track
    arrange(desc(start)) %>%
    mutate(cpg = factor(cpg, levels = unique(cpg)),
           cpg_element = factor(cpg_element, 
                                levels = c('enhancer', 'pmd', 'imprinted_gene_placenta',
                                           'imprinted_gene_general', 'imprinted_dmr_general',
                                           'imprinted_dmr_placenta', 'cpg_id')),
           presence = factor(ifelse(presence == "TRUE", 'Present',
                                    ifelse(presence == "FALSE", 'Absent', presence)),
                             levels = c('Present', 'Absent', ' ', '  ', 
                                        'sea', 'shelf', 'shore', 'island')),
           # for facet label
           group = 'Annotations') %>%
    
    # plot code
    ggplot(aes(x = cpg, y = cpg_element, fill = presence)) +
    geom_tile(color = '#fcfcfc') +
    guides(fill=guide_legend(ncol=3, override.aes = list(colour = 'black'))) +
    facet_wrap(vars(group), ncol = 1) +
    theme_bw(base_size = font_size) +
    theme(axis.text.x = element_blank(),
          axis.ticks = element_blank(),
          axis.text.y = element_text(vjust = 0.5),
          
          legend.text = element_text(size = font_size),
          legend.key.height = unit(0.25, 'cm'),
          legend.key.width = unit(0.25, 'cm'),
          
          plot.margin = margin(l=-0.1,unit="cm"),
          panel.border = element_blank(),
          axis.line.y = element_line(color = 'black', size = 0.5, linetype = 'solid'),
          
          strip.background = element_blank(),
          strip.text = element_text(face = 'bold', hjust = 0)) +
    scale_fill_manual(values = c('Present' = '#cccccc', 'Absent' = '#f7f7f7', 
                                 ' ' = 'white', '  ' = 'white',
                                 'sea' = '#ffffcc', 'shelf' = '#a1dab4', 'shore' = '#41b6c4',
                                 'island' = '#225ea8'), drop = T) +
    scale_y_discrete(expand = c(0,0)) +
    scale_x_discrete(expand = c(0,0)) +
    labs(x = '', y = '', fill = '')
  
  # transcript annotations
  
  # cpgs can have more than one mapping to a gene element per transcript
  # cpgs mapped to intronexonboundaries can also be part of exons and introns
  # 5 UTR and 3 UTR can also overlap with exons and introns
  
  # If there are multiple mappings then only the following will be displayed in order of priority:
  # 5UTR / 3UTR, exon / intron, intronexonboundary
  
  p3 <- betas_gene %>% 
    select(cpg, end, genes_id, genes_tx_id) %>%
    distinct() %>%
    
    # split comma separate list and put each entry into a new row
    mutate(genes_tx_id = str_split(genes_tx_id, ', '),
           genes_id = str_split(genes_id, ', ')) %>%
    unnest(c(genes_tx_id, genes_id)) %>%
    
    # order transcripts by position of start
    arrange(end) %>%
    mutate(genes_tx_id = 
             factor(genes_tx_id, levels = unique(genes_tx_id)) %>%
             fct_explicit_na()) %>%
    
    # fill all combinations of cpg and gene_tx_id with NA if missing, so that they show in plot
    complete(cpg, genes_tx_id) %>%
    
    # filter out to unique mappings using the priority described above
    # strategy is to order by factor level and take the first occurence
    mutate(genes_id = factor(genes_id, 
                             levels = c('1to5kb', 'promoter', '5UTR', '3UTR', 'exon',
                                        'intron', 'intronexonboundary', 'intergenic'))) %>%
    
    group_by(cpg, genes_tx_id) %>%
    arrange(genes_id)  %>%
    dplyr::slice(1) %>%
    
    # reorder factor levels for appearance in plot legend
    mutate(genes_id = factor(genes_id, 
                             levels = c('1to5kb', 'promoter', '5UTR', 'exon', 'intron',
                                        'intronexonboundary', '3UTR', 'intergenic')),
           group = 'Transcripts') %>%
    
    #plot
    ggplot(aes(x = cpg, y = genes_tx_id, fill = genes_id)) +
    geom_tile(color = '#fcfcfc', alpha = 0.6) +
    facet_wrap(vars(group), ncol = 1) +
    guides(fill=guide_legend(ncol=2, override.aes = list(alpha=0.7, 
                                                         color = 'black'))) +
    theme_bw(base_size = font_size) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), 
          axis.text.y = element_text(vjust = 0.5),
          axis.ticks = element_blank(),
          
          plot.margin = margin(l=-0.1,unit="cm"),
          
          legend.text = element_text(size = font_size,
                                     margin = margin(t = -1, b = -1)),
          legend.key.height = unit(0.25, 'cm'),
          legend.key.width = unit(0.25, 'cm'),
          
          panel.grid = element_blank(),
          panel.border = element_blank(),
          axis.line = element_line(color = 'black'),
          axis.line.y = element_line(color = 'black', size = 0.5, linetype = 'solid'),
          
          strip.background = element_blank(),
          strip.text = element_text(face = 'bold', hjust = 0)) +
    scale_fill_brewer(na.value = '#f7f7f7', breaks = c('1to5kb', 'promoter', '5UTR', 'exon', 
                                                        'intron', 'intronexonboundary', '3UTR',
                                                        'intergenic'),
                      palette = 'Paired', direction = -1) +
    scale_y_discrete(expand = c(0,0), 
                     labels = function(x)(if_else(x == '(Missing)', '', x))) +
    scale_x_discrete(expand = c(0,0)) +
    labs(x = '', y = '', fill = '')
  
  # the size of each plot depends on the number of unique transcripts
  # More transcripts means a longer transcript track
  # so the plot height depends on the number of transcripts
  
  # number of transcripts
  n_trans <- betas_gene$genes_tx_id %>% str_split(', ') %>% unlist() %>% unique %>% length()
  
  # size of p1
  p1_h <- 20
  
  # combine tracks plot
  ggarrange(p1, p2, p3, ncol = 1, heights = c(p1_h-(n_trans-7), 7, n_trans))
}

p <- stat_plot(gene_name = 'DNMT1', trimester = 'Third');p

```

# Figure S* gene plots

```{r}

ps1 <- stat_plot(gene_name = 'DNMT3A', trimester = 'Third');ps1 #h:3.5, w:10

ps2 <- stat_plot(gene_name = 'DNMT3B', trimester = 'Third');ps2 #h:3.5, w: 4.25
ps3 <-  stat_plot(gene_name = 'DNMT3L', trimester = 'Third');ps3 #h:3.5, w: 3.5


ps4 <-  stat_plot(gene_name = 'TET1', trimester = 'Third');ps4 #h:3.5, w: 5
```


# Figure 1 enrichment

```{r}
base_path <- file.path('data', 'main','interim')

tests <- readRDS(here(base_path, '2_5_enrich_tests.rds'))


plot_bar <- function(x){
  x %>%
    # ordering in plot
    mutate(# linerange ymin ymax values
           ymin = pmin(Observed_p_in, Expected_p_in),
           ymax = pmax(Observed_p_in, Expected_p_in)) %>%
    
    ggplot() +
    
    # for cell type dmc enrichment lines and points
    geom_linerange(aes(x = genomic_feature, 
                        ymin = ymin, ymax = ymax, 
                        color = Celltype),
             stat = 'identity', 
             position = position_dodge(width = 0.75),
             #fatten = 0.01, 
             size = 0.75) +
    #geom_point(aes(x = genomic_feature,  y = Observed_p_in,
    #               color = Celltype, alpha = bonferroni001),
    #         stat = 'identity',
    #         position = position_dodge(width = 0.75),
             #fatten = 0.01, 
    #         size = 0.03) +
    
    # for expected frequency in legend
    geom_linerange(aes(x = genomic_feature, 
                      ymin = Expected_p_in, 
                      ymax = Expected_p_in, 
                      linetype = 'Expected frequency')) +
    
    # for generating the expected frequency bars
    geom_errorbar(aes(x = genomic_feature, 
                      ymin = Expected_p_in, 
                      ymax = Expected_p_in, 
                      linetype = 'Expected frequency'),
                  show.legend = FALSE) +

    # add significance asterisks
    geom_text(data = . %>%
                filter(bonferroni001),
              aes(x = genomic_feature,
                  y = ymax + 0.035, 
                  group = Celltype, 
                  size = bonferroni001),
              position = position_dodge(0.85), 
              label = '*') +
    
    facet_grid(genomic_feature_category~ Direction , 
               scales = 'free', space = 'free', drop = TRUE) +
    coord_flip() +
    theme_bw(base_size = 6) + 
    theme(legend.position = 'right', legend.direction = 'vertical',
          legend.spacing.y = unit(-0.2, 'cm'),
          strip.background = element_blank(),
          strip.text.x = element_text(face = 'bold'),
          strip.text.y = element_blank(),
          axis.ticks.y = element_blank(),
          panel.grid.major.y = element_blank(),
          panel.border = element_rect(color = 'grey'),
          panel.spacing.y = unit(0.15, 'cm')) +
    

    labs(color = '', x = '', y = '', shape = '', linetype = '', size = '',
         title = 'Percentage of cell-specific DMCs', subtitle = 'located in various genomic features') +
    scale_color_manual(values = colors,
                       guide = guide_legend(override.aes = list(size = 0.75),
                                            order = 3)) +
    scale_linetype(guide =guide_legend(order = 2,
                                       override.aes = list(size = 1))) +
    scale_size_manual(values = c('TRUE' = 1), 
                      na.translate = F, 
                      labels = c('Enriched\n(bonferroni p < 0.001)'), 
                      guide = guide_legend(override.aes = list(size = 4),
                                            order = 1)) +
    scale_y_continuous(limits = c(0,1), expand = c(0.075,0),
                       breaks = c(0, 0.5, 1),
                       labels = function(x)scales::percent(x, accuracy = 1)) 
    #scale_alpha_discrete(range = c(0.6,1), breaks = c('TRUE', 'FALSE'), guide = FALSE)
}

p1 <- tests %>% 
  filter(Trimester == 'Third',
         genomic_feature_category %in% c('pmd', 'gene', 'enhancer', 'cpg_island')) %>%
  arrange(genomic_feature_category, genomic_feature) %>%
  mutate(genomic_feature = as_factor(genomic_feature)) %>%
  plot_bar();p1

p2 <- tests %>% 
  filter(Trimester == 'First',
         genomic_feature_category %in% c('pmd', 'gene', 'enhancer', 'cpg_island')) %>%
  arrange(genomic_feature_category, genomic_feature) %>%
  mutate(genomic_feature = as_factor(genomic_feature)) %>%
  plot_bar()
```


# Number of DMCs

```{r}
dmcs <- readRDS(here(base_path, '2_4_dmcs.rds'))

# define cutoffs
p_thresh <- 0.01
b_thresh <- 0.25

# count number of significant cpgs per cell
dmcs_sig <- dmcs %>% 
  group_by(Group1) %>% 
  summarize(b001db50_all = sum(bonferroni < p_thresh & abs(delta_b) > b_thresh), 
            b001db50_hypo = sum(bonferroni < p_thresh & delta_b < -b_thresh),
            b001db50_hyper = sum(bonferroni < p_thresh & delta_b > b_thresh)) %>%
  separate(Group1, into = c('Trimester', 'Celltype'), sep = '\\.') %>%
  gather('group', 'n_cpg', -Trimester, -Celltype) %>%
  separate(group, into = c('pvalue', 'delta_b')) %>%
  mutate(Group1_label = paste0(Trimester, ' - ', case_when(
    Celltype == 'Endo_cs' ~ 'Endothelial cs',
    Celltype == 'Hofb_cs' ~ 'Hofbauer cs',
    Celltype == 'Strom_cs' ~ 'Stromal cs',
    Celltype == 'Troph_cs' ~ 'Trophoblasts cs'
  )))

dmcs_sig %>% 
  # calculate proportions
  spread(key = delta_b, value = n_cpg) %>%
  mutate(hyper_p = hyper*100/all,
         hypo_p = hypo*100/all) %>%
  select(-Trimester, -Celltype, -pvalue) %>%
  
  mutate(Hypermethylated = paste0(hyper, ' (', prettyNum(hyper_p, digits = 2), '%)'),
         Hypomethylated = paste0(hypo, ' (', prettyNum(hypo_p, digits = 2), '%)')) %>%
  
  # specify color of cells
  mutate(Hypermethylated = cell_spec(Hypermethylated, color = 'white', 
                                     background = spec_color(hyper_p, end = 0.5, direction = -1)),
         Hypomethylated = cell_spec(Hypomethylated, color = 'white', 
                                     background = spec_color(hypo_p, end = 0.5, direction = -1))) %>%
  separate(Group1_label, into = c('Trimester', 'Celltype'), sep = ' - ') %>%
  select(-(hyper:hypo_p)) %>%
  dplyr::rename(Num_sig = all) %>%
  kable(escape = F) %>%
  kable_styling(full_width = F) %>%
  column_spec(1, bold = T) %>%
  collapse_rows(columns = 1)

plot_object <- dmcs_sig %>% 
  # calculate proportions
  spread(key = delta_b, value = n_cpg) %>%
  select(-Trimester, -Celltype, -pvalue) %>%
  pivot_longer(cols = c(hyper, hypo),
               names_to = 'direction',
               values_to = 'n') %>%
  separate(Group1_label, into = c('Trimester', 'Celltype'), sep = ' - ') %>% {
    ggplot(data = ., aes(x = Celltype)) +
      geom_bar(stat = 'identity', aes(y = n, fill = direction)) +
      geom_text(data = select(., Trimester:all) %>%
                  distinct(),
                aes(label = scales::number(all), y = all),
                direction = 'y',
                size = 2,
                nudge_y = 15000) +
      facet_wrap(~Trimester, ncol = 1, labeller = 'label_both') +
      scale_x_discrete(labels = function(x)gsub(' cs', '', x)) +
      scale_y_continuous(labels = scales::number_format(), 
                         expand = c(0, 0),
                         limits = c(0, 160000)) +
      scale_fill_manual(values = c('hyper' = '#e6e6e6', 'hypo' = 'grey'), 
                        labels = c('Hypo-', 'Hyper-'),
                        breaks = c('hypo', 'hyper'),
                        guide = guide_legend(override.aes = list(size = 1))) +
      theme_bw(base_size = 6 ) +
      theme(panel.border = element_blank(),
            axis.text.x = element_blank(),
            axis.ticks = element_blank(),
            panel.grid = element_blank(),
            strip.background = element_blank(),
            strip.text = element_text(hjust = 0, face = 'bold'),
            legend.position = 'right') +
      coord_flip() +
      labs(x = '', y = '', fill = '')
    };plot_object
```

# PCA

```{r}
p <- pDat_filt %>%
  filter(Trimester != 'Second') %>%
  select(Sample_Name, Case_ID, Trimester, Tissue, contains('PC')) %>%
  ggplot(aes(x = PC1_raw, y = PC2_raw, color = Tissue)) +
  geom_point(alpha = 0.9, size = 0.5) +
  theme_bw(base_size = 6) +
  theme(panel.grid.minor = element_blank(),
        strip.background = element_blank(),
        axis.ticks = element_blank(),
        panel.grid = element_blank(),
        panel.border = element_blank(),
        axis.line = element_line(color = 'black'),
        legend.position = '') +
  scale_color_manual(values = color_code_tissue[pDat_filt$Tissue],
                     guide = guide_legend(override.aes = list(size = 3, alpha = 1),
                                          ncol = 3)) +
  labs(x = 'PC1 (41%)', y = 'PC2 (23%)', color = '') +
  coord_equal();p
  
```

# Number of DMCs in genes


just DNMTs and TET1 for supp figure

```{r}
dnmt <- anno %>%
  filter(grepl('DNMT|TET1', genes_symbol)) %>%
  inner_join(dmcs, by = c('cpg' = 'gene'))

         
dnmt %>%
  

  # boolean vector indicating if cpg is a differentially methylated for 1 or more cell type
  mutate(sig = (bonferroni < 0.01 & abs(delta_b > 0.1))) %>%
        # bonferroni = scales::pvalue(bonferroni, accuracy = 0.01), 
         #delta_b = scales::number(delta_b, accuracy = 0.01)) %>%
  #select(Group1, cpg, contains('gene'), sig, bonferroni, delta_b) %>% View()
  
  group_by(cpg) %>%
  summarize(genes_symbol = dplyr::first(genes_symbol), sig = any(sig)) %>%
  select(cpg, genes_symbol, sig) %>%
  
  # put gene symbol on rows
  separate_rows(genes_symbol, sep = ', ') %>%
  distinct() %>%
  
  # calculate % dmcs per gene
  group_by(genes_symbol) %>%
  summarize(p = sum(sig)/n())
```

all genes

```{r}
# what prop many cpgs are dm
dmcs %>% 
  group_by(gene) %>% 
  summarize(n = any(bonferroni < 0.01 & abs(delta_b) > 0.1)) %>% 
  summarize(sum(n)/n()) # 0.627

# what prop of cpgs are dm, outside of genes
dmcs %>% 
  filter(gene %in% (anno %>% filter(is.na(genes_symbol)) %>% pull(cpg))) %>%
  group_by(gene) %>% 
  summarize(n = any(bonferroni < 0.01 & abs(delta_b) > 0.1)) %>% 
  summarize(sum(n)/n()) # 0.800

# what prop of cpgs are dm, inside of genes
dmcs %>% 
  filter(gene %in% (anno %>% filter(!is.na(genes_symbol)) %>% pull(cpg))) %>%
  group_by(gene) %>% 
  summarize(n = any(bonferroni < 0.01 & abs(delta_b) > 0.1)) %>% 
  summarize(sum(n)/n()) # 0.574


all_genes <- anno %>%
  mutate(genes_symbol = gsub('NA, ', '', genes_symbol), # if cpg is otherwise associated with a gene,
         genes_symbol = ifelse(is.na(genes_symbol), 'NA', genes_symbol)) %>%
  inner_join(dmcs, by = c('cpg' = 'gene')) %>%
  
  # boolean vector indicating if cpg is a differentially methylated for 1 or more cell type
  mutate(sig = (bonferroni < 0.01 & abs(delta_b > 0.1))) %>%
        # bonferroni = scales::pvalue(bonferroni, accuracy = 0.01), 
         #delta_b = scales::number(delta_b, accuracy = 0.01)) %>%
  #select(Group1, cpg, contains('gene'), sig, bonferroni, delta_b) %>% View()
  
  group_by(cpg) %>%
  summarize(genes_symbol = dplyr::first(genes_symbol), sig = any(sig)) %>%
  select(cpg, genes_symbol, sig) %>%
  
  # put gene symbol on rows
  separate_rows(genes_symbol, sep = ', ') %>%
  distinct() %>%
  
  # calculate % dmcs per gene
  group_by(genes_symbol) %>%
  summarize(n_dm =  sum(sig),
            p_dm = n_dm/n(),
            n = n())

all_genes %>% 
  mutate(n_interval = cut(n, c(0, 10, 25, 50, 100, 250, Inf))) %>%
  ggplot(aes(x = n_interval, y = p_dm)) +
  geom_violin()

p3 <- all_genes %>% 
  mutate(n_interval = cut(n, c(0, 10, 25, 50, 100, 250, Inf))) %>%
  filter(n > 10, n < 250) %>%
  {
    ggplot(data = ., aes(x = n, y = n_dm)) +
      geom_point(alpha = 0.1, size = 0.1) +
      stat_smooth(se = FALSE, color = '#1b9e77', 
                  alpha = 0.25, size = 0.25, method = 'loess') +
      geom_abline(intercept = 0, slope = 1, color = '#7570b3', alpha = 1, size = 0.25) +
      
      # highlight genes
      geom_text_repel(data = . %>% 
                        filter(genes_symbol %in% c('DNMT1', 'DNMT3A', 'DNMT3B', 'DNMT3L', 'TET1')),
                      aes(label = genes_symbol),
                      segment.color = '#d95f02',
                      size = 1.5,
                      segment.size = 0.1,
                      nudge_y = 150) +
      geom_point(data = . %>% 
                        filter(genes_symbol %in% c('DNMT1', 'DNMT3A', 'DNMT3B', 'DNMT3L', 'TET1')),
                 color = '#c95353', size = 0.5) +
      
      theme_bw(base_size = 6) +
      theme(panel.grid = element_blank(),
            panel.border = element_blank(),
            axis.line = element_line(),
            axis.ticks = element_blank(),
            axis.title.y = element_text(angle = 0, vjust = 0.5, hjust = 0),
            plot.margin = margin(0.1, 0.5, 0.1, 0.1, "cm")) +
      scale_x_continuous(expand = c(0,0)) +
      scale_y_continuous(expand = c(0,0)) +
      coord_cartesian(xlim = c(10, 250),
                      ylim = c(0, 250)) +
      labs(x = 'CpGs per gene\n(n)', y = 'Differentially\nmethylated\n(n)')
    }; p3 # h: 1.25, w: 2.5

p4 <- all_genes %>%
  filter(n > 10) %>%
  ggplot(aes(x = p_dm)) +
  geom_histogram(binwidth = 0.05, color = 'white', size = 1,
                 fill = '#c2c2c2') +
  geom_vline(xintercept = all_genes %>%
               filter(n > 10) %>% summarize(m = mean(p_dm)) %>% pull(m), 
             color = 'black', 
             linetype = 'dashed',
             size = 0.75) + # average
  theme_bw(base_size = 6) +
  theme(axis.title.y = element_text(angle = 0, vjust = 0.5),
        panel.grid = element_blank(),
        panel.grid.major.y = element_line(color = 'grey'),
        panel.border = element_blank(),
        axis.ticks = element_blank(),
        axis.line = element_line()) +
  scale_x_continuous(expand = c(0,0), labels = scales::percent) +
  scale_y_continuous(expand = c(0,0)) +
  labs(x = 'Gene-wise percentage of CpGs\nthat are differentially methylated', y = 'Count');p4 #h: 1.25, w:2.5



p5 <- tibble(
  set = c('All', 'Outside genes', 'In genes'),
  `Differentially methylated` = c(0.627, 0.800, 0.574),
  `Not differentially methylated` = 1-`Differentially methylated`
) %>%
  pivot_longer(
    cols = c(`Differentially methylated`, `Not differentially methylated`),
    names_to = 'DM',
    values_to = 'p'
  ) %>%
  mutate(DM = fct_rev(DM)) %>%
  filter(DM == 'Differentially methylated') %>%
  {
    ggplot(data = ., aes(x = set, y = p, fill = DM)) +
      geom_bar(stat = 'identity') +
      geom_text(aes(label = scales::percent(p, accuracy = 1)),
                nudge_y = +0.1,
                size = 1.75) +
      theme_bw(base_size = 6) +
      theme(panel.border = element_blank(),
            panel.grid = element_blank(),
            axis.ticks = element_blank(),
            axis.text.y = element_blank(),
            axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, size = 4),
            axis.line.x = element_line()) +
      scale_y_continuous(expand = c(0,0), breaks = c(0,0.5,1), limits = c(0,1),
                         labels = scales::percent) +
      scale_fill_manual(values = c('grey'), guide = FALSE) +
      labs(x = '', fill = '', y = '', x = '') 
  };p5 #h: 1.25, w: 0.75

  geom_vline(xintercept = 0.627, color = 'black', linetype = 'dashed') + # out of all cpgs
  geom_vline(xintercept = 0.800, color = 'red', linetype = 'dashed') + # outside of genes
  geom_vline(xintercept = 0.574, color = 'green', linetype = 'dashed') + # inside genes
    
all_genes %>% 
  mutate(n_interval = cut(n, c(0, 10, 25, 50, 100, 250, Inf))) %>%
  arrange(desc(n_interval), desc(p_dm)) %>% DT::datatable()
```


