---
title: "4_2_figure_4_5"
author: "Victor Yuan"
date: "25/11/2019"
output: html_document
editor_options: 
  chunk_output_type: console
---

# PMDs

```{r, message = FALSE, warning = FALSE}
# libraries and data
fontsize <- 8
library(annotatr)
library(tidyverse); theme_set(theme_bw(base_size = fontsize))
library(glue)
library(scales)
library(DT)
library(viridisLite)
library(here)
library(ggsignif)


base_path <- file.path("data", "main", "interim")

pDat <- readRDS(here(base_path, '2_3_pDat_contam.rds'))
pDat <- pDat %>%
  mutate(Tissue = case_when(
    !(Tissue %in% c('Villi', 'Villi maternal', 'Syncytiotrophoblast')) ~ paste(Tissue, 'cs'),
    Tissue == 'Syncytiotrophoblast' ~ 'Trophoblasts enz',
    TRUE ~ Tissue
  )) 


# raw methylation data
betas <- readRDS(here(base_path, '1_4_betas_noob_filt.rds'))
colnames(betas) <- pDat$Sample_Name

# annotation
anno <- readRDS('Z:/Victor/Repositories/EPIC_annotation/hg19_epic_annotation.rds')

#color code
color_code <- readRDS(here(base_path, '2_3_color_code.rds'))
color_code_tissue <- setNames(color_code$Colors_Tissue, color_code$label)

pDat_filt <- pDat %>% 
  filter(maternal_contamination_norm_flip < 0.35,
         !Sample_Name %in% c('PM364_hofb_cs', 'PL293_v_R2', 'PM366_vc_R2', 'P131_hofb_cs',
                             'PM324_V1', 'PM324_V4'),
         !Tissue %in% c('Villi maternal', 'Trophoblasts enz', 'Mixture cs', 'Dead Cells and Lymphocytes cs'))

# filter to first trimester
betas_filt <- betas[,pDat_filt$Sample_Name]

densities_pmds_plots <- readRDS(here(base_path, '2_7_pmd_density_plots.rds'))
coverage <- readRDS(here(base_path, '2_7_pmd_coverage.rds')) 
densities_high_res <- readRDS(here(base_path, '2_7_pmd_high_res_density.rds'))

#color code
color_code <- readRDS(here(base_path, '2_3_color_code.rds'))
color_code_tissue <- setNames(color_code$Colors_Tissue, color_code$label)

w1 <- 5.5
f1 <- 8
e1 <- TRUE
```


# PMD intervals

## together


```{r, fig.height = 7}
p1 <- densities_high_res %>%
  
  # extract data
  filter(filt == 'filtered', 
         array == 'EPIC',
         Trimester != 'Second') %>%
  
  # remove density coords
  select(Trimester, Celltype, range, prop) %>%
  distinct() %>%
  
  # widen
  pivot_wider(id_cols = c(Trimester, Celltype),
              names_from = range,
              values_from = prop) %>%
  # sum intervals
  mutate(`(0, 0.2]` = `(-Inf,0.1]` + `(0.1,0.2]`,
         `(0.2, 0.4]` = `(0.2,0.3]` + `(0.3,0.4]`,
         `(0.4, 0.6]` =  `(0.4,0.5]` + `(0.5,0.6]`,
         `(0.6, 0.8]` = `(0.6,0.7]` + `(0.7,0.8]`,
         `(0.8, 1]` =  `(0.8,0.9]` + `(0.9, Inf]`) %>%
  select(Trimester, Celltype, `(0, 0.2]`:`(0.8, 1]`) %>%
  pivot_longer(cols = c(`(0, 0.2]`:`(0.8, 1]`),
               names_to = 'range',
               values_to = 'prop') %>%
  
  {
    ggplot(data = ., aes(x = Trimester, y = prop, col = Celltype)) +
      geom_point(size = 0.9) +
      geom_line(aes(group = Celltype), show.legend = FALSE) +
      facet_wrap(~range, ncol = 5,
                 labeller = labeller(
                   range = c('(0, 0.2]' = '0-20',
                             '(0.2, 0.4]' = '20-40',
                             '(0.4, 0.6]' = '40-60',
                             '(0.6, 0.8]' = '60-80',
                             '(0.8, 1]' = '80-100')
                 )) +
      scale_color_manual(values= color_code_tissue[unique(.$Celltype)],
                         labels = function(x)gsub(' cs', '', x),
                         guide = guide_legend(override.aes = list(size = 2))) +
      scale_x_discrete(labels = c('1', '3')) +
      scale_y_continuous(expand = c(0, 0), 
                         limits = c(0, 0.56), 
                         breaks = seq(0,1, by = 0.1), 
                         labels = function(x)percent(x, accuracy = 1)) +
      labs(color = '', 
           #title = 'Percent of total CpGs',
           #subtitle = 'in different methylation intervals',
           y = '') +
      theme(legend.position = 'right',
            legend.text = element_text(size = 6),
            legend.spacing.y = unit(0.1, 'mm'),
            legend.key.size = unit(0.5, 'lines'),
            axis.ticks.x = element_blank(),
            strip.background = element_blank(),
            strip.text = element_text(face = 'bold', size = 6),
            panel.grid = element_blank(),
            
            panel.border = element_blank(),
            axis.line = element_line(color = 'black'))
  };p1 # h: 2.25, w: 2.75
```

## by trimester

```{r, fig.height = 7}
p1 <- densities_high_res %>%
  
  # extract data
  filter(filt == 'filtered', 
         array == 'EPIC',
         Trimester != 'Second') %>%
  
  # remove density coords
  select(Trimester, Celltype, range, prop) %>%
  distinct() %>%
  
  # widen
  pivot_wider(id_cols = c(Trimester, Celltype),
              names_from = range,
              values_from = prop) %>%
  # sum intervals
  mutate(`(0, 0.2]` = `(-Inf,0.1]` + `(0.1,0.2]`,
         `(0.2, 0.4]` = `(0.2,0.3]` + `(0.3,0.4]`,
         `(0.4, 0.6]` =  `(0.4,0.5]` + `(0.5,0.6]`,
         `(0.6, 0.8]` = `(0.6,0.7]` + `(0.7,0.8]`,
         `(0.8, 1]` =  `(0.8,0.9]` + `(0.9, Inf]`) %>%
  select(Trimester, Celltype, `(0, 0.2]`:`(0.8, 1]`) %>%
  pivot_longer(cols = c(`(0, 0.2]`:`(0.8, 1]`),
               names_to = 'range',
               values_to = 'prop') %>%
  
  mutate(range_num = case_when(
    range == '(0, 0.2]' ~ 0.1,
    range == '(0.2, 0.4]' ~ 0.3,
    range == '(0.4, 0.6]' ~ 0.5,
    range == '(0.6, 0.8]' ~ 0.7,
    range == '(0.8, 1]' ~ 0.9
  )) %>%
  
  {
    ggplot(data = ., aes(x = range_num, y = prop, col = Celltype)) +
      geom_point(size = 1) +
      geom_line(aes(group = Celltype), show.legend = FALSE, size = 0.5) +
      facet_wrap(vars(Trimester), ncol = 1, 
                 labeller = labeller(Trimester = c('First' = 'First trimester',
                                                   'Third' = 'Term'))) +
      scale_color_manual(values= color_code_tissue[unique(.$Celltype)],
                         labels = function(x)gsub(' cs', '', x),
                         guide = guide_legend(override.aes = list(size = 2))) +
      scale_x_continuous(breaks = c(0.1, 0.3, 0.5, 0.7, 0.9),
                         labels = c('<20', '20-40', '40-60', '60-80', '>80')) +
      scale_y_continuous(expand = c(0, 0), 
                         limits = c(0, 0.56), 
                         breaks = seq(0,1, by = 0.25), 
                         labels = function(x)percent(x, accuracy = 1)) +
      labs(color = '', 
           #title = 'Percent of total CpGs',
           #subtitle = 'in different methylation intervals',
           y = 'Percentage of CpGs',
           x = 'DNAm interval (%)') +
      theme(legend.position = 'right',
            legend.text = element_text(size = fontsize),
            legend.spacing.y = unit(0.1, 'mm'),
            legend.key.size = unit(0.5, 'lines'),
            axis.ticks.x = element_blank(),
            strip.background = element_blank(),
            strip.text = element_text(face = 'bold', size = fontsize, hjust = 0),
            panel.grid = element_blank(),
            
            panel.border = element_blank(),
            axis.line = element_line(color = 'black'))
  };p1 # h: 2.5, w: 2.75
```

# imprinting

```{r}
imprint_density <- readRDS(here::here(base_path, '2_12_imprint_density.rds'))
imprint_density_allele <- readRDS(here::here(base_path, '2_12_imprint_density_allele.rds'))
```

## other and placental-specific

```{r}
p1a <- imprint_density %>%
  filter(imprint_tissue_specificity == 'other') %>%
  {
    ggplot(data = ., aes(x = x, y = y, color = Tissue)) +
      geom_line(size = 0.75,) +
      geom_area(data = . %>%
                  filter(x < 0.75, x > 0.25),
                aes(fill = Tissue, color = NULL),
                alpha = 0.35) +
      geom_text(data = . %>%
                   select(Tissue,  Trimester, 
                          imprint_tissue_specificity, contains('p')) %>%
                   distinct() %>%
                   mutate(p_25_75 = scales::percent(p_25_75, accuracy = 1)),
                 x = 0.5,
                 y = 3.5,
                 aes(label = p_25_75),
                 show.legend = FALSE,
                size = 2.5) +
      facet_grid(cols = vars(Tissue), 
                 rows = vars(Trimester),
                 labeller = labeller(.cols = function(x)gsub(' cs', '', x),
                                     .rows = function(x)ifelse(x == 'First', 
                                                               paste0(x, ' Trimester'),
                                                               'Term')),
                 switch = 'y') +
      scale_color_manual(values = color_code_tissue[unique(imprint_density$Tissue)], 
                         na.value = '#636363',
                         guide = 'none') +
      scale_fill_manual(values = color_code_tissue[unique(imprint_density$Tissue)], 
                         na.value = '#636363',
                         labels = function(x)gsub(' cs', '', x),
                        guide = 'none') +
      scale_y_continuous(expand = c(0,0),
                         limits = c(0,3.75)) +
      scale_x_continuous(#limits = c(0,1), 
                         expand = c(0,0), 
                         breaks = c(0, 0.25, 0.5, 0.75, 1), 
                         labels = c('0', '', '50%', '', '100%')) +
      theme_bw(base_size = fontsize) +
      theme(panel.border = element_blank(),
            panel.grid = element_blank(),
            panel.spacing.x = unit(2, "mm"),
            strip.background = element_blank(),
            strip.placement = 'outside',
            strip.text.y = element_text(angle = 180),
            axis.line = element_line(),
            axis.title.y = element_text(angle = 0, vjust = 0.5)) +
      labs(x = 'DNA methylation', y = '', fill = '', 
           title = 'Regions imprinted in multiple tissues')+
      annotate("segment", x=-Inf, xend=Inf, y=-Inf, yend=-Inf)
  };p1a # h: 2, w: 5.75

p1b <- imprint_density %>%
  filter(imprint_tissue_specificity == 'placental-specific') %>%
  {
    ggplot(data = ., aes(x = x, y = y, color = Tissue)) +
      geom_line(size = 0.75) +
      geom_area(data = . %>%
                  filter(x < 0.75, x > 0.25),
                aes(fill = Tissue, color = NULL),
                alpha = 0.35) +
      geom_text(data = . %>%
                   select(Tissue,  Trimester, 
                          imprint_tissue_specificity, contains('p')) %>%
                   distinct() %>%
                   mutate(p_25_75 = scales::percent(p_25_75, accuracy = 1)),
                 x = 0.5,
                 y = 3.5,
                 aes(label = p_25_75),
                 show.legend = FALSE,
                size = 2.5) +
      facet_grid(cols = vars(Tissue), 
                 rows = vars(Trimester),
                 labeller = labeller(.cols = function(x)gsub(' cs', '', x),
                                     .rows = function(x)ifelse(x == 'First', 
                                                               paste0(x, ' Trimester'),
                                                               'Term')),
                 switch = 'y') +
      scale_color_manual(values = color_code_tissue[unique(imprint_density$Tissue)], 
                         na.value = '#636363',
                         guide = 'none') +
      scale_fill_manual(values = color_code_tissue[unique(imprint_density$Tissue)], 
                         na.value = '#636363',
                         labels = function(x)gsub(' cs', '', x),
                        guide = 'none') +
      scale_y_continuous(expand = c(0,0)) +
      scale_x_continuous(#limits = c(0,1), 
                         expand = c(0,0), 
                         breaks = c(0, 0.25, 0.5, 0.75, 1), 
                         labels = c('0', '', '50%', '', '100%')) +
      theme_bw(base_size = fontsize) +
      theme(panel.border = element_blank(),
            panel.grid = element_blank(),
            panel.spacing.x = unit(2, "mm"),
            strip.background = element_blank(),
            strip.placement = 'outside',
            strip.text.y = element_text(angle = 180),
            axis.line = element_line(),
            axis.title.y = element_text(angle = 0, vjust = 0.5)) +
      labs(x = 'DNA methylation', y = '', fill = '', 
           title = 'Regions imprinted only in placenta')+
      annotate("segment", x=-Inf, xend=Inf, y=-Inf, yend=-Inf)
  };p1b # h: 2, w: 5.75
```

## Imprints by methylated allele

```{r}
p2a <- imprint_density_allele %>%
  filter(imprint_methylated_allele == 'M') %>%
  {
    ggplot(data = ., aes(x = x, y = y, color = Tissue)) +
      geom_line(size = 0.75) +
      geom_area(data = . %>%
                  filter(x < 0.75, x > 0.25),
                aes(fill = Tissue, color = NULL),
                alpha = 0.35) +
      geom_text(data = . %>%
                   select(-x, -y) %>%
                   distinct() %>%
                   mutate(p_25_75 = scales::percent(p_25_75, accuracy = 1)),
                 x = 0.5,
                 y = 3.5,
                 aes(label = p_25_75),
                size = 2.5,
                 show.legend = FALSE) +
      facet_grid(cols = vars(Tissue), 
                 rows = vars(Trimester),
                 labeller = labeller(.cols = function(x)gsub(' cs', '', x),
                   .rows = function(x)gsub(
                     'Third', 'Term', gsub(
                       'First', "First Trimester", x))),
                 switch = 'both') +
      scale_color_manual(values = color_code_tissue[unique(imprint_density$Tissue)], 
                         na.value = '#636363',
                         guide = 'none') +
      scale_fill_manual(values = color_code_tissue[unique(imprint_density$Tissue)], 
                         na.value = '#636363',
                         labels = function(x)gsub(' cs', '', x),
                        guide = 'none') +
      scale_y_continuous(expand = c(0,0), limits = c(0,4.9)) +
      scale_x_continuous(limits = c(0,1), 
                         expand = c(0,0), 
                         breaks = c(0, 0.25, 0.5, 0.75, 1), 
                         labels = c('', '', '50%', '', '100%')) +
      theme_bw(base_size = fontsize) +
      theme(panel.border = element_blank(),
            panel.grid = element_blank(),
            panel.spacing.x = unit(2, "mm"),
            strip.background = element_blank(),
            strip.placement = 'outside',
            strip.text.y = element_text(angle = 180),
            axis.line = element_line(),
            axis.title.y = element_text(angle = 0, vjust = 0.5)) +
      labs(x = 'DNA methylation', title = 'Maternal (# CpGs = 895)', y = '', fill = '')+
      annotate("segment", x=-Inf, xend=Inf, y=-Inf, yend=-Inf)
  };p2a

p2b <- imprint_density_allele %>%
  filter(imprint_methylated_allele == 'P') %>%
  {
    ggplot(data = ., aes(x = x, y = y, color = Tissue)) +
      geom_line(size = 0.75) +
      geom_area(data = . %>%
                  filter(x < 0.75, x > 0.25),
                aes(fill = Tissue, color = NULL),
                alpha = 0.35) +
      geom_text(data = . %>%
                   select(-x,-y) %>%
                   distinct() %>%
                   mutate(p_25_75 = scales::percent(p_25_75, accuracy = 1)),
                 x = 0.5,
                 y = 3.5,
                size = 2.5,
                 aes(label = p_25_75),
                 show.legend = FALSE) +
      facet_grid(cols = vars(Tissue), 
                 rows = vars(Trimester),
                 labeller = labeller(.cols = function(x)gsub(' cs', '', x),
                   .rows = function(x)gsub(
                     'Third', 'Term', gsub(
                       'First', "First Trimester", x))),
                 switch = 'both') +
      scale_color_manual(values = color_code_tissue[unique(imprint_density$Tissue)], 
                         na.value = '#636363',
                         guide = 'none') +
      scale_fill_manual(values = color_code_tissue[unique(imprint_density$Tissue)], 
                         na.value = '#636363',
                         labels = function(x)gsub(' cs', '', x),
                        guide = 'none') +
      scale_y_continuous(expand = c(0,0), limits = c(0,4.9)) +
      scale_x_continuous(limits = c(0,1), 
                         expand = c(0,0), 
                         breaks = c(0, 0.25, 0.5, 0.75, 1), 
                         labels = c('', '', '50%', '', '100%')) +
      theme_bw(base_size = fontsize) +
      theme(panel.border = element_blank(),
            panel.grid = element_blank(),
            panel.spacing.x = unit(2, "mm"),
            strip.background = element_blank(),
            strip.placement = 'outside',
            strip.text.y = element_text(angle = 180),
            axis.line = element_line(),
            axis.title.y = element_text(angle = 0, vjust = 0.5)) +
      labs(x = 'DNA methylation', title = 'Paternal\n(# CpGs = 190)', y = '', fill = '')+
      annotate("segment", x=-Inf, xend=Inf, y=-Inf, yend=-Inf)
  };p2b

```

## just by trimester

```{r}
p_i1 <- imprint_density %>%
  filter(Trimester == 'Third') %>%
  {
    ggplot(data = ., aes(x = x, y = y, color = Tissue)) +
      geom_line(size = 0.75) +
      geom_area(data = . %>%
                  filter(x < 0.75, x > 0.25),
                aes(fill = Tissue, color = NULL),
                alpha = 0.35) +
      geom_text(data = . %>%
                   select(Tissue, imprint_tissue_specificity, contains('p')) %>%
                   distinct() %>%
                   mutate(p_25_75 = scales::percent(p_25_75, accuracy = 1)),
                 x = 0.5,
                 y = 3.5,
                 aes(label = p_25_75),
                 show.legend = FALSE,
                size = 2.5) +
      facet_grid(cols = vars(Tissue), 
                 rows = vars(imprint_tissue_specificity),
                 labeller = labeller(
                   .cols = function(x)gsub(' cs', '', x),        
                   .rows = c('other' = 'Imprinted in\nmultiple tissues\n(# CpGs = 1085)',
                             'placental-specific' = 'Imprinted in\nplacenta\n(# CpGs = 981)')),
                 switch = 'y') +
      scale_color_manual(values = color_code_tissue[unique(imprint_density$Tissue)], 
                         na.value = '#636363',
                         guide = 'none') +
      scale_fill_manual(values = color_code_tissue[unique(imprint_density$Tissue)], 
                         na.value = '#636363',
                         labels = function(x)gsub(' cs', '', x),
                        guide = 'none') +
      scale_y_continuous(expand = c(0,0)) +
      scale_x_continuous(limits = c(0,1), 
                         expand = c(0,0), 
                         breaks = c(0, 0.25, 0.5, 0.75, 1), 
                         labels = c('', '', '50%', '', '100%')) +
      theme_bw(base_size = fontsize) +
      theme(panel.border = element_blank(),
            panel.grid = element_blank(),
            panel.spacing.x = unit(2, "mm"),
            strip.background = element_blank(),
            strip.placement = 'outside',
            strip.text.y = element_text(angle = 180),
            axis.line = element_line(),
            axis.title.y = element_text(angle = 0, vjust = 0.5)) +
      labs(x = 'DNA methylation', y = '', fill = '')+
      annotate("segment", x=-Inf, xend=Inf, y=-Inf, yend=-Inf)
  };p_i1## h: 2, w: 6

p_i1b <- imprint_density %>%
  filter(Trimester == 'First') %>%
  {
    ggplot(data = ., aes(x = x, y = y, color = Tissue)) +
      geom_line(size = 0.75) +
      geom_area(data = . %>%
                  filter(x < 0.75, x > 0.25),
                aes(fill = Tissue, color = NULL),
                alpha = 0.35) +
      geom_text(data = . %>%
                   select(Tissue, imprint_tissue_specificity, contains('p')) %>%
                   distinct() %>%
                   mutate(p_25_75 = scales::percent(p_25_75, accuracy = 1)),
                 x = 0.5,
                 y = 3.5,
                 aes(label = p_25_75),
                 show.legend = FALSE,
                size = 2.5) +
      facet_grid(cols = vars(Tissue), 
                 rows = vars(imprint_tissue_specificity),
                 labeller = labeller(
                   .cols = function(x)gsub(' cs', '', x),        
                   .rows = c('other' = 'Ubiquitous\n(# CpGs = 1085)',
                             'placental-specific' = 'Placental-specific\n(# CpGs = 981)')),
                 switch = 'both') +
      scale_color_manual(values = color_code_tissue[unique(imprint_density$Tissue)], 
                         na.value = '#636363',
                         guide = 'none') +
      scale_fill_manual(values = color_code_tissue[unique(imprint_density$Tissue)], 
                         na.value = '#636363',
                         labels = function(x)gsub(' cs', '', x),
                        guide = 'none') +
      scale_y_continuous(expand = c(0,0)) +
      scale_x_continuous(limits = c(0,1), 
                         expand = c(0,0), 
                         breaks = c(0, 0.25, 0.5, 0.75, 1), 
                         labels = c('', '', '50%', '', '100%')) +
      theme_bw(base_size = fontsize) +
      theme(panel.border = element_blank(),
            panel.grid = element_blank(),
            panel.spacing.x = unit(2, "mm"),
            strip.background = element_blank(),
            strip.placement = 'outside',
            strip.text.y = element_text(angle = 180),
            axis.line = element_line(),
            axis.title.y = element_text(angle = 0, vjust = 0.5)) +
      labs(x = 'DNA methylation', y = '', fill = '')+
      annotate("segment", x=-Inf, xend=Inf, y=-Inf, yend=-Inf)
  };p_i1b# h: 2.15, w: 6
```

# Specific Imprinted Regions

## set up igv and gene function

```{r}
#ect annotations for intersection with regions
# Note inclusion of custom annotation, and use of shortcuts
annots <- c('hg19_cpgs', 'hg19_basicgenes', 'hg19_genes_intergenic',
           'hg19_genes_intronexonboundaries')

# Build the annotations (a single GRanges object)
annotations <- annotatr::build_annotations(genome = 'hg19', annotations = annots)
annotations <- annotations %>%
  as_tibble() %>%
  mutate(type = gsub('genes_', '',
                     gsub('hg19_', '', type)),
         type = fct_relevel(type,
                            c('1to5kb', 'promoters', '5UTRs', 
                              'exons', 'introns', 'intronexonboundaries', 
                              '3UTRs',
                              'cpg_islands', 'cpg_shores', 'cpg_shelves', 'cpg_inter'))) 

anno <- anno %>%
  filter(cpg %in% rownames(betas_filt))

font_size <- 12
igv <- function(gene_name = NULL, 
                
                chr = NULL,
                start = NULL,
                end = NULL,
                
                start_extend = 2500, # how far to extend the plot left, in genomic coordinates 
                end_extend = 2500,   # how far to extend right
                
                show_transcript_gene = TRUE, # show transcripts associated gene symbol
                breaks_width = 15000, # determines the spacing of x axis breaks
                
                highlight_region = NULL, # a df with start end columns, indicating regions to highlight
                
                trimester = 'Third', # can be Third, First, or Both
                sex = 'Both', # can be Male, Female, or Both
                
                plot_sizes = c(3,1), # a 2-element numeric vector indicating the relative plot heights 
                                    # of the methylation plot to the transcript plot
                transcript_height = 1 # scale factor for height of transcript elements. e.g. 1.1 = 110%
                
                # required components:
                # pDat_filt with columns Sample_Name, Trimester, Sex, and Tissue
                # betas_filt data.frame with column names == pDat$Sample_Name
                # anno data frame that is contains a column cpg, genes_symbol, chr, start, end
                # the cpgs in anno and betas_filt need to be exactly the same, and same order
                # annotation data.frame that I built from annotatr, and slightly cleaned
                # colors, a named vector corresponding to the color codes for each celltype
                ) {
  
  # subset to relevant samples
  if (trimester %in% c('Third', 'First')) {
    pDat_filt <- pDat_filt %>%
      filter(Trimester == trimester)
  }
  
  if (!trimester %in% c('Third', 'First', 'Both')) {
    stop('trimester must be set to one of "Third", "First", or "Both"')
  }
  
  if (sex %in% c('Male', 'Female')) {
    pDat_filt <- pDat_filt %>%
      filter(Sex == sex)
  }
  
  if (!sex %in% c('Male', 'Female', 'Both')) {
    stop('sex must be set to one of "Male", "Female", or "Both"')
  }
  
   
  # subset to cpgs associated with gene or region
  
  if (!is.null(gene_name)){
    anno_subset <- anno %>%
      filter(grepl(paste0('\\<', gene_name, '\\>'), genes_symbol)) %>% 
      dplyr::select(chr, start, end) 
    
      c <- unique(anno_subset$chr)
      s <- min(anno_subset$start) # first cpg
      e <- max(anno_subset$start) # last cpg
      s_extend <- s - start_extend
      e_extend <- e + end_extend
  
  
  } else if (!is.null(chr) & !is.null(start) & !is.null(end)) {
    anno_subset <- anno %>%
      filter(chr == chr, start > start, end < end)
    
    c <- chr
    s <- start 
    e <- end 
    s_extend <- s - start_extend
    e_extend <- e + end_extend
  } else {
    stop('Input either a gene symbol, or a chromosome, start and end positions')
  }
  
  region_of_interest <- anno %>%
    filter(chr == c, start >= s_extend, end <= e_extend)
  
  # subset betas
  b <- betas_filt[region_of_interest$cpg, pDat_filt$Sample_Name] %>%
    as.data.frame() %>%
    bind_cols(cpg = rownames(.), .) %>%
    pivot_longer(cols = -cpg,
                 names_to = 'Sample_Name',
                 values_to = 'beta') %>%
    left_join(anno, by = 'cpg') %>% 
    right_join(pDat_filt, by = 'Sample_Name') 
  
  
  p1 <- b %>%
    
    # calculate mean beta for each cpg for each tissue
    group_by(cpg, Tissue) %>%
    summarize(mean = mean(beta), start = dplyr::first(start)) %>%
    
    # plot methylation means
    ggplot(aes(x = start)) +
    geom_linerange(ymin = 0, aes(ymax = mean, color = Tissue)) +
    
    facet_wrap(vars(Tissue), ncol = 1, strip.position = 'left',
               labeller = labeller(Tissue = function(x)gsub(' cs', '', x))) +
    
    theme_bw(base_size = font_size) +
    theme(strip.background = element_blank(),
          strip.text.y = element_text(angle = 180, hjust = 1),
          strip.placement = 'outside',
          axis.ticks = element_blank(),
          axis.text.x = element_blank(),
          axis.title.x = element_blank(),
          axis.line = element_line(size = 0.5),
          panel.border = element_blank(),
          panel.grid.minor = element_blank(),
          panel.spacing = unit(0, 'lines'),
          #panel.grid = element_blank()
          ) +
    scale_y_continuous(limits = c(0,1), breaks = c(0,1), 
                       expand = c(0,0), labels = NULL) +
    scale_x_continuous(limits = c(s_extend, e_extend), expand = c(0,0), 
                       breaks = breaks_width(breaks_width),
                       labels = scales::number) +
    scale_color_manual(values = color_code_tissue, guide = 'none') +
    labs(y = '', x = '', 
         title = ifelse(!is.null(gene_name), 
                        gene_name,
                        paste0(c, ':', s_extend, '-', e_extend)))
  
  p2 <- annotations %>%
    filter(seqnames == as.character(c), start > s_extend, end < e_extend,
           type %in% c('exons', 'introns', 'promoters')) %>%
    
    mutate(symbol = ifelse(is.na(symbol), '', symbol)) %>%
    ggplot(aes(x = start, xend = end)) +
    geom_segment(y= 0.5, yend = 0.5, aes(size = type, color = type)) +
    theme_bw(base_size = font_size)+
    theme(strip.background = element_blank(),
          strip.text.y  = element_text(angle = 180),
          strip.placement = 'outside',
          legend.position = 'bottom',
          plot.title =  element_blank(),
          plot.subtitle = element_blank(),
          axis.text.y = element_blank(),
          axis.ticks = element_blank(),
          axis.line = element_line(size = 0.5),
          panel.spacing = unit(0, 'lines'),
          panel.grid.major.y = element_blank(),
          panel.grid.minor.y = element_blank(),
          panel.grid.minor.x = element_blank(),
          panel.border = element_blank(),
          legend.key.size = unit(0.5, 'cm'),
          legend.spacing = unit(0.25, 'cm')) +
    scale_size_manual(values = c('exons' = 3, 'introns' = 1, 'intronexonboundaries' = 3,
                                 '5UTRs' = 5, '3UTRs' = 6,
                                 '1to5kb' = 2,
                                 'promoters' = 2,
                                 
                                 'cpg_inter' = 1,
                                 'cpg_islands' = 5,
                                 'cpg_shores' = 4,
                                 'cpg_shelves' = 3)*transcript_height) +
    scale_color_manual(values = c('exons' = 'black', 
                                  'introns' = 'black', 
                                  'intronexonboundaries' = 'black',
                                 '5UTRs' = 'grey', '3UTRs' = 'grey',
                                 '1to5kb' = 'grey',
                                 'promoters' = 'forestgreen',
                                 
                                 'cpg_inter' = 1,
                                 'cpg_islands' = 5,
                                 'cpg_shores' = 4,
                                 'cpg_shelves' = 3)) +
    scale_x_continuous(limits = c(s_extend, e_extend), 
                       expand = c(0,0), 
                       labels = scales::number,
                       breaks = breaks_width(breaks_width)) +
    scale_y_continuous(expand = c(0,0)) +
    labs(x = 'position', size = '', color = '')
  
  if (show_transcript_gene) {
    p2 <- p2 + 
      facet_wrap(~symbol + tx_id, ncol = 1, strip.position = 'left') 
  } else if (show_transcript_gene == FALSE) {
    p2 <- p2 + 
      facet_wrap(~tx_id, ncol = 1, strip.position = 'left') 
  }
  
  if (!is.logical(show_transcript_gene)) {
    stop('show_transcript_gene must be set to TRUE or FALSE')
  }
  
  if (!is.null(highlight_region)) {
    p1 <- p1 + 
      geom_rect(data = highlight_region,
                aes(xmin = start, xmax = end),
                ymin = 0, ymax = 1,
                alpha = 0.1)
    p2 <- p2 + 
      geom_rect(data = highlight_region,
                aes(xmin = start, xmax = end),
                ymin = 0, ymax = 1,
                alpha = 0.1)
  }
  egg::ggarrange(p1,p2, ncol = 1, heights = plot_sizes) #H:4.25, #w: 3.25

}
```

```{r}

plot_cpgs <- function(gene_name = NULL, c = NULL, s = NULL, e = NULL,
                      plot_heights = c(4, 1, 1),
                      title = NULL) {
  
  if (!is.null(gene_name)) {
    anno_subset <- anno %>%
        filter(grepl(paste0('\\<', gene_name, '\\>'), genes_symbol))
    
  } else {
    anno_subset <- anno %>%
      filter(chr == c,
             start > s,
             end < e)
  }
  
  if (nrow(anno_subset) == 0) {
    stop('no cpgs found')
  }
  
betas_gene <- t(betas_filt[anno_subset$cpg,,drop = FALSE]*100) %>% as_tibble %>% 
    mutate(Sample_Name = colnames(betas_filt)) %>%
    
    # reshape into longer format
    pivot_longer(cols = -Sample_Name,
                 names_to = 'cpg', 
                 values_to = 'beta')  %>%
    
    # add tissue and trimester info
    left_join(pDat_filt %>% dplyr::select(Sample_Name, Trimester, Tissue), 
              by = 'Sample_Name') %>%
    
    # calculate mean and sd for each cpg for each group
    group_by(Tissue, Trimester, cpg) %>%
    summarize(mean = mean(beta),
              sd = sd(beta)) %>%
    mutate(lower = mean-sd, upper = mean+sd) %>%
    
    # add cpg info
    left_join(anno_subset, by = 'cpg') %>%
    
    # order on cpg position
    ungroup() %>%
    arrange(start) %>%
    mutate(cpg = factor(cpg, levels = unique(cpg)),
           Trimester_Tissue = paste0(Trimester, ' - ', Tissue)) %>%
    filter(Trimester == 'Third')
  
  ##### GENERATE INDIVIDUAL PLOT TRACKS #####
  point_size <- 0.5
  font_size <- 8
  
  # Generate  methylation track
  p1 <- betas_gene %>%
    ggplot() +
    
    # geom_ribbon is to generate the alternating shaded background
    geom_ribbon(aes(x = as.numeric(cpg), ymin = 0, ymax = 25),
                fill = 'grey', alpha = 0.25)+
    geom_ribbon(aes(x = as.numeric(cpg), ymin = 50, ymax = 75),
                fill = 'grey', alpha = 0.25)+
    
    # geom_linerange and geom_point is for the actual methylation data
    geom_linerange(alpha = 0.5, size = point_size, 
                   aes(x = cpg, ymin =lower, ymax = upper, color = Tissue),
                   show.legend = FALSE) +
    geom_point(alpha = 0.75, aes(x = cpg, y = mean, color = Tissue), 
               size = point_size*2) +
    
    # we want to facet by trimester
    facet_wrap(vars(Trimester), ncol = 1, 
               labeller = labeller(Trimester = c('Third' = 'Term',
                                     'First' = 'First trimester'))) +
    
    # cosmetics
    theme_bw(base_size = font_size) +
    theme(#axis.text.x = element_text(angle = 90, vjust = 0.5, size = font_size-2),
      axis.text.x = element_blank(),
          axis.text.y = element_text(size = font_size-2),
          axis.ticks = element_blank(),
          axis.title.y = element_text(angle = 0, vjust = 0.5, hjust = 0, 
                                      size = font_size),
          
          legend.text = element_text(),
          legend.key.size = unit(0, 'lines'),
          
          panel.grid.minor.y = element_blank(),
          panel.grid.major.y = element_blank(),
      panel.grid.major.x = element_blank(),
          panel.spacing.y = unit(0.5, 'cm'),
          panel.border = element_blank(),
          axis.line = element_line(color = 'black', size = 0.5, linetype = 'solid'),
          
          strip.background = element_blank(),
          strip.text = element_text(face = 'bold', hjust = 0),
          
          plot.margin=margin(l=-0.1, unit="cm")) +
    scale_y_continuous(limits = c(0, 100), 
                       expand = c(0, 0), 
                       breaks = c(0, 50, 100),
                       labels = function(x)paste0(x, '%')) +
    scale_x_discrete(expand = c(0.01,0.01))+
    scale_color_manual(values= color_code_tissue[unique(betas_gene$Tissue)],
                       guide = guide_legend(override.aes = list(size = point_size*2)),
                       labels = function(x)gsub(' cs', '', x)) +
    labs(y = 'DNA\nmethylation', 
         x = '', 
         color = '', 
         title = ifelse(!is.null(title),
                        title,
                        ifelse(!is.null(gene_name),
                               gene_name, 
                               paste0(c, ':', s, '-', e)))) 
  
  # plot cpg annotations
  p2 <- betas_gene %>% 
    
    # First we need to process the annotation data
    mutate(# absent/presence for different genomic elements
           enhancer = !is.na(enhancers_id),
           pmd = !is.na(pmd_id),
           imprinted_dmr_general = 
             !is.na(imprint_tissue_specificity) & imprint_tissue_specificity == 'other',
           imprinted_dmr_placenta = 
             !is.na(imprint_tissue_specificity) & imprint_tissue_specificity ==
             'placental-specific')%>%
    
    dplyr::select(cpg, enhancer, pmd, 
           imprinted_dmr_general, imprinted_dmr_placenta, cpg_id, start) %>%
    mutate_if(is.logical, as.character) %>%
    
    # reshape
    pivot_longer(cols = c(-cpg, -start),
               names_to = 'cpg_element', 
               values_to = 'presence') %>%
    distinct() %>%
    
    # arrange plot orde for each track
    arrange(start) %>%
    mutate(cpg = factor(cpg, levels = unique(cpg)),
           cpg_element = factor(cpg_element, 
                                levels = c('enhancer', 'pmd', 'imprinted_dmr_general',
                                           'imprinted_dmr_placenta', 'cpg_id')),
           presence = factor(ifelse(presence == "TRUE", 'Present',
                                    ifelse(presence == "FALSE", 'Absent', presence)),
                             levels = c('sea', 'shelf', 'shore', 'island',
                                        'Present', 'Absent', ' ', '  ')),
           # for facet label
           group = 'Annotations') %>%
    
    # plot code
    ggplot(aes(x = cpg, y = cpg_element, fill = presence)) +
    geom_tile(color = '#fcfcfc') +
    guides(fill=guide_legend(ncol=3, override.aes = list(colour = 'black'))) +
    facet_wrap(vars(group), ncol = 1) +
    theme_bw(base_size = font_size) +
    theme(axis.text.x = element_blank(),
          axis.ticks = element_blank(),
          axis.text.y = element_text(vjust = 0.5),
          
          legend.text = element_text(),
          legend.key.height = unit(0.25, 'cm'),
          legend.key.width = unit(0.25, 'cm'),
          
          plot.margin = margin(l=-0.1,unit="cm"),
          panel.border = element_blank(),
          axis.line.y = element_line(color = 'black', size = 0.5, linetype = 'solid'),
          
          strip.background = element_blank(),
          strip.text = element_text(face = 'bold', hjust = 0)) +
    scale_fill_manual(values = c('Present' = '#cccccc', 'Absent' = '#f7f7f7', 
                                 ' ' = 'white', '  ' = 'white',
                                 'sea' = '#ffffcc', 'shelf' = '#a1dab4', 'shore' = '#41b6c4',
                                 'island' = '#225ea8'), drop = F,
                      labels = stringr::str_to_title) +
    scale_y_discrete(expand = c(0,0),
                     labels = function(x)gsub('cpg_id', 'Relation to CpG Islands',
                       gsub('imprinted_dmr_placenta', 'Imprinted DMR (Placenta)',
                       gsub('imprinted_dmr_general', 'Imprinted DMR (General)',
                       gsub('pmd', 'Placental PMD',
                       gsub('enhancer', 'Enhancer', 
                            x)))))) +
    scale_x_discrete(expand = c(0,0)) +
    guides(fill=guide_legend(ncol=2, override.aes = list(colour = 'white'))) +
    labs(x = '', y = '', fill = '') 
  
   p3 <- betas_gene %>% 
    dplyr::select(cpg, end, genes_id, genes_tx_id, genes_symbol) %>%
    distinct() %>%
    
    # split comma separate list and put each entry into a new row
    mutate(genes_tx_id = str_split(genes_tx_id, ', '),
           genes_id = str_split(genes_id, ', '),
           genes_symbol = str_split(genes_symbol, ', ')) %>%
    unnest(c(genes_tx_id, genes_id, genes_symbol)) %>%
    
    # paste together transcript ID and genes symbol
    mutate(genes_tx_id_symbol = if_else(genes_symbol != 'no_associated_gene', 
                                        paste0(genes_tx_id, ' (', genes_symbol, ')'),
                                        genes_tx_id)) %>%
    
    ### ORDER TRANSCRIPTS IN PLOT 
    #
    # note that the first level appears on the bottom of the graph (y = 0 if it was continuous)
    
    # calculate earliest end of each symbol
    group_by(genes_symbol) %>%
    mutate(genes_symbol_end = min(end)) %>%
    ungroup() %>%
    
    # reorder gene symbols by their earliest end position
    mutate(genes_symbol = fct_reorder(genes_symbol, genes_symbol_end)) %>%
    
    # reorder transcripts based on gene symbol levels and then by their end position
    arrange(genes_symbol, end) %>%
    mutate(genes_tx_id_symbol = 
             factor(genes_tx_id_symbol, levels = unique(genes_tx_id_symbol)) %>%
             fct_explicit_na()) %>%
    
    ####
    # fill all combinations of cpg and gene_tx_id with NA if missing, so that they show in plot
    complete(cpg, genes_tx_id_symbol) %>%
    
    # filter out to unique mappings using the priority described above
    # strategy is to order by factor level and take the first occurence
    mutate(genes_id = factor(genes_id, 
                             levels = c('1to5kb', 'promoter', '5UTR', '3UTR', 'exon',
                                        'intron', 'intronexonboundary', 'intergenic'))) %>%
    
    group_by(cpg, genes_tx_id_symbol) %>%
    arrange(genes_id)  %>%
    dplyr::slice(1) %>%
    
    # reorder factor levels for appearance in plot legend
    mutate(genes_id = factor(genes_id, 
                             levels = c('1to5kb', 'promoter', '5UTR', 'exon', 'intron',
                                        'intronexonboundary', '3UTR', 'intergenic')),
           group = 'Transcripts') %>%
    
    #plot
    ggplot(aes(x = cpg, y = genes_tx_id_symbol, fill = genes_id)) +
    geom_tile(color = '#fcfcfc', alpha = 0.6) +
    facet_wrap(vars(group), ncol = 1) +
    guides(fill=guide_legend(ncol=2, override.aes = list(alpha=0.7, 
                                                         color = 'black'))) +
    theme_bw(base_size = font_size) +
    theme(
      axis.text.x = element_blank(),
          axis.text.y = element_text(vjust = 0.5),
          axis.ticks = element_blank(),
          
          plot.margin = margin(l=-0.1,unit="cm"),
          
          legend.text = element_text(margin = margin(t = -1, b = -1)),
          legend.key.height = unit(0.25, 'cm'),
          legend.key.width = unit(0.25, 'cm'),
          
          panel.grid = element_blank(),
          panel.border = element_blank(),
          axis.line = element_line(color = 'black'),
          axis.line.y = element_line(color = 'black', size = 0.5, linetype = 'solid'),
          
          strip.background = element_blank(),
          strip.text = element_text(face = 'bold', hjust = 0)) +
    scale_fill_manual(na.value = '#f7f7f7', 
                      breaks = c('1to5kb', 'promoter', '5UTR', 'exon',
                                 'intron', 'intronexonboundary', '3UTR',
                                 'intergenic'),
                      values = c('1to5kb' = '#FF7F00',
                                 'promoter' = "#FDBF6F", 
                                 '5UTR' = "#E31A1C",
                                 'exon' = "#FB9A99",
                                 'intron' = "#33A02C",
                                 'intronexonboundary' = "#B2DF8A",
                                 '3UTR'= "#1F78B4",
                                 'intergenic' = "#A6CEE3")) +
    scale_y_discrete(expand = c(0,0), 
                     labels = function(x)(if_else(x == '(Missing)', '', x))) +
    scale_x_discrete(expand = c(0,0)) +
    guides(fill=guide_legend(override.aes = list(colour = 'white'))) +
    labs(x = 'CpG', y = '', fill = '')

egg::ggarrange(p1, p2, p3, heights = plot_heights)
  
}
```

## plot genes

```{r}
#C19MC
p_c19 <- plot_cpgs(c ='chr19', s = 54150515, e = 54155608+15000,
                   title = 'C19MC imprinted DMR') 
#fig_height <- 3.75
#fig_width <- 4.25

igv(chr= 'chr19', start = 54150515, end= 54155608+15000,
    start_extend = 100,
    highlight_region = data.frame(start = 54150515, end = 54155608),
    breaks_width = 10000)

#RASGRF1
p_ras <- plot_cpgs(c = 'chr15', s = 79382548-45000, e = 79383980,
          title = 'RASGRF1')

#DCAF10
p_dcaf10 <- plot_cpgs(gene_name = 'DCAF10')

#JMJD1C
p_jmjd1c <- plot_cpgs(c = 'chr10', s = 65224441-25000, e = 65225999+25000,
          title = 'JMJD1C',
          plot_heights = c(4, 1, 0.75))

#h: 3.75. w:5
p_fgf8 <- plot_cpgs(gene_name = 'FGF8', plot_heights = c(4, 1, 2))
p_fgf12 <- plot_cpgs(c = 'chr3', s = 192124589-35000, e = 192127457+35000, 
                     plot_heights = c(4, 1, 0.5),
                     title = 'FGF12')

p_fgf12 <- plot_cpgs(c = 'chr13', s = 102568126-50000, e = 102569981+100000, 
          plot_heights = c(4, 1, 0.75),
          title = 'FGF14')



```


# PMD regions

```{r}
densities_pmds_plots <- readRDS(here(base_path, '2_7_pmd_density_plots.rds'))
coverage <- readRDS(here(base_path, '2_7_pmd_coverage.rds'))
densities_high_res <- readRDS(here(base_path, '2_7_pmd_high_res_density.rds'))
```

```{r}
# define plotting function
plot_region <- function(c, s, e, span = 0.1, cpg_size = NULL,
                        filtered = TRUE, array = 'EPIC') {
  
  # determine which cpgs to use, based on c, s, e
  if (is.null(cpg_size)) {
    cpg_size <- (e-s)*0.0005
  }
  
  # requires anno, betas_filt, anno_450k, pDat_filt
  if (filtered) {
    region1 <- anno %>%
      filter(chr == c, between(start, s, e),
             cpg %in% rownames(betas_filt),
             !cpg_id %in% c('island', 'shore'),
             !grepl('promoter', genes_id)) %>%
      pull(cpg)
  } else {
    region1 <- anno %>%
      filter(chr == c, between(start, s, e),
             cpg %in% rownames(betas_filt)) %>%
      pull(cpg)
  }
  
  if (array == '450K') {
    region1 <- intersect(anno_450k$Name, region1)
  } 
  
  # subset to betas to cpg in region, then make in plottable format
  region1_data <- betas_filt[region1, ] %>%
    
    t() %>%
    as.data.frame() %>%
    bind_cols(Sample_Name = rownames(.), .) %>%
    
    # melt
    gather(cpg, beta, -Sample_Name) %>%
    
    # add trimester and sex
    inner_join(pDat_filt %>% select(Sample_Name, Trimester, Tissue)) %>%
    
    # add cpg coordinates
    inner_join(anno %>% select(cpg, chr, pos = start)) %>%
    
    # set a grouping variable to control alpha for placenta/troph
    mutate(alpha_group = if_else(Tissue %in% c('Trophoblasts cs', 'Villi'),
                                 1, 0.6)) %>%
    
    # remove second trimester data
    filter(!Trimester %in% c('First', 'Second'))
  
  # get pmd start/end for plotting pmd track
  pmd_data <- anno %>%
    filter(chr == c, between(start, s, e), 
           !is.na(pmd_id)) %>%
    select(pmd_id) %>%
    separate(pmd_id, into = c('chr', 'start', 'end'), remove = FALSE) %>%
    mutate(chr = factor(chr, levels = paste0('chr', c(1:22, 'X'))),
           start = as.numeric(start),
           end = as.numeric(end)) %>%
    arrange(chr, start) %>%
    distinct()
  
  ggplot(region1_data) +
    
    # pmd track
    geom_rect(data = pmd_data, 
              aes(xmin = start, xmax = end), 
              ymin = 0, ymax = 1, fill = '#e8e8e8', alpha = 1) +
    
    # methylation track
    geom_line(stat = 'smooth', span = span, method = 'loess', se = FALSE,
              aes(x = pos, y = beta, color = Tissue, alpha = alpha_group),
              size = 0.6) +
    
    theme(panel.grid = element_blank(),
          panel.border = element_blank(),
          axis.line = element_line(color = 'black'),
          axis.ticks.x =element_blank()) +
    scale_color_manual(values= color_code_tissue[unique(region1_data$Tissue)],
                       labels = function(x)gsub(' cs', '', x),
                       guide = FALSE) +
    scale_y_continuous(expand = c(0,0), 
                       limits = c(0, 1), 
                       breaks = c(0, 0.5, 1), 
                       labels = function(x)percent(x, accuracy = 1)) +
    scale_x_continuous(limits = c(s, e), labels = function(x){x/10^6},
                       expand = c(0, 0)) +
    scale_alpha_identity() +
    labs(x = 'Position (Mbp)', y = 'DNAm',
         title = c, color = '')
}

p3 <- plot_region(c = 'chr4', s = 1500000, e = 8500000, 
            span = 0.07, filtered = FALSE, array = 'EPIC')

p2 <- plot_region(c = 'chr21', s = 25000000, e = 47000000, span = 0.05, filtered = T, #cpg_size = 5000,
            array = 'EPIC') #h:1.1, w: 2.75
```

# Repetitive elements & global

```{r}
repeat_plot_data <- readRDS(here::here(base_path, '2_13_repeat_plot_data.rds'))
global_data <- readRDS(here::here(base_path, '2_13_global_data.rds'))

p_re_1 <- repeat_plot_data %>%
  filter(Repeat %in% c('Alu', 'L1')) %>%
  
  ggplot(aes(x = Repeat, y = mean, color = Tissue)) +
  geom_violin(position = position_dodge(width = 0.75),
              draw_quantiles = 0.5,
              size = 0.25,
              scale = 'width') +
  geom_jitter(alpha = 1, 
              size = 0.01,
              position = position_jitterdodge(dodge.width = 0.75, jitter.width = 0.05))+
  facet_grid(cols = vars(Trimester), labeller = label_both) +
  scale_color_manual(values = color_code_tissue[unique(repeat_plot_data$Tissue)], 
                         na.value = '#636363',
                         guide = 'none') +
  scale_y_continuous(expand = c(0,0),
                     limits = c(0.47, 0.9), 
                     breaks = c(0.5, 0.75, 1), 
                     labels = percent) +
  labs(color = '', x = '', y = 'Mean DNAm') +
  theme(strip.background = element_blank(),
        strip.text = element_text(face = 'bold'),
        axis.title.y = element_text(angle = 0, vjust = 0.5),
        panel.border = element_blank(),
        axis.line = element_line(),
        panel.grid.minor = element_blank(),
        panel.grid.major.x = element_blank()); p_re_1 #h:1.75, w:3

p_re_2 <- repeat_plot_data %>%
  filter(Repeat_class %in% c('LTR')) %>%
  ggplot(aes(x = Repeat, y = mean, color = Tissue)) +
  geom_violin(position = position_dodge(width = 0.75),
              draw_quantiles = 0.5,
              size = 0.25,
              scale = 'width') +
  geom_jitter(alpha = 1, 
              size = 0.01,
              position = position_jitterdodge(dodge.width = 0.75, jitter.width = 0.05))+
  facet_wrap(vars(Trimester), ncol = 1, labeller = label_both) +
  scale_color_manual(values = color_code_tissue[unique(repeat_plot_data$Tissue)], 
                         na.value = '#636363',
                         guide = 'none') +
  scale_y_continuous(expand = c(0,0),
                     limits = c(0.34, 0.8), 
                     breaks = c(0.5, 0.75, 1), 
                     labels = percent) +
  labs(color = '', x = '', y = 'Mean DNAm') +
   theme(strip.background = element_blank(),
        strip.text = element_text(face = 'bold'),
        axis.title.y = element_text(angle = 0, vjust = 0.5),
        panel.grid.minor = element_blank(),
        panel.grid.major.x = element_blank()); p_re_2 #h:3, w:6

# Just REMP
p_re_3 <- repeat_plot_data %>%
  filter(grepl('REMP', Repeat),
         Trimester == 'Third') %>%
  ggplot(aes(x = Repeat, y = mean, color = Tissue)) +
  geom_violin(position = position_dodge(width = 0.75),
              draw_quantiles = 0.5,
              size = 0.25,
              scale = 'width') +
  geom_jitter(alpha = 1, 
              size = 0.01,
              position = position_jitterdodge(dodge.width = 0.75, jitter.width = 0.05))+
  geom_signif(color = 'black',
              textsize = 1.5,
              size = 0.25,
              annotations = c("p<0.001", "p<0.001"),
              y_position = c(0.96, 0.935),
              xmin=c(0.7,  0.85), xmax=rep(1.3, 2))+
  geom_signif(color = 'black',
              textsize = 1.5,
              size = 0.25,
              annotations = c("p<0.001", "p<0.001", "p<0.001"),
              y_position = c(0.96, 0.935, 0.91),
              xmin=c(1.7,  1.85, 2), xmax=rep(2.3, 3))+
  scale_color_manual(values = color_code_tissue[unique(repeat_plot_data$Tissue)], 
                         na.value = '#636363',
                         guide = 'none') +
  scale_y_continuous(expand = c(0,0),
                     limits = c(0.65, 1), 
                     breaks = c(0.5, 0.75, 1), 
                     labels = percent) +
  labs(color = '', x = '', y = 'Mean DNAm') +
  theme(strip.background = element_blank(),
        strip.text = element_text(face = 'bold'),
        axis.title.y = element_text(angle = 0, vjust = 0.5),
        panel.border = element_blank(),
        axis.line = element_line(),
        panel.grid.minor = element_blank(),
        panel.grid.major.x = element_blank()); p_re_3 #h:1.75, w:3

p_re_4 <- global_data %>%
  ggplot(aes(x = Measure, y = mean, color = Tissue)) +
  geom_violin(position = position_dodge(width = 0.75),
              draw_quantiles = 0.5,
              size = 0.25,
              scale = 'width') +
  geom_jitter(alpha = 1, 
              size = 0.01,
              position = position_jitterdodge(dodge.width = 0.75, jitter.width = 0.05))+
  facet_grid(cols = vars(Trimester), labeller = label_both) +
  scale_color_manual(values = color_code_tissue[unique(pDat_filt$Tissue)], 
                         na.value = '#636363',
                         guide = 'none') +
  scale_y_continuous(expand = c(0,0),
                     #limits = c(0.4, 0.6), 
                     breaks = c(0, 0.4, 0.5, 0.6, 0.7, 0.8, 1), 
                     labels = percent) +
  labs(color = '', x = '', y = 'Mean DNAm') +
  theme(strip.background = element_blank(),
        strip.text = element_text(face = 'bold'),
        axis.title.y = element_text(angle = 0, vjust = 0.5),
        panel.border = element_blank(),
        axis.line = element_line(),
        panel.grid.minor = element_blank(),
        panel.grid.major.x = element_blank());p_re_4  #h:1.75, w:5.5

# just 3rd, figure 3E
p_re_5 <- global_data %>%
  filter(Trimester == 'Third') %>%
  ggplot(aes(x = Measure, y = mean, color = Tissue)) +
  geom_violin(position = position_dodge(width = 0.75),
              draw_quantiles = 0.5,
              size = 0.25,
              scale = 'width') +
  geom_jitter(alpha = 1, 
              size = 0.1,
              position = position_jitterdodge(dodge.width = 0.75, jitter.width = 0.075))+
  geom_signif(color = 'black',
              textsize = 2,
              size = 0.25,
              annotations = c("p<0.001", "p=0.04"),
              y_position = c(0.8, 0.72),
              xmin=c(2.85,  3.15), xmax=rep(3.33, 2))+
  geom_signif(color = 'black',
              textsize = 2,
              size = 0.25,
              annotations = c("p<0.001", "p<0.001", "p=0.02"),
              y_position = c(0.9, 0.84, 0.78),
              xmin=c(1.74, 1.85,  2.15), xmax=rep(2.33, 3))+
  geom_signif(color = 'black',
              size = 0.25,
              textsize = 2,
              annotations = c("p<0.001", "p<0.001", "p=0.036"),
              y_position = c(0.675, 0.625, 0.575),
              xmin=c(0.74, 0.85,  1), xmax=rep(1.33, 3))+
  scale_color_manual(values = color_code_tissue[unique(pDat_filt$Tissue)], 
                         na.value = '#636363',
                         guide = 'none') +
  scale_y_continuous(expand = c(0,0),
                     limits = c(0.47, 0.92), 
                     breaks = c(0, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1), 
                     labels = percent) +
  labs(color = '', x = '', y = 'Mean DNAm') +
  theme(strip.background = element_blank(),
        strip.text = element_text(face = 'bold'),
        axis.title.y = element_text(angle = 0, vjust = 0.5),
        panel.border = element_blank(),
        axis.line = element_line(),
        axis.ticks = element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major.x = element_blank());p_re_5  #h:1.75, w:5.75

# just first trim, Figure S3
p_re_6 <- global_data %>%
  filter(Trimester == 'First') %>%
  ggplot(aes(x = Measure, y = mean, color = Tissue)) +
  geom_violin(position = position_dodge(width = 0.75),
              draw_quantiles = 0.5,
              size = 0.25,
              scale = 'width') +
  geom_jitter(alpha = 1, 
              size = 0.2,
              position = position_jitterdodge(dodge.width = 0.75, jitter.width = 0.05))+
  scale_color_manual(values = color_code_tissue[unique(pDat_filt$Tissue)], 
                         na.value = '#636363',
                         guide = 'none') +
  scale_y_continuous(expand = c(0,0),
                     limits = c(0.42, 0.85), 
                     breaks = c( 0.5, 0.6, 0.7, 0.8, 0.9, 1), 
                     labels = percent) +
  labs(color = '', x = '', y = 'Mean DNAm', title = 'First trimester') +
  theme(strip.background = element_blank(),
        strip.text = element_text(face = 'bold'),
        axis.title.y = element_text(angle = 0, vjust = 0.5),
        panel.border = element_blank(),
        axis.line = element_line(),
        panel.grid.minor = element_blank(),
        panel.grid.major.x = element_blank());p_re_6  #h:1.75, w:5.5
```


# mQTLs

```{r}
library(UpSetR)
data_results <- readRDS(here(base_path, '2_11_mqtl_results_data.rds'))
results <- readRDS(here(base_path, '2_11_mqtl_results.rds'))
pcs <- readRDS(here(base_path, '2_11_pcs_geno.rds'))
prop_var <- pcs$variance
pcs_eth <- pcs$pcs

locus_summary_plot_data <- readRDS(here(base_path, '2_11_locus_summary_plot_data.rds'))
```

## number of mqtls

```{r}
p4 <- results %>%
  group_by(tissue) %>%
  
  # how many are significant
  summarize(`FDR < 0.05` = sum(fdr < 0.05)) %>%
  
  # reshape for plot
  pivot_longer(cols = -tissue,
               names_to = 'Significance',
               values_to = 'n') %>%
  
  mutate(tissue = fct_rev(tissue)) %>%
  ggplot(aes(x = tissue, y = n)) +
  geom_bar(stat = 'identity', color = 'white') +
  geom_text(aes(label = n), nudge_y = 200, size = 2.5) +
  ggrepel::geom_text_repel(data = data.frame(tissue = "Endothelial cs",
                                             n = 2956,
                                             label = 'Out of a total of\n2956 tested mQTLs'), 
                           nudge_y = -650,
                           aes(label = label),
                           size = 2.75) +
  geom_hline(yintercept = 2956, linetype = 'dashed') +
  theme_bw(base_size = fontsize) +
  theme(panel.border = element_blank(),
        axis.line = element_line(),
        axis.ticks = element_blank(),
        axis.title.y = element_text(angle = 0, vjust = 0.5),
        panel.grid = element_blank(),
        strip.background = element_blank(),
        strip.text = element_text(hjust = 0, face = 'bold')) +
  scale_y_continuous(expand = c(0,0), limits = c(0, 3000)) +
  scale_x_discrete(labels = function(x)gsub(' cs', '', x)) +
  labs(x = '', y= 'Number of significant (FDR < 0.05)\nCpG-SNP associations') + 
  coord_flip();p4 #h: 2.5, w: 3
```

## Overlap (upset)

```{r}
results_dummy <- results %>%
  mutate(significant = as.numeric(fdr < 0.05),
         tissue = gsub(' cs', '', tissue)) %>%
  select(cpg, tissue, snpid, significant) %>%
  pivot_wider(id_cols = c(cpg, snpid),
              names_from = 'tissue',
              values_from = 'significant') %>% as.data.frame

p5 <- upset(results_dummy, 
            point.size = 3.25,
            line.size = 0.9,
            text.scale = c(1.25, 1, 1, 1, 1.25, 1.5)*0.85,
            mainbar.y.label = 'number of intersecting mQTLs', 
            sets.x.label = 'mQTLs per cell type',
            mb.ratio = c(0.55, 0.45),
            intersections = list(
              
              list('Villi'),
              list('Trophoblasts'),
              list('Hofbauer'),
              list('Endothelial'),
              list('Stromal'),
              
              list('Villi', 'Trophoblasts'),
              list('Villi', 'Hofbauer'),
              list('Villi', 'Endothelial'),
              list('Villi', 'Stromal'),
              
              list('Villi', 'Stromal', 'Endothelial', 'Hofbauer', 'Trophoblasts')
            ),
            queries = list(
              list(query = intersects,
                   params = list('Villi'),
                   active = TRUE),
              list(query = intersects,
                   params = list('Villi', 'Trophoblasts'),
                   active = TRUE),
              list(query = intersects,
                   params = list('Villi', 'Stromal', 'Endothelial', 'Hofbauer', 'Trophoblasts'),
                   active = TRUE)
            ));p5 #h: 2.5, w : 3.75 #edit 
```

## example mQTLs

```{r}
p6 <- data_results %>%
  filter(cpg %in% c('cg00167916', 'cg01643090', 'cg11966998', 'cg11861562', 'cg11030744'),
         SNPID %in% c('rs6938690', 'kgp11830586', 'rs11023685', 'rs8521', 'rs35682'))%>%
  
  mutate(Tissue = fct_rev(Tissue),
         gene_name = fct_relevel(gene_name, 
                                 c('TAP2', 'CAMTA1', 'TAGLN', 'SOX6', 'GHRL')),
         cpg_snp = fct_reorder(cpg_snp, gene_name, function(x)mean(as.numeric(x))),
         
         Tissue_color = case_when(
           Villi == 1 & Tissue == 'Villi' ~ 'Villi',
           Trophoblasts == 1 & Tissue == 'Trophoblasts cs' ~ 'Trophoblasts cs',
           Endothelial == 1 & Tissue == 'Endothelial cs' ~ 'Endothelial cs',
           Stromal == 1 & Tissue == 'Stromal cs' ~ 'Stromal cs',
           Hofbauer == 1 & Tissue == 'Hofbauer cs' ~ 'Hofbauer cs',
           TRUE ~ NA_character_
         )) %>%
  
 ggplot(aes(x = Tissue, y = betas, 
            color = Tissue_color, 
            fill = Tissue_color)) +
  geom_boxplot(position = position_dodge(width = 0.75), 
               aes(alpha = genotype),
               #color = '#000000',
               outlier.shape = NA,
               lwd = 0.1,
               fatten = 0.75) +
  #geom_jitter(size = 0.25, 
  #            alpha = 0.75,
  #            aes(group = genotype),
  #            position = position_jitterdodge(dodge.width = 0.75,
  #                                            jitter.width = 0.1)) +
  theme_bw(base_size = fontsize) +
  theme(axis.ticks = element_blank(),
        panel.grid.major.x = element_blank(),
        panel.grid.minor.y = element_blank(),
        strip.background = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
  labs(alpha = '', y = 'CpG DNA methylation', x = '') +
  facet_wrap(vars(cpg_snp, gene_name), ncol = 5) +
  
  scale_color_manual(values = color_code_tissue[unique(data_results$Tissue)], 
                     na.value = '#424242') +
  scale_fill_manual(values = color_code_tissue[unique(data_results$Tissue)],
                    na.value = '#424242') +
  scale_alpha_manual(values = c(0.2, 0.5, 0.8), guide = 'legend',
                     labels = c('AA', 'AB', 'BB')) +
  scale_y_continuous(expand = c(0,0), 
                     limits = c(0,1), 
                     labels = function(x)percent(x, accuracy = 1)) +
  scale_x_discrete(labels = function(x)gsub(' cs', '', x)) +
  
  # manually create legend for alpha
  guides(alpha = guide_legend(override.aes = list(fill = '#636363',
                                                  alpha = c(0.2, 0.5, 1),
                                                  color = 'black')),
         color = 'none',
         fill = 'none');p6 #h: 2.5, w: 6.5

```

# Ancestry

```{r}
p7 <- prop_var %>%
  ggplot(aes(x = PC, y = p_variance)) +
  geom_bar(stat = 'identity') +
  scale_x_continuous(expand = c(0,0),
                     limits = c(0,21),
                     breaks = 1:20) +
  theme(panel.grid.major.x = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        axis.ticks = element_blank(),
        axis.line = element_line()) +
  scale_y_continuous(expand = c(0,0), labels = scales::percent) +
  labs(y = '', title = 'Percent variance explained');p7 #h:2.5, w:6

p8 <- pcs_eth %>%
  ggplot(aes(x = PC1, y = PC2, color = Predicted_ethnicity)) +
  geom_point(size = 1, alpha = 0.75) +
  theme_classic(base_size = fontsize) +
  theme(legend.spacing.y = unit(0, 'cm'),
        legend.key.size = unit(0, 'cm'),
        legend.position = 'none')+
  guides(color = guide_legend(override.aes = list(size = 3))) +
  scale_color_brewer(breaks = c('Ambiguous', 'Asian', 'Caucasian'), type = 'qual',
                     palette = 2,
                     na.value = 'grey') +
  labs(x = paste0('PC1 (', prop_var$p_variance[1] %>% scales::percent(accuracy = 0.1),')'),
       y = paste0('PC2 (', prop_var$p_variance[2] %>% scales::percent(accuracy = 0.1),')'),
       color = 'PLANET-predicted\nethnicity');p8 # h:2.5, w: 2

p9 <- pcs_eth %>%
  ggplot(aes(x = PC3, y = PC4, color = Predicted_ethnicity)) +
  geom_point(size = 1, alpha = 0.75) +
  theme_classic(base_size = fontsize) +
  theme(legend.spacing.y = unit(0, 'cm'),
        legend.key.size = unit(0, 'cm'),
        legend.position = 'none')+
  scale_color_brewer(breaks = c('Ambiguous', 'Asian', 'Caucasian'), type = 'qual',
                     palette = 2,
                     na.value = 'grey') +
  guides(color = guide_legend(override.aes = list(size = 3))) +
  labs(x = paste0('PC3 (', prop_var$p_variance[3] %>% scales::percent(accuracy = 0.1),')'),
       y = paste0('PC4 (', prop_var$p_variance[4] %>% scales::percent(accuracy = 0.1),')'),
       color = 'PLANET-predicted\nethnicity') +
  coord_equal();p9 # h:2.5, w: 2

p9b <- pcs_eth %>%
  ggplot(aes(x = PC5, y = PC6, color = Predicted_ethnicity)) +
  geom_point(size = 1, alpha = 0.75) +
  theme_classic(base_size = fontsize) +
  theme(legend.spacing.y = unit(0, 'cm'),
        legend.key.size = unit(0, 'cm'))+
  scale_color_brewer(breaks = c('Ambiguous', 'Asian', 'Caucasian'), type = 'qual',
                     palette = 2,
                     na.value = 'grey') +
  guides(color = guide_legend(override.aes = list(size = 3))) +
  labs(x = paste0('PC5 (', prop_var$p_variance[5] %>% scales::percent(accuracy = 0.1),')'),
       y = paste0('PC6 (', prop_var$p_variance[6] %>% scales::percent(accuracy = 0.1),')'),
       color = 'PLANET-predicted\nethnicity') +
  coord_equal();p9b# h:2.5, w: 2
```

# filtering / processing

```{r}
p10 <- locus_summary_plot_data %>%
  ggplot(aes(x = step, y = n)) +
  geom_bar(stat = 'identity') +
  geom_text(aes(label = scales::number(n)), nudge_y = 150000,
            size = 2.5) +
  theme(panel.border = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.grid.minor = element_blank(),
        axis.ticks = element_blank()) +
  scale_y_continuous(labels = scales::number, breaks = c(0, 500000, 1000000)) +
  labs(x = '', title = 'Number of SNPs removed', y = '') +
  coord_flip();p10 #h:1.5, w:2.5

# calculate MAF
p11 <- data_results %>%
  select(Case_ID, SNPID, genotype) %>%
  distinct() %>%
  mutate(genotype = fct_explicit_na(genotype, 'No call')) %>%
  
  # number of b alleles
  group_by(SNPID, genotype) %>%
  summarize(n = n()) %>% 
  mutate(n_b = (as.numeric(genotype)-1)*n) %>%
  group_by(SNPID) %>%
  
  # BAF
  mutate(BAF = sum(n_b) / (sum(n)*2),
         MAF = if_else(BAF > 0.5, 1-BAF, BAF)) %>%
  
  # reorder SNP names based on BAF
  ungroup() %>%
  mutate(SNPID = fct_reorder(SNPID, BAF)) %>%
  
  ggplot(aes(x = SNPID, y = n, fill = genotype, color = genotype)) +
  geom_bar(stat = 'identity') +
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.y = element_text(angle = 0, vjust = 0.5),
        legend.key.size = unit(0.5, 'cm')) +
  scale_fill_manual(values = c(viridis(3), 'grey')) +
  scale_color_manual(values = c(viridis(3), 'grey'), guide = 'none') +
  scale_y_continuous(expand = c(0,0)) +
  labs(y = 'Count', x = 'SNP', fill = '# B alleles') ;p11 #h: 1.5, w: 3.5
```