---
title: "Placental Cell DNA Methylation"
author: "Victor Yuan"
output:
    html_document:
    keep_md: no
    self_contained: yes
    theme: spacelab
    toc: yes
    toc_depth: 3
    toc_float:
      collapsed: no
editor_options:
  chunk_output_type: console
---

```{r, echo = F, message = F, warning = F}
library(dplyr)
library(plyr)
library(kableExtra)
```

# README

For NIH EPIC cells project

R scripts and folders are named in the following system:

* Directories are numbered by the order their contained scripts in the workflow. For example, scripts
in `0_data` are ran before `1_qc_preprocess`, which is before `2_analysis.`
* Scripts are numbered by their directory and workflow order as well. `0_1_load_data.Rmd` is in the 
`0_data` directory and is ran before `0_2_check_data.Rmd.`
* Output from scripts is stored in either `main/interim` or `main/processed`. Output is prefixed by the
same prefix of the script used to generate them. For example, `0_1_load_data.Rmd` produces 
`0_1_rgset.rds`, which is stored in `main/interim`.

# Directory structure

```
Placental Cell DNA Methylation
|   README.html
│   README.md                   
|   README.Rmd 
│   
└───data
│   └───external                  <- Data from third party sources (e.g. annotations)
│   └───main
|           └───raw               <- The original, immutable data (e.g. DNAme, pData)          
|           └───interim           <- Intermediate data that has been transformed
│           └───processed         <- The final, canonical data sets for analysis/sharing
|
└───R
|   └───0_data                   <- Downloading & loading data into R
|   └───1_qc_preprocess          <- Quality control and preprocessing data
|   └───2_analysis               <- Analyzing processed data
|   └───functions                 
|
└───outs                          <- Figures, tables, reports
```

# data

## data/main/raw/

```{r, echo = F}
main_raw <- data.frame(
  File = list.files('./data/main/raw/'),
  Description = c('final sample sheet created after week 2',
                  'sample sheet for week 1 samples (not used)')
)
kable(main_raw)  %>%
  kable_styling(c("striped", "bordered"))
```

## data/main/interim/

```{r, echo = F}

main_interim <- data.frame(
  File = list.files('./data/main/interim/'),
  Description = c('IDATs parsed/loaded using ewastools',
                  'IDATs parsed/loaded using minfi',
                  'Control probes data parsed/loaded using ewastools',
                  'Correlation matrix on snp betas',
                  'UMAP and TSNE coordinates computed on raw betas', #5
                  'PCA association plot data',
                  'Phenodata with QC info',
                  'Probe QC information',
                  'SNP probe data',
                  'Called genotypes on SNP data using ewastools', #10
                  'Custom Chromosome lengths',
                  'Conumee determined binned probes for CNV calling',
                  'Probes output from conumee::CNV.write',
                  'Segments output from conumee::CNV.write',
                  'pData with maternal contamination info', # 15
                  'Noob-normalized filtered autosomal probes',
                  'Noob-normalized mset object'),
  Object = c('list',
             'rgchannelset',
             'dataframe',
             'matrix',
             'dataframe', #5
             'dataframe',
             'dataframe',
             'dataframe',
             'matrix',
             'list', # 10
             'dataframe?',
             'dataframe',
             'dataframe',
             'dataframe',
             'dataframe', # 15
             'matrix',
             'methylset') 
)

kable(main_interim)  %>%
  kable_styling(c("striped", "bordered"))
  
```

# Analysis

# References


