---
title: "2_16_wendy_genes"
author: "Victor Yuan"
date: "20/02/2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

for producing plots for wendy

# Setup

## Libraries

```{r, message = FALSE, warning = FALSE}
# libraries and data
library(tidyverse)
library(scales)
library(here)
library(readxl)
library(janitor)
theme_set(theme_bw())
```

## Data

```{r eval = TRUE}
base_path <- file.path('data', 'main', 'interim')

# pData
pDat <- readRDS(here(base_path, '2_3_pDat_contam.rds'))
pDat <- pDat %>%
  mutate(Tissue = case_when(
    !(Tissue %in% c('Villi', 'Villi maternal', 'Syncytiotrophoblast')) ~ paste(Tissue, 'cs'),
    Tissue == 'Syncytiotrophoblast' ~ 'Trophoblasts enz',
    TRUE ~ Tissue
  )) 

# raw methylation data
betas <- readRDS(here(base_path, '1_4_betas_noob_filt_XYkeep.rds'))

# annotation
anno <- readRDS('Z:/Victor/Repositories/EPIC_annotation/hg19_epic_annotation.rds')
anno <- anno %>%
  as_tibble() %>%
  filter(cpg %in% rownames(betas)) # filter to filtered betas cpgs

# DMCs
dmcs <- readRDS(here(base_path, '2_4_dmcs.rds'))

# color key
color_code <- readRDS(here(base_path, '2_3_color_code.rds'))
color_code_tissue <- setNames(color_code$Colors_Tissue, color_code$label)
```

## Remove samples

```{r}
pDat_filt <- pDat %>% 
  filter(maternal_contamination_norm_flip < 0.35,
         !Sample_Name %in% c('PM364_hofb_cs', 'PL293_v_R2', 'PM366_vc_R2', 'P131_hofb_cs', 
                             'PM324_V4', 'PM324_V1', 'PM139_vc', 'PM77_vc'),
         !Tissue %in% c('Villi maternal', 'Trophoblasts enz', 'Mixture cs', 
                        'Dead Cells and Lymphocytes cs'),
         Trimester != 'Second')

betas_filt <- betas[,pDat_filt$Sentrix]
colnames(betas_filt) <- pDat_filt$Sample_Name
```


## define function 

```{r}
# function
annoPlot_with_tracks <- function(
  cpg_name = NULL, gene_name = NULL, Chr = NULL, Start = NULL, End = NULL, 
  cpg_number = 50, first_cpg = NULL, end_cpg = NULL, 
  Tissue_type = unique(pDat_filt$Tissue),
  font_size_scale = 0.75){
  
find_cpg_from_region <- function(Chr, Start, End){
  
  cpg_name <- anno %>% 
    filter(chr == Chr, start > Start, end < End) %>%
    pull(cpg) %>%
    unique()
  
  cpg_name <- cpg_name[cpg_name %in% rownames(betas)]
  
  cpg_name
}

######## FILTER ANNOTATION TO RELEVANT CPGS, DEPENDING ON INPUT #################
  if (!is.null(gene_name) & is.null(first_cpg) & is.null(end_cpg)){

    # If a gene symbol is supplied,
    # filter to gene in annotation
    anno_gene <- anno %>%
      filter(grepl(paste0('(?<!-)\\b', gene_name, '\\b(?!-)'), genes_symbol, perl = TRUE),
             cpg %in% rownames(betas)) %>%
      arrange(desc(start)) %>% slice(1:cpg_number)
    
    # stop if gene symbol is invalid
    if (nrow(anno_gene) == 0) {
      stop('No cpgs found, maybe they were filtered out.')
    }
  }  else if (!is.null(gene_name) & !is.null(first_cpg) & !is.null(end_cpg)){

    # If a gene symbol is supplied,
    # filter to gene in annotation
    anno_gene <- anno %>%
      filter(grepl(paste0('(?<!-)\\b', gene_name, '\\b(?!-)'), genes_symbol, perl = TRUE),
             cpg %in% rownames(betas)) %>%
      arrange(desc(start)) %>% slice(first_cpg: end_cpg)
    
    # stop if gene symbol is invalid
    if (nrow(anno_gene) == 0) {
      stop('No cpgs found, maybe they were filtered out.')
    }
  }  else if (!is.null(cpg_name)){
      
      # If a cpg is provided,
      # filter annotation to cpgs
      anno_gene <- anno %>%
        filter(cpg %in% cpg_name,
               cpg %in% rownames(betas)) %>% 
        arrange(chr, desc(start))
      
      # stop if no cpgs were found
      if (nrow(anno_gene) == 0) {
        stop('None of the selected CpGs exist in our processed data.')
      }
      
      if (length(cpg_name) < length(intersect(cpg_name, rownames(betas)))) {
        print(paste0('The following selected CpGs do not exist in our processed data:\n', 
                     setdiff(cpg_name, rownames(betas))))
      }
      
  }      else if (!is.null(Chr) & !is.null(Start) & !is.null(End) & is.null(first_cpg) & is.null(end_cpg)){
        
        #get cpg sites within a given region
        cpg_name <- find_cpg_from_region(Chr, Start, End)
        # filter annotation to cpgs
        anno_gene <- anno %>%
        filter(cpg %in% cpg_name,
               cpg %in% rownames(betas)) %>% 
        arrange(chr, desc(start)) %>% slice(1:cpg_number)
        
      # stop if no cpgs were found
      if (nrow(anno_gene) == 0) {
      stop('No cpgs found in the selected region')
    }
      }     else if (!is.null(Chr) & !is.null(Start) & !is.null(End) & !is.null(first_cpg) & !is.null(end_cpg)){
        
        #get cpg sites within a given region
        cpg_name <- find_cpg_from_region(Chr, Start, End)
        # filter annotation to cpgs
        anno_gene <- anno %>%
        filter(cpg %in% cpg_name,
               cpg %in% rownames(betas)) %>% 
        arrange(chr, desc(start)) %>% slice(first_cpg : end_cpg)
      # stop if no cpgs were found
      if (nrow(anno_gene) == 0) {
      stop('No cpgs found in the selected region')
    }
      }    else {
      stop('Enter valid cpg(s) OR one specific gene.')
  }
  
  ######## SETUP INPUT TO PLOTS ###############
  #
  # Title
  #
  title <- paste0(
    unique(anno_gene$chr), ':',
    min(anno_gene$start), '-',
    max(anno_gene$start)
  )
  
  if (is.null(cpg_name)){
    title <- paste0(gene_name, ', ', title)
  } 
  
  
  #
  # now we wrangle the betas/annotations/pdata information into a format usable for plotting
  #
    betas_gene <- t(betas_in[anno_gene$cpg,,drop = FALSE]*100) %>% as_tibble %>% 
    mutate(Sample_Name = colnames(betas_in)) %>%
    
    # reshape into longer format
    pivot_longer(cols = -Sample_Name,
                 names_to = 'cpg', 
                 values_to = 'beta')  %>%
    
    # add tissue and trimester info
    left_join(pDat_filt %>% select(Sample_Name, Trimester, Tissue, Sex), by = 'Sample_Name') %>%
    
    # calculate mean and sd for each cpg for each group
    group_by(Tissue, Trimester, cpg, Sex) %>%
    summarize(mean = mean(beta),
              sd = sd(beta)) %>%
    mutate(lower = mean-sd, upper = mean+sd) %>%
    
    # add cpg info
    left_join(anno_gene, by = 'cpg') %>%
    
    # order on cpg position
    ungroup() %>%
    arrange(desc(start)) %>%
    mutate(cpg = factor(cpg, levels = unique(cpg)),
           Trimester_Tissue = paste0(Trimester, ' - ', Tissue)) %>%
    filter(Tissue %in% Tissue_type)
  
  ##### GENERATE INDIVIDUAL PLOT TRACKS #####
  font_size <- 12*font_size_scale
  point_size <- 2
  line_size <- 0.9
  
  # Generate  methylation track
  p1 <- betas_gene %>%
    mutate(cpg_num = as.numeric(cpg)) %>%
    ggplot() +
    
    # geom_ribbon is to generate the alternating shaded background
    #geom_ribbon(aes(x = cpg_num, ymin = 0, ymax = 25),
    #            fill = 'grey', alpha = 0.15)+
    #geom_ribbon(aes(x = cpg_num, ymin = 50, ymax = 75),
    #            fill = 'grey', alpha = 0.15)+
    
    # geom_linerange and geom_point is for the actual methylation data
    #geom_linerange(alpha = 0.5, size = line_size, 
    #               aes(x = cpg_num, ymin =lower, ymax = upper, color = Tissue),
    #               show.legend = FALSE) +
    geom_point(alpha = 0.75, aes(x = cpg_num, y = mean, color = Tissue), size = point_size) +
    geom_line(aes(x = cpg_num, y = mean, color = Tissue, linetype = Sex), 
              alpha = 0.75, size = line_size)+
    
    #geom_ribbon(alpha = 0.5, aes(x = cpg_num, ymin = lower, ymax = upper, fill = Tissue))+
    
    # we want to facet by trimester
    facet_wrap(vars(Trimester), ncol = 1, labeller = label_both) +
    
    # cosmetics
    theme_bw(base_size = font_size) +
    theme(axis.title.x = element_blank(),
          axis.text.x = element_blank(), 
          axis.text.y = element_text(size = font_size*0.9),
          axis.ticks.x = element_blank(),
          axis.title.y = element_text(angle = 0, vjust = 0.5, hjust = 0, size = font_size),
          
          legend.text = element_text(size = font_size*0.9),
          
          panel.grid = element_blank(),
          panel.spacing.y = unit(0.25, 'cm'),
          
          strip.background = element_blank(),
          strip.text = element_text(face = 'bold', hjust = 0),
          
          plot.margin=margin(l=-0.1, unit="cm")) +
    scale_y_continuous(limits = c(0, 100), 
                       expand = c(0, 0), 
                       labels = function(x)paste0(x, '%')) +
    scale_x_continuous() +
    scale_color_manual(values= color_code_tissue[unique(betas_gene$Tissue)],
                       guide = guide_legend(override.aes = list(size = 4)),
                       labels = function(x)gsub(' cs', '', x)) +
    labs(y = 'DNA\nmethylation', 
         x = '', 

         color = '', 
         title = title)

  # plot cpg annotations
  p2 <- betas_gene %>% 
    
    # First we need to process the annotation data
    mutate(# absent/presence for different genomic elements
           enhancer = !is.na(enhancers_id),
           pmd = !is.na(pmd_id),
           imprint_tissue_specificity = !is.na(imprint_tissue_specificity)) %>%
    mutate_if(is.logical, as.character) %>%
    select(cpg, enhancer, pmd, 
           imprint_tissue_specificity, cpg_id, start) %>%
    
    # reshape
    pivot_longer(cols = c(-cpg, -start),
               names_to = 'cpg_element', 
               values_to = 'presence') %>%
    distinct() %>%
    
    # arrange plot orde for each track
    arrange(desc(start)) %>%
    mutate(cpg = factor(cpg, levels = unique(cpg)),
           cpg_element = factor(cpg_element, 
                                levels = c('enhancer', 'pmd', 'imprint_tissue_specificity', 'cpg_id')),
           presence = factor(ifelse(presence == "TRUE", 'Present',
                                    ifelse(presence == "FALSE", 'Absent', presence)),
                             levels = c('sea', 'shelf', 'shore', 'island',
                                        'Present', 'Absent')),
           cpg_num = as.numeric(cpg),
           
           # for facet label
           group = 'Annotations') %>%
    
    # plot code
    ggplot(aes(x = cpg_num, y = cpg_element, fill = presence)) +
    geom_tile(color = 'white') +
    guides(fill=guide_legend(ncol=3, override.aes = list(colour = 'white'))) +
    facet_wrap(vars(group), ncol = 2) +
    theme_bw(base_size = font_size) +
    theme(axis.title.x = element_blank(),
          axis.text.x = element_blank(),
          axis.ticks = element_blank(),
          axis.text.y = element_text(vjust = 0.5),
          
          legend.text = element_text(size = font_size*0.9),
          
          plot.margin = margin(l=-0.1,unit="cm"),
          
          strip.background = element_blank(),
          strip.text = element_text(face = 'bold', hjust = 0)) +
    scale_fill_manual(values = c('Present' = '#cccccc', 'Absent' = '#f7f7f7', 
                                 'sea' = '#ffffcc', 'shelf' = '#a1dab4', 
                                 'shore' = '#41b6c4', 'island' = '#225ea8'), drop = F,
                      labels = stringr::str_to_title) +
    scale_x_continuous(expand = c(0,0)) +
    scale_y_discrete(expand = c(0,0),
                     labels = function(x)gsub('cpg_id', 'Relation to CpG Islands',
                       gsub('imprint_tissue_specificity', 'Imprinted DMR',
                       gsub('pmd', 'Placental PMD',
                       gsub('enhancer', 'Enhancer', 
                            x))))) +
    labs(x = '', y = '', fill = '')
  
  # transcript annotations
  
  # cpgs can have more than one mapping to a gene element per transcript
  # cpgs mapped to intronexonboundaries can also be part of exons and introns
  # 5 UTR and 3 UTR can also overlap with exons and introns
  
  # If there are multiple mappings then only the following will be displayed in order of priority:
  # 5UTR / 3UTR, exon / intron, intronexonboundary
  
  p3 <- betas_gene %>% 
    select(cpg, end, genes_id, genes_tx_id, genes_symbol) %>%
    distinct() %>%
    
    # split comma separate list and put each entry into a new row
    mutate(genes_tx_id = str_split(genes_tx_id, ', '),
           genes_id = str_split(genes_id, ', '),
           genes_symbol = str_split(genes_symbol, ', ')) %>%
    unnest(c(genes_tx_id, genes_id, genes_symbol)) %>%
    
    # paste together transcript ID and genes symbol
    mutate(genes_tx_id_symbol = if_else(genes_symbol != 'no_associated_gene', 
                                        paste0(genes_tx_id, ' (', genes_symbol, ')'),
                                        genes_tx_id)) %>%
    
    ### ORDER TRANSCRIPTS IN PLOT 
    #
    # note that the first level appears on the bottom of the graph (y = 0 if it was continuous)
    
    # calculate earliest end of each symbol
    group_by(genes_symbol) %>%
    mutate(genes_symbol_end = min(end)) %>%
    ungroup() %>%
    
    # reorder gene symbols by their earliest end position
    mutate(genes_symbol = fct_reorder(genes_symbol, genes_symbol_end)) %>%
    
    # reorder transcripts based on gene symbol levels and then by their end position
    arrange(genes_symbol, end) %>%
    mutate(genes_tx_id_symbol = 
             factor(genes_tx_id_symbol, levels = unique(genes_tx_id_symbol)) %>%
             fct_explicit_na()) %>%
    
    
    ####
    # fill all combinations of cpg and gene_tx_id with NA if missing, so that they show in plot
    complete(cpg, genes_tx_id_symbol) %>%
    
    # filter out to unique mappings using the priority described above
    # strategy is to order by factor level and take the first occurence
    mutate(genes_id = factor(genes_id, 
                             levels = c('1to5kb', 'promoter', '5UTR', '3UTR', 'exon',
                                        'intron', 'intronexonboundary', 'intergenic'))) %>%
    
    group_by(cpg, genes_tx_id_symbol) %>%
    arrange(genes_id)  %>%
    dplyr::slice(1) %>%
    
    # reorder factor levels for appearance in plot legend
    mutate(genes_id = factor(genes_id, 
                             levels = c('1to5kb', 'promoter', '5UTR', 'exon', 'intron',
                                        'intronexonboundary', '3UTR', 'intergenic')),
           group = 'Transcripts') %>%
    
    #plot
    ggplot(aes(x = cpg, y = genes_tx_id_symbol, fill = genes_id)) +
    geom_tile(color = 'white', alpha = 0.6) +
    facet_wrap(vars(group), ncol = 1) +
    guides(fill=guide_legend(ncol=3, override.aes = list(alpha=0.7, color = 'white'))) +
    theme_bw(base_size = font_size) +
    theme(axis.title.x = element_blank(),
          axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), 
          axis.text.y = element_text(vjust = 0.5),
          axis.ticks = element_blank(),
          
          plot.margin = margin(l=-0.1,unit="cm"),
          
          legend.text = element_text(size = font_size*0.9),
          
          panel.grid = element_blank(),
          
          strip.background = element_blank(),
          strip.text = element_text(face = 'bold', hjust = 0)) +
    scale_fill_brewer(na.value = '#f7f7f7', breaks = c('1to5kb', 'promoter', '5UTR', 'exon', 
                                                        'intron', 'intronexonboundary', '3UTR',
                                                        'intergenic'),
                      palette = 'Paired', direction = -1) +
    scale_y_discrete(expand = c(0,0), 
                     labels = function(x)(if_else(x == '(Missing)', '', x))) +
    scale_x_discrete(expand = c(0,0)) +
    coord_cartesian(clip = "off") +
    labs(x = '', y = '', fill = '')
  
  # the size of each plot depends on the number of unique transcripts
  # More transcripts means a longer transcript track
  # so the plot height depends on the number of transcripts
  
  # number of transcripts
  n_trans <- betas_gene$genes_tx_id %>% str_split(', ') %>% unlist() %>% unique %>% length()
  
  # size of p1
  p1_h <- 40
  
  # combine tracks plot
  egg::ggarrange(p1, p2, p3, ncol = 1, heights = c(p1_h-(n_trans-6), 6, n_trans))
}
```

## plot

```{r}
#troph and villi
betas_in <- betas_filt[,pDat_filt %>% 
                         filter(Tissue %in% c('Villi', 'Trophoblasts cs')) %>% 
                         pull(Sample_Name)]
p <- annoPlot_with_tracks(gene_name = 'OGT');p
p <- annoPlot_with_tracks(gene_name = 'ZNF300');p
```

## FACET BY CELL TYPE

```{R}
# function
annoPlot_with_tracks <- function(
  cpg_name = NULL, gene_name = NULL, Chr = NULL, Start = NULL, End = NULL, 
  cpg_number = 50, first_cpg = NULL, end_cpg = NULL, 
  Tissue_type = unique(pDat_filt$Tissue),
  font_size_scale = 0.75){
  
find_cpg_from_region <- function(Chr, Start, End){
  
  cpg_name <- anno %>% 
    filter(chr == Chr, start > Start, end < End) %>%
    pull(cpg) %>%
    unique()
  
  cpg_name <- cpg_name[cpg_name %in% rownames(betas)]
  
  cpg_name
}

######## FILTER ANNOTATION TO RELEVANT CPGS, DEPENDING ON INPUT #################
  if (!is.null(gene_name) & is.null(first_cpg) & is.null(end_cpg)){

    # If a gene symbol is supplied,
    # filter to gene in annotation
    anno_gene <- anno %>%
      filter(grepl(paste0('(?<!-)\\b', gene_name, '\\b(?!-)'), genes_symbol, perl = TRUE),
             cpg %in% rownames(betas)) %>%
      arrange(desc(start)) %>% slice(1:cpg_number)
    
    # stop if gene symbol is invalid
    if (nrow(anno_gene) == 0) {
      stop('No cpgs found, maybe they were filtered out.')
    }
  }  else if (!is.null(gene_name) & !is.null(first_cpg) & !is.null(end_cpg)){

    # If a gene symbol is supplied,
    # filter to gene in annotation
    anno_gene <- anno %>%
      filter(grepl(paste0('(?<!-)\\b', gene_name, '\\b(?!-)'), genes_symbol, perl = TRUE),
             cpg %in% rownames(betas)) %>%
      arrange(desc(start)) %>% slice(first_cpg: end_cpg)
    
    # stop if gene symbol is invalid
    if (nrow(anno_gene) == 0) {
      stop('No cpgs found, maybe they were filtered out.')
    }
  }  else if (!is.null(cpg_name)){
      
      # If a cpg is provided,
      # filter annotation to cpgs
      anno_gene <- anno %>%
        filter(cpg %in% cpg_name,
               cpg %in% rownames(betas)) %>% 
        arrange(chr, desc(start))
      
      # stop if no cpgs were found
      if (nrow(anno_gene) == 0) {
        stop('None of the selected CpGs exist in our processed data.')
      }
      
      if (length(cpg_name) < length(intersect(cpg_name, rownames(betas)))) {
        print(paste0('The following selected CpGs do not exist in our processed data:\n', 
                     setdiff(cpg_name, rownames(betas))))
      }
      
  }      else if (!is.null(Chr) & !is.null(Start) & !is.null(End) & is.null(first_cpg) & is.null(end_cpg)){
        
        #get cpg sites within a given region
        cpg_name <- find_cpg_from_region(Chr, Start, End)
        # filter annotation to cpgs
        anno_gene <- anno %>%
        filter(cpg %in% cpg_name,
               cpg %in% rownames(betas)) %>% 
        arrange(chr, desc(start)) %>% slice(1:cpg_number)
        
      # stop if no cpgs were found
      if (nrow(anno_gene) == 0) {
      stop('No cpgs found in the selected region')
    }
      }     else if (!is.null(Chr) & !is.null(Start) & !is.null(End) & !is.null(first_cpg) & !is.null(end_cpg)){
        
        #get cpg sites within a given region
        cpg_name <- find_cpg_from_region(Chr, Start, End)
        # filter annotation to cpgs
        anno_gene <- anno %>%
        filter(cpg %in% cpg_name,
               cpg %in% rownames(betas)) %>% 
        arrange(chr, desc(start)) %>% slice(first_cpg : end_cpg)
      # stop if no cpgs were found
      if (nrow(anno_gene) == 0) {
      stop('No cpgs found in the selected region')
    }
      }    else {
      stop('Enter valid cpg(s) OR one specific gene.')
  }
  
  ######## SETUP INPUT TO PLOTS ###############
  #
  # Title
  #
  title <- paste0(
    unique(anno_gene$chr), ':',
    min(anno_gene$start), '-',
    max(anno_gene$start)
  )
  
  if (is.null(cpg_name)){
    title <- paste0(gene_name, ', ', title)
  } 
  
  
  #
  # now we wrangle the betas/annotations/pdata information into a format usable for plotting
  #
    betas_gene <- t(betas_in[anno_gene$cpg,,drop = FALSE]*100) %>% as_tibble %>% 
    mutate(Sample_Name = colnames(betas_in)) %>%
    
    # reshape into longer format
    pivot_longer(cols = -Sample_Name,
                 names_to = 'cpg', 
                 values_to = 'beta')  %>%
    
    # add tissue and trimester info
    left_join(pDat_filt %>% select(Sample_Name, Trimester, Tissue, Sex), by = 'Sample_Name') %>%
    
    # calculate mean and sd for each cpg for each group
    group_by(Tissue, Trimester, cpg, Sex) %>%
    summarize(mean = mean(beta),
              sd = sd(beta)) %>%
    mutate(lower = mean-sd, upper = mean+sd) %>%
    
    # add cpg info
    left_join(anno_gene, by = 'cpg') %>%
    
    # order on cpg position
    ungroup() %>%
    arrange(desc(start)) %>%
    mutate(cpg = factor(cpg, levels = unique(cpg)),
           Trimester_Tissue = paste0(Trimester, ' - ', Tissue)) %>%
    filter(Tissue %in% Tissue_type)
  
  ##### GENERATE INDIVIDUAL PLOT TRACKS #####
  font_size <- 12*font_size_scale
  point_size <- 2
  line_size <- 0.9
  
  # Generate  methylation track
  p1 <- betas_gene %>%
    mutate(cpg_num = as.numeric(cpg)) %>%
    ggplot() +
    
    # geom_ribbon is to generate the alternating shaded background
    #geom_ribbon(aes(x = cpg_num, ymin = 0, ymax = 25),
    #            fill = 'grey', alpha = 0.15)+
    #geom_ribbon(aes(x = cpg_num, ymin = 50, ymax = 75),
    #            fill = 'grey', alpha = 0.15)+
    
    # geom_linerange and geom_point is for the actual methylation data
    #geom_linerange(alpha = 0.5, size = line_size, 
    #               aes(x = cpg_num, ymin =lower, ymax = upper, color = Tissue),
    #               show.legend = FALSE) +
    geom_point(alpha = 0.75, aes(x = cpg_num, y = mean, color = Tissue), size = point_size) +
    geom_line(aes(x = cpg_num, y = mean, color = Tissue, linetype = Sex), 
              alpha = 0.75, size = line_size)+
    
    #geom_ribbon(alpha = 0.5, aes(x = cpg_num, ymin = lower, ymax = upper, fill = Tissue))+
    
    # we want to facet by trimester
    facet_wrap(~Tissue, ncol = 1, labeller = labeller(Tissue = function(x)gsub(' cs', '', x))) +
    
    # cosmetics
    theme_bw(base_size = font_size) +
    theme(axis.title.x = element_blank(),
          axis.text.x = element_blank(), 
          axis.text.y = element_text(size = font_size*0.9),
          axis.ticks.x = element_blank(),
          axis.title.y = element_text(angle = 0, vjust = 0.5, hjust = 0, size = font_size),
          
          legend.text = element_text(size = font_size*0.9),
          
          panel.grid = element_blank(),
          panel.spacing.y = unit(0.25, 'cm'),
          
          strip.background = element_blank(),
          strip.text = element_text(face = 'bold', hjust = 0),
          
          plot.margin=margin(l=-0.1, unit="cm")) +
    scale_y_continuous(limits = c(0, 100), 
                       expand = c(0, 0), 
                       labels = function(x)paste0(x, '%')) +
    scale_x_continuous() +
    scale_color_manual(values= color_code_tissue[unique(betas_gene$Tissue)],
                       guide = guide_legend(override.aes = list(size = 4)),
                       labels = function(x)gsub(' cs', '', x)) +
    labs(y = 'DNA\nmethylation', 
         x = '', 

         color = '', 
         title = title)

  # plot cpg annotations
  p2 <- betas_gene %>% 
    
    # First we need to process the annotation data
    mutate(# absent/presence for different genomic elements
           enhancer = !is.na(enhancers_id),
           pmd = !is.na(pmd_id),
           imprint_tissue_specificity = !is.na(imprint_tissue_specificity)) %>%
    mutate_if(is.logical, as.character) %>%
    select(cpg, enhancer, pmd, 
           imprint_tissue_specificity, cpg_id, start) %>%
    
    # reshape
    pivot_longer(cols = c(-cpg, -start),
               names_to = 'cpg_element', 
               values_to = 'presence') %>%
    distinct() %>%
    
    # arrange plot orde for each track
    arrange(desc(start)) %>%
    mutate(cpg = factor(cpg, levels = unique(cpg)),
           cpg_element = factor(cpg_element, 
                                levels = c('enhancer', 'pmd', 'imprint_tissue_specificity', 'cpg_id')),
           presence = factor(ifelse(presence == "TRUE", 'Present',
                                    ifelse(presence == "FALSE", 'Absent', presence)),
                             levels = c('sea', 'shelf', 'shore', 'island',
                                        'Present', 'Absent')),
           cpg_num = as.numeric(cpg),
           
           # for facet label
           group = 'Annotations') %>%
    
    # plot code
    ggplot(aes(x = cpg_num, y = cpg_element, fill = presence)) +
    geom_tile(color = 'white') +
    guides(fill=guide_legend(ncol=3, override.aes = list(colour = 'white'))) +
    facet_wrap(vars(group), ncol = 2) +
    theme_bw(base_size = font_size) +
    theme(axis.title.x = element_blank(),
          axis.text.x = element_blank(),
          axis.ticks = element_blank(),
          axis.text.y = element_text(vjust = 0.5),
          
          legend.text = element_text(size = font_size*0.9),
          
          plot.margin = margin(l=-0.1,unit="cm"),
          
          strip.background = element_blank(),
          strip.text = element_text(face = 'bold', hjust = 0)) +
    scale_fill_manual(values = c('Present' = '#cccccc', 'Absent' = '#f7f7f7', 
                                 'sea' = '#ffffcc', 'shelf' = '#a1dab4', 
                                 'shore' = '#41b6c4', 'island' = '#225ea8'), drop = F,
                      labels = stringr::str_to_title) +
    scale_x_continuous(expand = c(0,0)) +
    scale_y_discrete(expand = c(0,0),
                     labels = function(x)gsub('cpg_id', 'Relation to CpG Islands',
                       gsub('imprint_tissue_specificity', 'Imprinted DMR',
                       gsub('pmd', 'Placental PMD',
                       gsub('enhancer', 'Enhancer', 
                            x))))) +
    labs(x = '', y = '', fill = '')
  
  # transcript annotations
  
  # cpgs can have more than one mapping to a gene element per transcript
  # cpgs mapped to intronexonboundaries can also be part of exons and introns
  # 5 UTR and 3 UTR can also overlap with exons and introns
  
  # If there are multiple mappings then only the following will be displayed in order of priority:
  # 5UTR / 3UTR, exon / intron, intronexonboundary
  
  p3 <- betas_gene %>% 
    select(cpg, end, genes_id, genes_tx_id, genes_symbol) %>%
    distinct() %>%
    
    # split comma separate list and put each entry into a new row
    mutate(genes_tx_id = str_split(genes_tx_id, ', '),
           genes_id = str_split(genes_id, ', '),
           genes_symbol = str_split(genes_symbol, ', ')) %>%
    unnest(c(genes_tx_id, genes_id, genes_symbol)) %>%
    
    # paste together transcript ID and genes symbol
    mutate(genes_tx_id_symbol = if_else(genes_symbol != 'no_associated_gene', 
                                        paste0(genes_tx_id, ' (', genes_symbol, ')'),
                                        genes_tx_id)) %>%
    
    ### ORDER TRANSCRIPTS IN PLOT 
    #
    # note that the first level appears on the bottom of the graph (y = 0 if it was continuous)
    
    # calculate earliest end of each symbol
    group_by(genes_symbol) %>%
    mutate(genes_symbol_end = min(end)) %>%
    ungroup() %>%
    
    # reorder gene symbols by their earliest end position
    mutate(genes_symbol = fct_reorder(genes_symbol, genes_symbol_end)) %>%
    
    # reorder transcripts based on gene symbol levels and then by their end position
    arrange(genes_symbol, end) %>%
    mutate(genes_tx_id_symbol = 
             factor(genes_tx_id_symbol, levels = unique(genes_tx_id_symbol)) %>%
             fct_explicit_na()) %>%
    
    
    ####
    # fill all combinations of cpg and gene_tx_id with NA if missing, so that they show in plot
    complete(cpg, genes_tx_id_symbol) %>%
    
    # filter out to unique mappings using the priority described above
    # strategy is to order by factor level and take the first occurence
    mutate(genes_id = factor(genes_id, 
                             levels = c('1to5kb', 'promoter', '5UTR', '3UTR', 'exon',
                                        'intron', 'intronexonboundary', 'intergenic'))) %>%
    
    group_by(cpg, genes_tx_id_symbol) %>%
    arrange(genes_id)  %>%
    dplyr::slice(1) %>%
    
    # reorder factor levels for appearance in plot legend
    mutate(genes_id = factor(genes_id, 
                             levels = c('1to5kb', 'promoter', '5UTR', 'exon', 'intron',
                                        'intronexonboundary', '3UTR', 'intergenic')),
           group = 'Transcripts') %>%
    
    #plot
    ggplot(aes(x = cpg, y = genes_tx_id_symbol, fill = genes_id)) +
    geom_tile(color = 'white', alpha = 0.6) +
    facet_wrap(vars(group), ncol = 1) +
    guides(fill=guide_legend(ncol=3, override.aes = list(alpha=0.7, color = 'white'))) +
    theme_bw(base_size = font_size) +
    theme(axis.title.x = element_blank(),
          axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1), 
          axis.text.y = element_text(vjust = 0.5),
          axis.ticks = element_blank(),
          
          plot.margin = margin(l=-0.1,unit="cm"),
          
          legend.text = element_text(size = font_size*0.9),
          
          panel.grid = element_blank(),
          
          strip.background = element_blank(),
          strip.text = element_text(face = 'bold', hjust = 0)) +
    scale_fill_brewer(na.value = '#f7f7f7', breaks = c('1to5kb', 'promoter', '5UTR', 'exon', 
                                                        'intron', 'intronexonboundary', '3UTR',
                                                        'intergenic'),
                      palette = 'Paired', direction = -1) +
    scale_y_discrete(expand = c(0,0), 
                     labels = function(x)(if_else(x == '(Missing)', '', x))) +
    scale_x_discrete(expand = c(0,0)) +
    coord_cartesian(clip = "off") +
    labs(x = '', y = '', fill = '')
  
  # the size of each plot depends on the number of unique transcripts
  # More transcripts means a longer transcript track
  # so the plot height depends on the number of transcripts
  
  # number of transcripts
  n_trans <- betas_gene$genes_tx_id %>% str_split(', ') %>% unlist() %>% unique %>% length()
  
  # size of p1
  p1_h <- 40
  
  # combine tracks plot
  egg::ggarrange(p1, p2, p3, ncol = 1, heights = c(p1_h-(n_trans-6)+20, 6, n_trans))
}
```

## plot
```{r}
# all cell types, first trimester
betas_in <- betas_filt[,pDat_filt %>% 
                         filter(Trimester == 'First') %>%
                         pull(Sample_Name)]
p <- annoPlot_with_tracks(gene_name = 'OGT');p
p <- annoPlot_with_tracks(gene_name = 'ZNF300')

annoPlot_with_tracks(gene_name = 'DUSP9')

# Third trimester
betas_in <- betas_filt[,pDat_filt %>% 
                         filter(Trimester == 'Third') %>%
                         pull(Sample_Name)]
p <- annoPlot_with_tracks(gene_name = 'OGT')
p <- annoPlot_with_tracks(gene_name = 'ZNF300')

annoPlot_with_tracks(gene_name = 'DUSP9')
```