---
title: "2_3_Firsttrim_DMCs"
author: "Victor Yuan"
date: "17/06/2019"
output:
  html_document:
    keep_md: true
    toc: true
    toc_depth: 4
    toc_float:
      collapsed: false
    theme: spacelab
editor_options: 
  chunk_output_type: console
---

# Setup

## Load libraries and data

```{r message = F, warning = F}
# libraries and data
library(ggplot2)
library(tidyverse)
library(ggrepel)
library(viridis)
library(scales)
library(RColorBrewer)
library(pheatmap)
library(irlba)
library(plomics) # 0.2 github/wvictor14/plomics
library(GGally)
library(cowplot)
library(umap)
library(dendextend)
library(stringr)
library(minfi)
library(limma)
library(biobroom)
library(missMethyl)
library(ggridges)
library(ggpmisc)
library(egg)
```

## Data

Here I call 'syncytiotrophoblasts' into 'Trophoblasts enz' and add 'cs' suffixes to the cell sorted
samples.

```{r}
# pdata
pDat <- readRDS('../../data/main/interim/1_3_pDat.rds')

# filter to first trimester samples
pDat_first <- pDat %>%
  filter(Trimester == 'First', Tissue != 'Dead Cells and Lymphocytes') %>%
  
  # update tissue annotation
  mutate(Tissue = case_when(Tissue != 'Villi' ~ paste(Tissue, 'cs'),
                            Tissue == 'Villi' ~ Tissue))

# raw methylation data
betas <- readRDS('../../data/main/interim/1_4_betas_noob_filt.rds')
mset_noob <- readRDS('../../data/main/interim/1_4_mset_noob.rds')
colnames(mset_noob) <- colnames(betas) <- pDat$Sample_Name
mvals <- getM(mset_noob)

# filter to first trimester
mvals_first <- mvals[rownames(betas),pDat_first$Sample_Name]
betas_first <- betas[,pDat_first$Sample_Name]

# annotation
anno <- getAnnotation(mset_noob)
anno <- anno %>%
  as_tibble() %>%
  filter(Name %in% rownames(betas_first)) # filter to filtered betas cpgs
probe_anno <- readRDS('../../data/main/interim/1_1_probe_anno.rds')

# snp data
snp_betas <- readRDS('../../data/main/interim/1_1_snp_betas.rds')

# color key
color_code <- readRDS('../../data/main/interim/1_1_color_code.rds')
color_code[[1]] <- color_code[[1]] %>% 
  mutate(label = ifelse(Tissue == 'Syncytiotrophoblast', 'Trophoblasts enz', Tissue)) %>%
  mutate(label = ifelse(!label %in% c('Trophoblasts enz', 'Villi', 'Villi maternal'), 
                        paste(label, 'cs'),
                        label),
         Colors_Tissue = ifelse(label == 'Villi', '#721111', Colors_Tissue),
         Colors_Tissue = ifelse(label == 'Trophoblasts enz', '#f4702e', Colors_Tissue))
saveRDS(color_code[[1]], '../../data/main/interim/2_3_color_code.rds')

color_code_tissue <- setNames(color_code[[1]]$Colors_Tissue, color_code[[1]]$label)

```

# Contamination

## Training

These are the levels of contamination that we are dealing with:
```{r}
pDat_first %>% arrange(Tissue, cor_to_reference) %>%
  mutate(Sample_Name = factor(Sample_Name, levels = Sample_Name)) %>%
  ggplot(aes(x = Sample_Name, y = cor_to_reference)) +
  geom_point(shape = 21, aes(fill = Tissue), color = 'grey', size = 3) +
  scale_fill_manual(values = color_code_tissue[unique(pDat_first$Tissue)]) +
  facet_wrap(~Tissue, scales = 'free_y', ncol = 1) + theme_bw() + labs(x = '') + coord_flip()
```

But we can't equate this to % contamination because we don't have a clear reference point.

We can kind of tell based using the normalized Y intensity in males, because it will go to 0 as
they get closer to 100% maternal contamination



```{r}
p_c1 <- ggplot(pDat, aes(x = cor_to_reference, y = normalized_Y_intensity, color = Sex)) +
  geom_point() + theme_bw() + stat_smooth(method = 'lm');p_c1  

ggplot(pDat, aes(x = cor_to_reference, y = normalized_Y_intensity, Shape = Sex, color = Tissue,
                 shape = Trimester)) +
  geom_point() + theme_bw()  + facet_wrap(~Tissue)
```

```{r}
fit_ycor <-lm(normalized_Y_intensity ~ cor_to_reference, pDat %>% filter(Sex == 'M'))
summary(fit_ycor)
```


`normalized_Y_intensity` is our measure of contamination for males.  With the maximum values 
representing samples that are nearly 100% fetal and 0 % contaminated with maternal DNA. The minimum 
value for males represent the most contaminated male samples. However we since we don't know the 
true levels of contamination of these samples, we can't comment if the minimum values in males 
represent 100% contamination / 0% fetal. However, instead we can say that the normalized_Y_intensity
mean for female samples can represent *what a male's normalized Y intensity value would look like if it was 100% contaminated with maternal cells*. 

Therefore, it makes sense to transform / normalize this range to:

1. Maximum -> the average `normalized_Y_intensity` for males when contamination is lowest 
2. Minimum -> the average `normalized_Y_intensity` for females.


For 1. It makes sense to use male samples with a low probability that SNPs are outliers instead of
the correlation to the reference villi. Because `cor_to_reference` assumes villi samples are 100%
not contaminated, whereas the probability of a SNP being an outlier is an unsupervised measure of
contamination. The only thing we need to be careful about is when a male sample is 100% contaminated
with maternal DNA, it's probability will also be 0, but these should be easy to spot if we look at
normalized_Y_intensity as well.

Here I look the top 20 male samples with the lowest probability of their SNPs being outliers, and 
calculate their average normalized Y intensity, this will represent our maximum value of the 
normalized range:

```{r}
# calculate minmum maximum of normalized range
pDat %>% 
  filter(Sex == 'M') %>% 
  arrange(Prob_SNP_outlier) %>% 
  select(Sample_Name, Sex, Prob_SNP_outlier, normalized_Y_intensity) %>% 
  
  # display top 30
  dplyr::slice(1:20) %>%
  as.data.frame() %>% {
    ggplot(., aes(x = Prob_SNP_outlier, y = normalized_Y_intensity)) +
      geom_smooth(method = 'lm') + geom_point() +
      geom_hline(., yintercept = mean(.$normalized_Y_intensity), 
                 linetype = 'dashed', color = 'red') +
      geom_text(x = min(.$Prob_SNP_outlier) + 0.0005, y = mean(.$normalized_Y_intensity) + 0.01,
                label = paste0('mean = ', formatC(mean(.$normalized_Y_intensity)), digits = 2), 
                col = 'red') +
      theme_bw()
}

max_Y <- pDat %>% 
  filter(Sex == 'M') %>% 
  arrange(Prob_SNP_outlier) %>%
  dplyr::slice(1:20) %>%
  pull(normalized_Y_intensity) %>%
  mean()
```

Now for 1., the minimum of our normalized range will be the average female y intensity:

```{r}
min_Y <- pDat %>% filter(Sex == 'F') %>%
  pull(normalized_Y_intensity) %>%
  mean()

min_Y
```

## Predict on females

To calculate our measure of contamination for females:

1. predict normalized_Y_intensity based on cor_to_reference using model trained in males
2. Normalize to 0 and 1, based on min and max
3. Samples falling outside of min and max are set to 0 and 1 respectively
4. Reverse the scale such that 1 is 100% maternal contamination and 0 is 0%

For females, we skip step 1, and do step 2. and 3.


```{r}
pDat <- pDat %>% 
  
  # 1.
  ## predict values for females
  mutate(maternal_contamination = predict(fit_ycor, pDat %>% select(cor_to_reference))) %>%
  
  ## set as normalized_y_intensity for males
  mutate(maternal_contamination = ifelse(Sex == 'M', normalized_Y_intensity, maternal_contamination)) %>%
  

  mutate(maternal_contamination_norm = case_when(
    
    #2.
    maternal_contamination > max_Y ~ 1,
    maternal_contamination < min_Y ~ 0,
    
    #3.
    TRUE ~ ((maternal_contamination - min_Y) / (max_Y - min_Y))
  )) %>%
  
  #4.
  mutate(maternal_contamination_norm_flip = 1-maternal_contamination_norm)

pDat %>%
  select(Sex, contains('maternal_contamination'), cor_to_reference, normalized_Y_intensity) %>%
  gather(key = step, value = Contamination, -Sex, -cor_to_reference) %>%
  mutate(step = factor(case_when(
    step == 'normalized_Y_intensity' ~ '1. Raw data',
    step == 'maternal_contamination' ~ '2. Predicted value for females\nTrained values for males',
    step == 'maternal_contamination_norm' ~ '3. Normalized and constrained to 0 and 1',
    step == 'maternal_contamination_norm_flip' ~ '4. Reversed, such that 1 = 100%'
  ), levels = c('1. Raw data', 
                '2. Predicted value for females\nTrained values for males',
                '3. Normalized and constrained to 0 and 1', 
                '4. Reversed, such that 1 = 100%'))) %>% 
  {
    ggplot(., aes(x = cor_to_reference, y = Contamination, color = Sex)) +
      geom_smooth(data = . %>% filter(step == '1. Raw data', Sex == 'M'), method = 'lm',
                  show.legend = F) +
      geom_abline(data = data.frame(slope = fit_ycor$coefficients[2], 
                                    intercept = fit_ycor$coefficients[1],
                                    step = c('2. Predicted value for females\nTrained values for males'),
                                    Sex = c('M')),
                  aes(slope = slope, intercept = intercept, color = Sex)) +
      geom_point(alpha = 0.7, size = 1.5) +
      facet_wrap(~step, ncol = 2) +
      theme_bw()
}

```

One thing to note is that there are a large number of samples that are estimated to have 
contamination > 50%

```{r}
pDat %>% arrange(Tissue, maternal_contamination_norm_flip) %>% 
  mutate(Sample_Name = factor(as.character(Sample_Name), levels = Sample_Name)) %>%
  gather(key = Contamination, value = value, maternal_contamination_norm_flip, Prob_SNP_outlier) %>%
  ggplot(aes(x = Sample_Name, y = value, color = Tissue)) +
  geom_point() + 
  geom_segment(aes(y = 0, yend = value, 
                   x = Sample_Name, xend = Sample_Name))+
  scale_color_brewer(palette = 'Set1') +
  scale_y_continuous(limits = c(0,1), expand = c(0,0), breaks = seq(0,1, 0.1)) +
  scale_x_discrete(expand = c(0.05, 0.05)) +
  facet_grid(Contamination~Trimester, scales = 'free_x', space = 'free_x') +
  theme_bw() +
  theme(axis.text.x = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.grid.major.x = element_blank()) 

ggplot(pDat, aes(y = maternal_contamination_norm_flip, x = Prob_SNP_outlier)) +
  geom_point() 
```

## Predict on mat blood

First I define the function of our trained model

```{r}
# trained parameters
intercept <- fit_ycor$coefficients[1]
slope <- fit_ycor$coefficients[2]

infer_maternal <- function(x) {
  
  # 1. predict normalized_Y_intensity based on cor_to_reference using model trained in males
  y <- slope*x + intercept
  
  # 2 & 3 Normalize to 0 and 1, based on min and max, constrain to min max
  y_norm <- 
    
    case_when(
      #2.
      y > max_Y ~ 1,
      y < min_Y ~ 0,
      
      #3.
      TRUE ~ ((y - min_Y) / (max_Y - min_Y))
    )
  
  # 4. Reverse the scale such that 1 is 100% maternal contamination and 0 is 0%
  y_norm_flip <- 1 - y_norm
  
  return(y_norm_flip)
}

# check that this matches our previous manual calculation
pDat %>% filter(Sex == 'F') %>%
  mutate(out = infer_maternal(cor_to_reference)) %>%
  select(out, maternal_contamination_norm_flip)
```

```{r}
ss <- readxl::read_xlsx('Z:/ROBLAB6 Infinium450k John/Master_Sample_Sheet.xlsx')

table(ss$Tissue)

ss_matched <- ss %>% 
  filter(Platform != '27k') %>%
  group_by(Case_ID) %>%
  filter(any(Tissue %in% c('Maternal Blood', 'Amnion', 'Chorion', 'Decidua')) & 
           any(Tissue == 'Villi')) %>%
  ungroup() %>% arrange(desc(Platform), Case_ID, Tissue) %>%
  
  # remove duplicated rows
  group_by(Basename) %>% 
  dplyr::slice(1) %>%
  ungroup()

rgset_matched_450k <- read.metharray.exp(targets = ss_matched %>% filter(Platform == '450k'))
rgset_matched_850k <- read.metharray.exp(targets = ss_matched %>% filter(Platform == '850k'))

snps_matched_450k <- getSnpBeta(rgset_matched_450k)
snps_matched_850k <- getSnpBeta(rgset_matched_850k)

colnames(snps_matched_450k) <- ss_matched %>% filter(Platform == '450k') %>% pull(Sample_Name)
colnames(snps_matched_850k) <- ss_matched %>% filter(Platform == '850k') %>% pull(Sample_Name)

# calculate for all pairwise comparisons
cor(snps_matched_450k)

res_450k <- cor(snps_matched_450k) %>% as_tibble() %>% 
  mutate(Sample1 = colnames(.)) %>% 
  
  # add case_ID and tissue
  left_join(ss_matched %>% select(Sample_Name, Case_ID, Tissue), by = c('Sample1' = 'Sample_Name')) %>%
  
  # longate the dataframe, each row is a single pairwise comparison
  gather(key = 'Sample2', value = 'Cor', -Sample1, -Case_ID, -Tissue) %>%
  
  # pData needs to refer to the right sample (1 or 2)
  dplyr::rename(Case_ID1 = Case_ID, Tissue1 = Tissue) %>%
  
  # add info for sample 2
  left_join(ss_matched %>% select(Sample_Name, Case_ID, Tissue), by = c('Sample2' = 'Sample_Name')) %>%
  dplyr::rename(Case_ID2 = Case_ID, Tissue2 = Tissue) %>% 
  
  # only interested in within-case comparisons, not to each other
  filter(Case_ID1 == Case_ID2, Sample1 != Sample2,
         Tissue1 == 'Villi', Tissue2 != 'Villi') %>%
  
  # for samples with more than one villi, take the mean correlation to all of theme
  group_by(Sample2) %>%
  summarize(cor_to_villi = mean(Cor)) %>%
  
  left_join(ss_matched %>% select(Sample_Name, Tissue, Case_ID), by = c('Sample2' = 'Sample_Name')) %>%
  
  # calculat maternal contam
  mutate(maternal_contamination = infer_maternal(cor_to_villi))

# repeat for 850k
res_850k <- cor(snps_matched_850k) %>% as_tibble() %>% 
  mutate(Sample1 = colnames(.)) %>% 
  left_join(ss_matched %>% select(Sample_Name, Case_ID, Tissue), by = c('Sample1' = 'Sample_Name')) %>%
  gather(key = 'Sample2', value = 'Cor', -Sample1, -Case_ID, -Tissue) %>%
  dplyr::rename(Case_ID1 = Case_ID, Tissue1 = Tissue) %>%
  left_join(ss_matched %>% select(Sample_Name, Case_ID, Tissue), by = c('Sample2' = 'Sample_Name')) %>%
  dplyr::rename(Case_ID2 = Case_ID, Tissue2 = Tissue) %>% 
  filter(Case_ID1 == Case_ID2, Sample1 != Sample2,
         Tissue1 == 'Villi', Tissue2 != 'Villi') %>%
  group_by(Sample2) %>%
  summarize(cor_to_villi = mean(Cor)) %>%
  left_join(ss_matched %>% select(Sample_Name, Tissue, Case_ID), by = c('Sample2' = 'Sample_Name')) %>%
  mutate(maternal_contamination = infer_maternal(cor_to_villi))

res <- bind_rows(res_850k, res_450k) 

res %>%
  ggplot(aes(x = Sample2, y = maternal_contamination, color = Tissue)) +
  geom_point(size = 3) +
  geom_segment(aes(y = 0, yend = maternal_contamination, 
                   x = Sample2, xend = Sample2), size = 1) +
  geom_text(data = . %>%
              mutate(label = formatC(maternal_contamination, 2, format = 'f')),
            aes(label = label), nudge_y = 0.05, show.legend = F) +
  theme_bw() + 
  coord_flip()

```

TODO KEEP THIS SEPARATE FROM DMC ANALYSIS


# 1.0 Global methylation

## Density distributions

Here we plot the density distributions of the betas across each tissue

Density for each sample:

```{r}
sample_densities <- pDat_first %>%
  select(Sample_Name, Tissue) %>%
  mutate(densities = apply(betas_first, 2, density)) %>%
  mutate(x = map(densities, 'x'),
         y = map(densities, 'y')) %>%
  select(-densities) %>%
  unnest()

sample_densities %>%
  left_join(pDat_first %>% select(Sample_Name, cor_to_reference)) %>%
  mutate(cor_to_reference95 = ifelse(cor_to_reference < 0.95, 'Contaminated', 'Not contaminated')) %>%
  ggplot(aes(x = x, y = y, color = Tissue, group = Sample_Name)) +
  geom_line(size = 1, alpha = 0.5) + theme_bw() +
  scale_x_continuous(breaks = c(0, 0.25, 0.5, 0.75, 1), labels = c(0, 25, 50, 75, 100))  +
  facet_grid(cor_to_reference95~Tissue) +
  scale_color_manual(values = color_code_tissue[unique(pDat_first$Tissue)]) +
  guides(colour = guide_legend(override.aes = list(alpha = 1))) +
  geom_text_repel(data = sample_densities %>% 
                    filter(Sample_Name == 'PM364_hofb_cs', x > 0.55, x < 0.6) %>%
                    dplyr::slice(1),
                  aes(x = x, y = y, label = Sample_Name), inherit.aes = F,
                  nudge_y = 1, nudge_x = -0.1) +
  labs(x = '% methylation', y = 'density')

ggplot(sample_densities, aes(x = x, y = y, color = Tissue, group = Sample_Name)) +
  geom_line(size = 1, alpha = 0.5) + theme_bw() +
  scale_x_continuous(breaks = c(0, 0.25, 0.5, 0.75, 1), labels = c(0, 25, 50, 75, 100))  +
  facet_wrap(~Tissue) +
  scale_color_manual(values = color_code_tissue[unique(pDat_first$Tissue)]) +
  guides(colour = guide_legend(override.aes = list(alpha = 1))) +
  geom_text_repel(data = sample_densities %>% 
                    filter(Sample_Name == 'PM364_hofb_cs', x > 0.55, x < 0.6) %>%
                    dplyr::slice(1),
                  aes(x = x, y = y, label = Sample_Name), inherit.aes = F,
                  nudge_y = 1, nudge_x = -0.1) +
  labs(x = '% methylation', y = 'density')
```

## PCA

```{r}
# compute pca
set.seed(1)
pca <- prcomp_irlba(t(betas_first), n = 20, center = T, scale = F)

# add pc scores to pdata
pca_scores <- pca$x[,1:20] %>% as.data.frame()
colnames(pca_scores) <- paste0(colnames(pca_scores), '_processed')
pDat_first <- pDat_first %>% 
  select(-contains('processed')) %>%
  bind_cols(pca_scores)

# create proportion variance explained data frame
pc_info <- summary(pca)$importance %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column(var = 'variable') %>%
  gather(key = 'PC', value = 'value', -variable) %>%
  as_tibble() %>%
  mutate(PC = factor(as.character(PC), levels = paste0('PC', 1:20)),
         Label = ifelse(variable == 'Proportion of Variance',
                        paste0(PC, ' (', prettyNum(value*100, digits = 2), '%)'),
                        as.character(PC))) %>%
  arrange(variable, PC)

  
# correlate PCs with phenodata
pc_cor <- lmmatrix(dep = pca$x[,1:20, drop = F],
                   ind = pDat_first %>%
                   dplyr::select(Case_ID, Tissue, Sex,#bio
                                 Week, Chip_number, Row_numeric, Row_factor, Batch_BSC, # batch
                                 DNA_loaded, 
                                 failed_probes,
                                 cor_to_reference, Prob_SNP_outlier),
                   metric = 'Pvalue')
# plot data
pc_cor <- pc_cor %>% as.data.frame() %>% 
  
  # add dep variables
  mutate(dep = rownames(pc_cor)) %>%
  
  # reshape
  gather(PC, pval, -dep) %>%
  
  # pvalue categories
  mutate(pval_cat = factor(case_when(
    pval > 0.05  ~ '> 0.05',
    pval < 0.05 & pval > 0.01 ~ '< 0.05',
    pval < 0.01 & pval > 0.001 ~ '< 0.01',
    pval < 0.001 ~ '< 0.001'
  ), levels = c('> 0.05', '< 0.05','< 0.01', '< 0.001')),
  
  # make PC is encoded with proper levels!!!
  PC = factor(PC, levels = paste0('PC', 1:20))) %>% as_tibble()

# create color palette
colpal <- c('white', '#fee8c8', '#fdbb84', '#e34a33')
names(colpal) <- levels(pc_cor$pval_cat)

# relevel
pc_cor <- pc_cor %>% 
  mutate(dep = factor(dep, levels = c('Case_ID', 'Tissue', 'Sex',
                                     'Week', 'Chip_number', 'Row_numeric', 'Row_factor', 
                                     'Batch_BSC', # batch
                                     'DNA_loaded',
                                     'failed_probes',
                                     'cor_to_reference', 'Prob_SNP_outlier')))
                                     
p7 <- ggplot(pc_cor, aes(x = PC, y = dep, fill = pval_cat)) +
  geom_tile(col = 'lightgrey') + theme_bw() +
  scale_x_discrete(expand = c(0, 0), labels = 1:20) +
  scale_y_discrete(expand = c(0, 0)) +
  scale_fill_manual(values = colpal)  + 
  labs(y = '', fill = 'P value')

p8 <- ggplot(pc_info %>% filter(variable == 'Proportion of Variance') %>%
               mutate(value = value*100), 
             aes(x = PC, y = value)) +
  geom_bar(stat = 'identity') +
  theme_bw() + 
  scale_x_discrete(expand = c(0, 0), labels = 1:20) + 
  scale_y_continuous(expand = c(0, 0)) +
  labs(y = '% variance explained')

egg::ggarrange(p7, p8, heights = c(3,1))


# create another without all variables
var_interest <- c('Tissue', 'Sex', 'cor_to_reference','Case_ID',
                  'Week', 'Chip_number', 'Row_numeric', 'Row_factor', 'Batch_BSC')

p9 <- pc_cor %>% 
  filter(dep %in% var_interest) %>%
  mutate(dep = ifelse(dep %in% var_interest[5:10], paste0('TECH_', dep), as.character(dep))) %>%
  mutate(dep = factor(dep, levels = c('TECH_Week', 'TECH_Chip_number', 'TECH_Row_numeric',
                                      'TECH_Row_factor', 'TECH_Batch_BSC', 
                                      'cor_to_reference',  'Sex',
                                      'Case_ID', 'Tissue'))) %>%
  ggplot(aes(x = PC, y = dep, fill = pval_cat)) +
  geom_tile(col = 'lightgrey') + theme_bw() +
  scale_x_discrete(expand = c(0, 0), labels = 1:20) +
  scale_y_discrete(expand = c(0, 0)) +
  scale_fill_manual(values = colpal)  + 
  labs(y = '', fill = 'P value')

egg::ggarrange(p9, p8, heights = c(3,1))
```


Scatterplots

```{r message = F}
scatter <- function(x, y, fill, point_size = 1){
  xlab <- pc_info %>% filter(variable == 'Proportion of Variance', PC == x) %>% pull(Label)
  ylab <- pc_info %>% filter(variable == 'Proportion of Variance', PC == y) %>% pull(Label)
  
  x <- paste0(x, '_processed')
  y <- paste0(y, '_processed')
  
  out <- ggplot(pDat_first, aes_string(x = x, y = y, fill = fill)) +
    geom_point(shape = 21, size = point_size) + theme_bw() + labs(x = xlab, y = ylab) 
  
  if (is.numeric(as.data.frame(pDat_first)[,fill])){
    out <- out +
      scale_fill_viridis()
  } else {
    out <- out + 
      scale_fill_brewer(palette = 'Set1')
  }
    
  out
}

scatter(x = 'PC1', y = 'PC2', fill = 'Tissue', point_size = 2) +
  scale_fill_manual(values = color_code_tissue[unique(pDat_first$Tissue)])  +
  labs(fill = '') +
  geom_text_repel(data = pDat_first %>% filter(Tissue == 'Hofbauer cs', PC1_processed > -50), 
            aes(label = Sample_Name))
scatter(x = 'PC1', y = 'PC2', fill = 'cor_to_reference', point_size = 2) 

scatter2 <- function(x) {
  xlab <- pc_info %>% filter(variable == 'Proportion of Variance', PC == x) %>% pull(Label)
  
  x <- paste0(x, '_processed')
  
  out <- ggplot(pDat_first, aes_string(x = x, y = 'cor_to_reference', fill = 'cor_to_reference')) +
            geom_point(shape = 21, size = 2) +
            theme_bw() +
            scale_fill_viridis() +
            facet_wrap(~Tissue, scales = 'free_x') +
            geom_smooth(method = 'lm') +
            stat_fit_glance(method = "lm", 
                            method.args = list(formula = y~x),
                            label.x = "left",
                            label.y = "top",
                            aes(label = sprintf('R^2~"="~%.3f~~italic(P)~"="~%.4f',
                                stat(r.squared), stat(p.value))),
                            parse = TRUE) +
            labs(x = xlab, y = 'Correlation to reference')

  out
}

scatter2(x = 'PC1')+
  geom_text_repel(data = pDat_first %>% filter(Tissue == 'Hofbauer cs', PC1_processed > -50), 
            aes(label = Sample_Name))
scatter2(x = 'PC2')


  
scatter(x = 'PC3', y = 'PC4', fill = 'Tissue', point_size = 2) +
  scale_fill_manual(values = color_code_tissue[unique(pDat_first$Tissue)]) +
  labs(fill = '')
scatter(x = 'PC3', y = 'PC4', fill = 'cor_to_reference', point_size = 2) 
scatter2(x = 'PC3')
scatter2(x = 'PC4')

scatter(x = 'PC5', y = 'PC6', fill = 'Tissue', point_size = 2) +
  scale_fill_manual(values = color_code_tissue[unique(pDat_first$Tissue)]) +
  labs(fill = '')
scatter(x = 'PC5', y = 'PC6', fill = 'cor_to_reference', point_size = 2) 
scatter2(x = 'PC5')
scatter2(x = 'PC6')
```

## Pairwise Correlation

```{r}
cor_betas <- cor(betas_first)

# annotation bar
annotation <- pDat_first %>%
  select(Tissue, Sex, cor_to_reference) %>% as.data.frame()
rownames(annotation) <- pDat_first$Sample_Name

#colors for heatmap
anno_colors <- list(
  Tissue = color_code_tissue[unique(pDat_first$Tissue)],
  Sex = setNames(color_code[[2]]$Colors_Sex, color_code[[2]]$Sex),
  cor_to_reference = viridis(n = 3, option = 'B')
)


# pheatmap
pheatmap(cor_betas, annotation_col = annotation, annotation_row = annotation[,'Tissue', drop = F],
         show_rownames = F, show_colnames = T,
         annotation_colors = anno_colors,
         labels_col = ifelse(rownames(cor_betas) == 'P131_hofb_cs', 'P131_hofb_cs', ''),
         color = viridis(100, option = 'B'),
         cutree_cols = 5)
```

# Remove P131_hofb_cs

Based 

```{r}
pDat_first <- pDat_first %>% filter(Sample_Name != 'P131_hofb_cs')
betas_first <- betas_first[,pDat_first$Sample_Name]
```

# Repeat above analysis without P131_hofb_cs

## Densities

```{r}

sample_densities <- pDat_first %>%
  select(Sample_Name, Tissue) %>%
  mutate(densities = apply(betas_first, 2, density)) %>%
  mutate(x = map(densities, 'x'),
         y = map(densities, 'y')) %>%
  select(-densities) %>%
  unnest()


ggplot(sample_densities, aes(x = x, y = y, color = Tissue, group = Sample_Name)) +
  geom_line(size = 1, alpha = 0.5) + theme_bw() +
  scale_x_continuous(breaks = c(0, 0.25, 0.5, 0.75, 1), labels = c(0, 25, 50, 75, 100))  +
  facet_wrap(~Tissue) +
  scale_color_manual(values = color_code_tissue[unique(pDat_first$Tissue)]) +
  guides(colour = guide_legend(override.aes = list(alpha = 1))) +
  geom_text_repel(data = sample_densities %>% 
                    filter(Sample_Name == 'PM364_hofb_cs', x > 0.55, x < 0.6) %>%
                    dplyr::slice(1),
                  aes(x = x, y = y, label = Sample_Name), inherit.aes = F,
                  nudge_y = 1, nudge_x = -0.1) +
  labs(x = '% methylation', y = 'density')
```

## PCA

```{r}
# compute pca
set.seed(1)
pca <- prcomp_irlba(t(betas_first), n = 20, center = T, scale = F)

# add pc scores to pdata
pca_scores <- pca$x[,1:20] %>% as.data.frame()
colnames(pca_scores) <- paste0(colnames(pca_scores), '_processed')
pDat_first <- pDat_first %>% 
  select(-contains('processed')) %>%
  bind_cols(pca_scores)

# create proportion variance explained data frame
pc_info <- summary(pca)$importance %>% 
  as.data.frame() %>% 
  tibble::rownames_to_column(var = 'variable') %>%
  gather(key = 'PC', value = 'value', -variable) %>%
  as_tibble() %>%
  mutate(PC = factor(as.character(PC), levels = paste0('PC', 1:20)),
         Label = ifelse(variable == 'Proportion of Variance',
                        paste0(PC, ' (', prettyNum(value*100, digits = 2), '%)'),
                        as.character(PC))) %>%
  arrange(variable, PC)

  
# correlate PCs with phenodata
pc_cor <- lmmatrix(dep = pca$x[,1:20, drop = F],
                   ind = pDat_first %>%
                   dplyr::select(Case_ID, Tissue, Sex,#bio
                                 Week, Chip_number, Row_numeric, Row_factor, Batch_BSC, # batch
                                 DNA_loaded, 
                                 failed_probes,
                                 cor_to_reference, Prob_SNP_outlier),
                   metric = 'Pvalue')
# plot data
pc_cor <- pc_cor %>% as.data.frame() %>% 
  
  # add dep variables
  mutate(dep = rownames(pc_cor)) %>%
  
  # reshape
  gather(PC, pval, -dep) %>%
  
  # pvalue categories
  mutate(pval_cat = factor(case_when(
    pval > 0.05  ~ '> 0.05',
    pval < 0.05 & pval > 0.01 ~ '< 0.05',
    pval < 0.01 & pval > 0.001 ~ '< 0.01',
    pval < 0.001 ~ '< 0.001'
  ), levels = c('> 0.05', '< 0.05','< 0.01', '< 0.001')),
  
  # make PC is encoded with proper levels!!!
  PC = factor(PC, levels = paste0('PC', 1:20))) %>% as_tibble()

# create color palette
colpal <- c('white', '#fee8c8', '#fdbb84', '#e34a33')
names(colpal) <- levels(pc_cor$pval_cat)

# relevel
pc_cor <- pc_cor %>% 
  mutate(dep = factor(dep, levels = c('Case_ID', 'Tissue', 'Sex',
                                     'Week', 'Chip_number', 'Row_numeric', 'Row_factor', 
                                     'Batch_BSC', # batch
                                     'DNA_loaded',
                                     'failed_probes',
                                     'cor_to_reference', 'Prob_SNP_outlier')))
                                     
p7 <- ggplot(pc_cor, aes(x = PC, y = dep, fill = pval_cat)) +
  geom_tile(col = 'lightgrey') + theme_bw() +
  scale_x_discrete(expand = c(0, 0), labels = 1:20) +
  scale_y_discrete(expand = c(0, 0)) +
  scale_fill_manual(values = colpal)  + 
  labs(y = '', fill = 'P value')

p8 <- ggplot(pc_info %>% filter(variable == 'Proportion of Variance') %>%
               mutate(value = value*100), 
             aes(x = PC, y = value)) +
  geom_bar(stat = 'identity') +
  theme_bw() + 
  scale_x_discrete(expand = c(0, 0), labels = 1:20) + 
  scale_y_continuous(expand = c(0, 0)) +
  labs(y = '% variance explained')

egg::ggarrange(p7, p8, heights = c(3,1))


# create another without all variables
var_interest <- c('Tissue', 'Sex', 'cor_to_reference','Case_ID',
                  'Week', 'Chip_number', 'Row_numeric', 'Row_factor', 'Batch_BSC')

p9 <- pc_cor %>% 
  filter(dep %in% var_interest) %>%
  mutate(dep = ifelse(dep %in% var_interest[5:10], paste0('TECH_', dep), as.character(dep))) %>%
  mutate(dep = factor(dep, levels = c('TECH_Week', 'TECH_Chip_number', 'TECH_Row_numeric',
                                      'TECH_Row_factor', 'TECH_Batch_BSC', 
                                      'cor_to_reference',  'Sex',
                                      'Case_ID', 'Tissue'))) %>%
  ggplot(aes(x = PC, y = dep, fill = pval_cat)) +
  geom_tile(col = 'lightgrey') + theme_bw() +
  scale_x_discrete(expand = c(0, 0), labels = 1:20) +
  scale_y_discrete(expand = c(0, 0)) +
  scale_fill_manual(values = colpal)  + 
  labs(y = '', fill = 'P value')

egg::ggarrange(p9, p8, heights = c(3,1))


scatter(x = 'PC1', y = 'PC2', fill = 'Tissue', point_size = 2) +
  scale_fill_manual(values = color_code_tissue[unique(pDat_first$Tissue)])  +
  labs(fill = '') +
  geom_text_repel(data = pDat_first %>% filter(Tissue == 'Hofbauer cs', PC1_processed > -50), 
            aes(label = Sample_Name))
scatter(x = 'PC1', y = 'PC2', fill = 'cor_to_reference', point_size = 2) 

scatter2(x = 'PC1')+
  geom_text_repel(data = pDat_first %>% filter(Tissue == 'Hofbauer cs', PC1_processed > -50), 
            aes(label = Sample_Name))
scatter2(x = 'PC2')


  
scatter(x = 'PC3', y = 'PC4', fill = 'Tissue', point_size = 2) +
  scale_fill_manual(values = color_code_tissue[unique(pDat_first$Tissue)]) +
  labs(fill = '')
scatter(x = 'PC3', y = 'PC4', fill = 'cor_to_reference', point_size = 2) 
scatter2(x = 'PC3')
scatter2(x = 'PC4')

scatter(x = 'PC5', y = 'PC6', fill = 'Tissue', point_size = 2) +
  scale_fill_manual(values = color_code_tissue[unique(pDat_first$Tissue)]) +
  labs(fill = '')
scatter(x = 'PC5', y = 'PC6', fill = 'cor_to_reference', point_size = 2) 
scatter2(x = 'PC5')
scatter2(x = 'PC6')
```

Effect of correlation to reference and stromal cs is not associated with any of the top 6 pcs, so
I will not adjust for it

# 2.0 Linear modelling

## With  vs w/o adjustment

..for maternal contamination,

Drop the villi replicates:

*	PL293_v_R1 - keep, less failed probes
* PL293_v_R2 - drop

```{r}
# drop replicate 
n1 <- nrow(pDat_first)
pDat_first_filt <- pDat_first %>% filter(Tissue != 'PL293_v_R2')
n2 <- nrow(pDat_first_filt)
print(paste(n1, 'samples filtered down to', n2))

betas_first_filt <- betas_first[,pDat_first$Sample_Name]
mvals_first_filt <- mvals_first[,pDat_first$Sample_Name]
ncol(betas_first_filt) == n2
ncol(mvals_first_filt) == n2

# create a design matrix
design <- pDat_first_filt  %>% select(Tissue, Case_ID, cor_to_reference) %>%
  mutate(Tissue = gsub('Endothelial cs', 'Endo_cs',
                       gsub('Hofbauer cs', 'Hofb_cs',
                            gsub('Trophoblasts cs', 'Troph_cs',
                                 gsub('Stromal cs', 'Strom_cs',
                                      gsub(':', '\\.', 
                                           Tissue))))))

# specify the interactions between contam and specific tissues
design <- model.matrix(~0 + Tissue+ cor_to_reference + Case_ID , 
                       data = design)

colnames(design) <- gsub(':', '.', gsub('Tissue', '', gsub('Case_ID', '', colnames(design)))) # rename columns
design <- design[,setdiff(colnames(design),'Strom_cs.cor_to_reference')]

# fit the linear model 
fit_m <- lmFit(mvals_first_filt, design)

# fit contrasts
contMatrix_1vAll <- makeContrasts(Endo_cs - (Hofb_cs + Strom_cs + Troph_cs)/3,
                            Hofb_cs - (Endo_cs + Strom_cs + Troph_cs)/3,
                            Strom_cs - (Endo_cs + Hofb_cs + Troph_cs)/3,
                            Troph_cs - (Endo_cs + Hofb_cs + Strom_cs)/3,
                            levels=design)

fit_m <- contrasts.fit(fit_m, contMatrix_1vAll)
fit_m <- eBayes(fit_m) # moderate t stats

# pull out dmcs
dmcs <- broom::tidy(fit_m) %>%
  separate(term, into = c('Group1', 'Group2'), sep = '\\s-\\s', remove = F) %>%
  group_by(Group1, Group2) %>%
  mutate(fdr = p.adjust(p.value, method = "fdr"),
         bonferroni = p.adjust(p.value, method = 'bonferroni')) %>%
  ungroup() %>%
  select(-term, -Group2)

# add delta betas
fit_b <- lmFit(betas_first_filt, design) %>% 
  contrasts.fit(contMatrix_1vAll) %>%
  eBayes() %>%
  tidy() %>%
  separate(term, into = c('Group1', 'Group2'), sep = '\\s-\\s', remove = F) %>%
  dplyr::rename(delta_b = estimate) %>%
  select(gene, Group1, delta_b)
  
dmcs <- dmcs %>% left_join(fit_b) 
```

Without adjustment:

```{r}
# repeat without adjusting for contamination
design <- pDat_first_filt  %>% select(Tissue, Case_ID, cor_to_reference) %>%
  mutate(Tissue = gsub('Endothelial cs', 'Endo_cs',
                       gsub('Hofbauer cs', 'Hofb_cs',
                            gsub('Trophoblasts cs', 'Troph_cs',
                                 gsub('Stromal cs', 'Strom_cs',
                                      gsub(':', '\\.', 
                                           Tissue))))))

design_noadj <- model.matrix(~0+Tissue + Case_ID, data = design)
colnames(design_noadj) <- gsub(':', '\\.', 
                               gsub('Tissue', '',
                                    gsub('Case_ID', '', colnames(design_noadj)))) # rename columns

contMatrix_1vAll_noadj <- makeContrasts(Endo_cs - (Hofb_cs + Strom_cs + Troph_cs)/3,
                            Hofb_cs - (Endo_cs + Strom_cs + Troph_cs)/3,
                            Strom_cs - (Endo_cs + Hofb_cs + Troph_cs)/3,
                            Troph_cs - (Endo_cs + Hofb_cs + Strom_cs)/3,
                            levels=design_noadj)

# fit the linear model 
fit_m_noadj <- lmFit(mvals_first_filt, design_noadj) %>%
  contrasts.fit(contMatrix_1vAll_noadj) %>%
  eBayes() %>%
  broom::tidy() %>%
  separate(term, into = c('Group1', 'Group2'), sep = '\\s-\\s') %>%
  group_by(Group1) %>%
  mutate(fdr = p.adjust(p.value, method = "fdr"),
         bonferroni = p.adjust(p.value, method = 'bonferroni'))  %>%
  ungroup() %>%
  select(-Group2)

# add delta betas
fit_b_noadj <- lmFit(betas_first_filt, design_noadj) %>%
  contrasts.fit(contMatrix_1vAll_noadj) %>%
  eBayes() %>%
  tidy() %>%
  separate(term, into = c('Group1', 'Group2'), sep = '\\s-\\s') %>%
  dplyr::rename(delta_b = estimate) %>%
  select(gene, Group1, delta_b)

dmcs_noadj <- fit_m_noadj %>% left_join(fit_b_noadj)
```

compare: 

```{r}
effects <- full_join(
  dmcs %>% group_by(Group1) %>%
    summarize(with_adj = sum(fdr < 0.01 & abs(delta_b) > 0.1)),
  dmcs_noadj %>% group_by(Group1) %>%
    summarize(without_adj = sum(fdr < 0.01 & abs(delta_b) > 0.1))) %>%
  gather(key = key, value = count, -Group1) 

effects %>% 
  ggplot(aes(x = key, y = count, color = Group1)) +
  geom_point() +
  geom_line(aes(group = Group1)) + theme_bw() +
  scale_x_discrete(labels = c('w/ adjustment', 'w/o adjustment')) +
  scale_y_continuous(breaks = seq(0,350000, 50000), labels = seq(0,350000, 50000)/10000) +
  labs(y = 'Number CpGs in 10,000s', x = '', title = 'Effect of adjusting for contamination')
```

# Save Data

```{r, eval = F}
saveRDS(dmcs, '../../data/main/interim/2_3_dmcs_with_adjustment.rds')
saveRDS(effects, '../../data/main/interim/2_3_effects_of_adjustment.rds')
```

# SessionInfo
`r sessioninfo()`



