---
title: "2_4_2_combined_first_third"
author: "Victor Yuan"
date: "30/07/2019"
output:
  html_document:
    keep_md: true
    toc: true
    toc_depth: 4
    toc_float:
      collapsed: false
    theme: spacelab
editor_options: 
  chunk_output_type: console
---

# Setup

## Libraries


```{r}
# libraries and data
library(tidyverse)
library(readxl)
library(annotatr)
library(fuzzyjoin)
library(formattable)
library(kableExtra)
```

## Data

```{r}
pDat <- readRDS('../../data/main/interim/2_3_pDat_contam.rds')
pDat <- pDat %>%
  mutate(Tissue = case_when(
    !(Tissue %in% c('Villi', 'Villi maternal', 'Syncytiotrophoblast')) ~ paste(Tissue, 'cs'),
    Tissue == 'Syncytiotrophoblast' ~ 'Trophoblasts enz',
    TRUE ~ Tissue
  )) 

# raw methylation data
betas <- readRDS('../../data/main/interim/1_4_betas_noob_filt.rds')

mset_noob <- readRDS('../../data/main/interim/1_4_mset_noob.rds') # for mvals
colnames(mset_noob) <- colnames(betas) <- pDat$Sample_Name
mvals <- getM(mset_noob)

probe_anno <- readRDS('../../data/main/interim/1_1_probe_anno.rds')

# color key
pheatmap_color_code <- readRDS('../../data/main/interim/1_1_color_code.rds')

color_code <- readRDS('../../data/main/interim/2_3_color_code.rds')
color_code_tissue <- setNames(color_code$Colors_Tissue, color_code$label)

# DMCs
dmcs <- readRDS('../../data/main/interim/2_4_dmcs.rds')

# annotation
anno <- readRDS('Z:/Victor/Repositories/EPIC_annotation/hg19_epic_annotation.rds')

#color code
color_code <- readRDS('../../data/main/interim/2_3_color_code.rds')
color_code_tissue <- setNames(color_code$Colors_Tissue, color_code$label)
```

## Remove samples

remove contamined and non-interesting samples

```{r}
pDat_filt <- pDat %>% 
  filter(maternal_contamination_norm_flip < 0.35,
         !Sample_Name %in% c('PM364_hofb_cs', 'PL293_v_R2', 'PM366_vc_R2', 'P131_hofb_cs'),
         !Tissue %in% c('Villi maternal', 'Trophoblasts enz', 'Mixture cs', 'Dead Cells and Lymphocytes cs'))

# filter to first trimester
mvals_filt <- mvals[rownames(betas),pDat_filt$Sample_Name]
betas_filt <- betas[,pDat_filt$Sample_Name]
```

## Test for genomic enrichment

The strategy is to first dummy variable every genomic element that we want to test enrichment for. And then join that to the linear modelling results data frame.

Dummy variables:  for exmaple one column might be something like "intron, intron, exon, intron, 3'UTR, intergene, ...,". Which will be transformed so that we have separate columns for each category: intron: "T, T, F, T, F, F, ...", exon: "F, F, T, F, F, F, ..." etc.

Once every feature is "dummy variable'd" and we join this to the linear modelling data, then we can calculate the numbers required to run a fisher's exact test.

Note that the input list of dmcs is a 'long' table, where there is 4 celltypes/contrasts * ~800000 tested cpgs. Technically, we only need the ones defined as significant, so for now conditionally calculate the number of cpgs that are differentially methylated. An improvement to this code would be to expect as input the list of DMCS, and then another list of the background cpgs.

This took a long time to figure how to do, so eventually I would like to see this code put somewhere public that others can contribute towards.

```{r}
p_thresh <- 0.01
b_thresh <- 0.25

# dummy variable all categories
annotation <- anno %>% 
  
  # mutate spread is clean, but slow. Can only work with those that have non-duplicated strings in each element
  mutate(var = 1) %>%
  spread(cpg_id, var, fill = 0) %>% 
  
  mutate(var = 1) %>%
  spread(chr, var, fill = 0) %>% 
  
  mutate(
         # cannot follow the spread strategy above, since we have numerous occurences in each element
         `1to5kb` = ifelse(grepl('1to5kb', genes_id), 1, 0),
         `3UTR` = ifelse(grepl('3UTR', genes_id), 1, 0),
         `5UTR` = ifelse(grepl('5UTR', genes_id), 1, 0),
         exon = ifelse(grepl("exon", genes_id), 1, 0),
         intergenic = ifelse(grepl('intergenic', genes_id), 1, 0),
         intron = ifelse(grepl('intron', genes_id), 1, 0),
         intronexonboundary = ifelse(grepl('intronexonboundary', genes_id), 1, 0),
         promoter = ifelse(grepl("promoter", genes_id), 1, 0),
         
         enhancer = !is.na(enhancers_id),
         pmd = !is.na(pmd_id),
         imprinted_gene_placenta = !is.na(imprinted_gene_placenta),
         imprinted_gene_general = !is.na(imprinted_gene_general),
         imprinted_dmr_general = !is.na(imprinted_dmr_general),
         imprinted_dmr_placenta = !is.na(imprinted_dmr_placenta))

# add gene annotation to dmcs
dmcs <- dmcs %>% left_join(annotation, by = c('gene' = 'cpg'))

# filter to all tested cpgs
annotation <- annotation %>%
  filter(cpg %in% dmcs$gene)
  
# tabulate the background frequency per genomic element
expected <-  annotation %>%
  dplyr::select(cpg, island:pmd, contains('imprint')) %>%
  gather(key = genomic_feature, value = present, -cpg) %>%
  
  group_by(genomic_feature) %>%
  summarize(Expected_n_in = sum(present), 
            
            Expected_n_out = nrow(annotation) - Expected_n_in,
            Expected_p_in = Expected_n_in/ nrow(annotation),
            Expected_p_out = Expected_n_out / nrow(annotation))

# tabulate observed frequency for significant cpgs
observed <- dmcs %>% 
  
  # calculate total dmcs per cell type
  group_by(Group1) %>%
  mutate(n_total = sum(bonferroni < p_thresh & abs(delta_b) > b_thresh )) %>%
  
  # filter to just dmcs
  filter(bonferroni < p_thresh, abs(delta_b) > b_thresh) %>%
  select(Group1, gene, island:pmd, contains('imprint'), n_total) %>%

  # calculated the number of dmcs in each feature, and then the number out
  gather(key = genomic_feature, value = present, -gene, -Group1, -n_total) %>%
  group_by(Group1, genomic_feature) %>%
  summarize(Observed_n_in = sum(present), 
            
            Observed_n_out = unique(n_total) - Observed_n_in,
            Observed_p_in = Observed_n_in / unique(n_total),
            Observed_p_out = Observed_n_out / unique(n_total))

# fisher's test for enrichment
# (1) # of DMCs on var1 (already calcualted, 'Freq')
# (2) # of non-DMCs on var1
# (3) # of DMCs not on var1
# (4) # of non-DMCs not on var1


# calculate (2) (3) (4), and test enrichment
tests <- observed %>%
  
  ungroup() %>%
  
  # add in all_cpgs datarfame for calculations
  left_join(expected, by = 'genomic_feature') %>%
  
  # calculate (2), (4)
  mutate(Observed_notDMC_in = Expected_n_in - Observed_n_in,   #(2) # non-DMCs in var1
         
         # for (4)
         Observed_notDMC_out = Expected_n_out - Observed_n_out) %>%  # of non-DMCs out of Var1
  
  # test enrichment
  rowwise() %>%
  mutate(fisher_in = list(matrix(c(Observed_n_in, Observed_notDMC_in, 
                                   Observed_n_out, Observed_notDMC_out),2,2))) %>%
  ungroup() %>%
  
  mutate(fisher_out = map(fisher_in, ~fisher.test(., alternative = 'greater')),
         p = map_dbl(fisher_out, 'p.value'),
         odds_ratio = map_dbl(fisher_out, 'estimate'),
         FDR = p.adjust(p, method = 'fdr'),
         bonferroni = p.adjust(p, method = 'bonferroni'),
         FDR01 = ifelse(FDR < 0.01, T, F)) %>% 
  select(FDR, FDR01, bonferroni, p, odds_ratio, everything())


tests

# clean
tests <- tests %>%
  separate(Group1, into = c('Trimester', 'Celltype'), sep = '\\.') %>%
  mutate(Celltype = case_when(
    Celltype == 'Endo_cs' ~ 'Endothelial cs',
    Celltype == 'Hofb_cs' ~ 'Hofbauer cs',
    Celltype == 'Strom_cs' ~ 'Stromal cs',
    Celltype == 'Troph_cs' ~ 'Trophoblasts cs')) 
  
# color code
colors <- color_code %>% 
  filter(label %in% c('Endothelial cs', 'Hofbauer cs', 'Stromal cs', 'Trophoblasts cs'))
colors <- setNames(colors$Colors_Tissue, unique(tests$Celltype))

# categorize genomic features
tests <- tests %>% 
  mutate(genomic_feature_category = case_when(
    grepl('chr', genomic_feature) ~ 'chr',
    genomic_feature %in% c('1to5kb', '3UTR', '5UTR', 'exon', 'intron', 
                           'intergenic', 'intronexonboundary', 'promoter') ~ 'gene',
    genomic_feature %in% c('island', 'shore', 'shelf', 'sea') ~ 'cpg_island',
    grepl('imprint', genomic_feature) ~ 'imprinting',
    genomic_feature == 'pmd' ~ 'pmd',
    genomic_feature == 'enhancer' ~ 'enhancer'
  )) 
```

## Visualize enrichment

```{r}
tests %>% 
  filter(genomic_feature_category == 'chr',
         !genomic_feature %in% c('chrX', 'chrY')) %>%
  mutate(genomic_feature = factor(genomic_feature, levels = paste0('chr', 1:22)),
         
         Observed_p_minus_expected = Observed_p_in - Expected_p_in,
    
    # linerange ymin ymax values
         ymin = pmin(Observed_p_in, Expected_p_in),
         ymax = pmax(Observed_p_in, Expected_p_in)) %>%
  
  ggplot() +
  geom_pointrange(aes(x = genomic_feature, y = Observed_p_in, ymin = ymin, ymax = ymax, color =Celltype),
           stat = 'identity', position = position_dodge(width = 0.75),
           fatten = 2) +
  scale_color_manual(values = colors)  + 
  facet_grid(Trimester ~ . , scales = 'free_x', space = 'free') +
  geom_errorbar(aes(x = genomic_feature, ymin = Expected_p_in, ymax = Expected_p_in, linetype = 
                      'Expected frequency'), 
                size = 1) +
  
  
  coord_flip() +
  scale_y_continuous(limits = c(0,0.15), expand = c(0,0)) +
  theme_bw(base_size = 14) + 
  theme(legend.position = 'right', legend.direction = 'vertical',
        legend.spacing.y = unit(-0.2, 'cm')) +
  
  # add significance
  geom_point(aes(x = genomic_feature, y = ymax + 0.001, group = Celltype, shape = FDR01),
             position = position_dodge(0.75), size = 4) +
  labs(color = '', x = '', y = 'Frequency', shape = '', linetype = '') +
  scale_shape_manual(values = c('TRUE' = '*', 'FALSE' = ''), na.translate = F, 
                     labels = c('', 'Enriched (FDR < 0.01)'))
```

```{r}
tests %>% 
  
  # genomic features
  filter(genomic_feature_category == 'chr',
         !genomic_feature %in% c('chrX', 'chrY')) %>%
  
  # prepare text and shape of table
  mutate(genomic_feature = factor(genomic_feature, levels = paste0('chr', 1:22)),
         key = paste0(Trimester , '_', Celltype)) %>%
  select(key, genomic_feature, Observed_p_in, Expected_p_in, FDR01) %>%
  
   # prepare colors
  group_by(genomic_feature) %>%
  mutate(
    value = ifelse(Observed_p_in > Expected_p_in,
                   cell_spec(Observed_p_in, 
                             background = spec_color(Observed_p_in, 
                                                     begin = 0.5, 
                                                     direction = 1)),
                   cell_spec(Observed_p_in, 
                             background = spec_color(Observed_p_in, 
                                                     begin = 0,
                                                     end = 0.5, direction = -1)))) %>%
  select(genomic_feature, key, value, Expected_p_in) %>%
  spread(key= key, value = value) %>%
  kable(escape = F) %>%
  kable_styling()
  
  
  
```

# temp garbage

PMDs

```{r}
# load Schroeder 2013 list of pmds
pmds <- read_xls('../../data/external/placental_pmds.xls', col_names = F)
colnames(pmds) <- c('chr', 'start', 'end')

# match with the list of pmds
dmcs <- dmcs %>% 
  dplyr::rename(cpg = gene) %>%
  mutate_at(c('start', 'end'), as.numeric) %>%
  genome_left_join(pmds) %>%
  mutate(pmd_id = ifelse(!is.na(chr.y), paste0(chr.y, ':', start.y, '-', end.y), NA),
         pmd_width = end.y-start.y) %>%
  dplyr::rename(chr = chr.x, start = start.x, end = end.x) %>%
  select(-contains('.y'))

p_thresh <- 0.01
b_thresh <- 0.25

# number of DMCs in pmds
pmds_count <- dmcs %>% 
  separate(Group1, into = c('Trimester', 'Celltype'), sep = '\\.') %>%
  mutate(Celltype = case_when(
    Celltype == 'Endo_cs' ~ 'Endothelial cs',
    Celltype == 'Hofb_cs' ~ 'Hofbauer cs',
    Celltype == 'Strom_cs' ~ 'Stromal cs',
    Celltype == 'Troph_cs' ~ 'Trophoblasts cs'
  )) %>%
  group_by(Trimester, Celltype) %>%
  summarize(dmcs_in_pmds = sum(bonferroni < p_thresh & abs(delta_b) > b_thresh & !is.na(pmd_id)),
            dmcs_total = sum(bonferroni < p_thresh & abs(delta_b) > b_thresh)) %>%
  mutate(dmcs_notin_pmds = dmcs_total - dmcs_in_pmds) %>%
  gather(key = type, value = count, -Trimester, -Celltype) %>%
  mutate(type = case_when(
    type == 'dmcs_in_pmds' ~ 'In PMD',
    type == 'dmcs_total' ~ 'Total',
    type == 'dmcs_notin_pmds' ~ 'Out of PMD'
  ))

pmds_count %>% 
  filter(type != 'Total') %>%
  group_by(Trimester, Celltype) %>%
  mutate(proportion = count*100/sum(count),
         proportion_label = paste0(prettyNum(proportion, digits = 2), '%')) %>%
  ggplot(aes(x = Celltype, y = proportion, fill = type)) +
  geom_col(position = 'stack') +
  geom_text(aes(label = proportion_label), position = position_stack(0.5)) +
  facet_wrap(vars(Trimester)) +
  coord_cartesian(expand = F) +
  theme(panel.spacing = unit(2, 'lines')) +
  labs(x = '', y = 'Percent of DMCs', fill = '')
```
